---
title: "DE"
author: "Zoe Dellaert"
date: "2024-10-05"
output:
  html_document: default
  github_document: default
---

## Differential expression analysis of LCM RNA Data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Managing Packages Using Renv

To run this code in my project using the renv environment, run the following lines of code

```{r, eval=FALSE}
install.packages("renv") #install the package on the new computer (may not be necessary if renv bootstraps itself as expected)
renv::restore() #reinstall all the package versions in the renv lockfile
```

## Load packages

```{r packages}
require("genefilter")
require("DESeq2")
require("apeglm")
require("ashr")
require("ggplot2")
require("vsn")
require("hexbin")
require("pheatmap")
require("RColorBrewer")
require("EnhancedVolcano")
require("tidyverse")

sessionInfo() #provides list of loaded packages and version of R.
```

```{r}
save_ggplot <- function(plot, filename, width = 10, height = 7, units = "in", dpi = 300) {
  # Display plot
  print(plot)
  
  # Save plot
  ggsave(filename = paste0(filename, ".png"), plot = plot, width = width, height = height, units = units, dpi = dpi)
}

# Specify colors
ann_colors = list(
    Tissue = c(OralEpi = "palegreen3" ,Aboral = "mediumpurple1")
)
```


## Read and clean count matrix and metadata 

Read in raw count data
```{r}
counts_raw <- read.csv("../output_RNA/stringtie-GeneExt/LCM_RNA_gene_count_matrix.csv", row.names = 1) #load in data
```

gene_id,LCM_15,LCM_16,LCM_20,LCM_21,LCM_26,LCM_27,LCM_4,LCM_5,LCM_8,LCM_9

Read in metadata 
```{r}
meta <- read.csv("../data_RNA/LCM_RNA_metadata.csv") %>%
            dplyr::arrange(Sample) %>%
            mutate(across(c(Tissue, Fragment, Section_Date, LCM_Date), factor)) # Set variables as factors 

meta$Tissue <- factor(meta$Tissue, levels = c("OralEpi","Aboral")) #we want OralEpi to be the baseline
```

Data sanity checks!
```{r}
all(meta$Sample %in% colnames(counts_raw)) #are all of the sample names in the metadata column names in the gene count matrix? Should be TRUE
all(meta$Sample == colnames(counts_raw)) #are they the same in the same order? Should be TRUE
```

## pOverA filtering to reduce dataset

```{r}
ffun<-filterfun(pOverA(0.5,10))  #Keep genes expressed in at least 50% of samples -
counts_filt_poa <- genefilter((counts_raw), ffun) #apply filter

filtered_counts <- counts_raw[counts_filt_poa,] #keep only rows that passed filter

cat("Number of genes after filtering:", sum(counts_filt_poa))

write.csv(filtered_counts, "../output_RNA/differential_expression/filtered_counts.csv")
```

There are now 14464 genes in the filtered dataset.

Data sanity checks:
```{r}
all(meta$Sample %in% colnames(filtered_counts)) #are all of the sample names in the metadata column names in the gene count matrix?
all(meta$Sample == colnames(filtered_counts))  #are they the same in the same order? Should be TRUE
```

## [DESeq2](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

Create DESeq object and run DESeq2
```{r}
dds <- DESeqDataSetFromMatrix(countData = filtered_counts,
                              colData = meta,
                              design= ~ Fragment + Tissue)

dds <- DESeq(dds)
```

 ### Extract results for Aboral vs. OralEpi contrast

```{r}
res <- results(dds, contrast = c("Tissue","Aboral","OralEpi"))
resLFC <- lfcShrink(dds, coef="Tissue_Aboral_vs_OralEpi", res=res, type = "apeglm")
```

### Extract results for adjusted p-value < 0.05

```{r}
res <- resLFC

resOrdered <- res[order(res$pvalue),]# save differentially expressed genes

DE_05 <- as.data.frame(resOrdered) %>% filter(padj < 0.05)
DE_05_Up <- DE_05 %>% filter(log2FoldChange > 0) #Higher in Aboral, Lower in OralEpi
DE_05_Down <- DE_05 %>% filter(log2FoldChange < 0) #Lower in Aboral, Higher in OralEpi

nrow(DE_05)
nrow(DE_05_Up) #Higher in Aboral, Lower in OralEpi
nrow(DE_05_Down) #Lower in Aboral, Higher in OralEpi
```

```{r}
write.csv(as.data.frame(resOrdered), 
          file="../output_RNA/differential_expression/DESeq_results.csv")

write.csv(DE_05, 
          file="../output_RNA/differential_expression/DEG_05.csv")
```

### Visualizing Differential Expression

```{r}
EnhancedVolcano(resLFC, 
    lab = rownames(resLFC),
    x = 'log2FoldChange',
    y = 'pvalue')
```

## Plots
```{r}
plotMA(results(dds, contrast = c("Tissue","Aboral","OralEpi")), ylim=c(-20,20))
plotMA(resLFC, ylim=c(-20,20))
```

### Log2 Fold Change Comparison

```{r}
# because we are interested in the comparison and not the intercept, we set 'coef=2'
resNorm <- lfcShrink(dds, coef=6, type="normal")
resAsh <- lfcShrink(dds, coef=6, type="ashr")

par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-20,20)
plotMA(resLFC, xlim=xlim, ylim=ylim, main="apeglm")
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
plotMA(resAsh, xlim=xlim, ylim=ylim, main="ashr")
```

```{r}
plotCounts(dds, gene=which.max(res$log2FoldChange), intgroup="Tissue")
plotCounts(dds, gene=which.min(res$log2FoldChange), intgroup="Tissue")
```

Transforming count data for visualization 

```{r}
vsd <- vst(dds, blind=FALSE)
rld <- rlog(dds, blind=FALSE)
ntd <- normTransform(dds) # this gives log2(n + 1)

meanSdPlot(assay(vsd), main = "vsd")
meanSdPlot(assay(rld))
meanSdPlot(assay(ntd))
```

Will move forward with vst transformation for visualizations

## Heatmap of count matrix

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])

#view all genes
pheatmap(assay(vsd), cluster_rows=TRUE, show_rownames=FALSE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)),
         annotation_colors = ann_colors,color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200))

#view highest count genes
select <- order(rowMeans(counts(dds,normalized=TRUE)),
                decreasing=TRUE)[1:20]

pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)),
         annotation_colors = ann_colors, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200))


#view most significantly differentially expressed genes

select <- order(res$padj)[1:20]

pheatmap(assay(vsd)[select,], cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2, annotation_col=(df%>% select(Tissue)),
         annotation_colors = ann_colors,color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200))
```

## Heatmap of the sample-to-sample distances
```{r}
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Tissue, vsd$Fragment, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

## Principal component plot of the samples

```{r}
pcaData <- plotPCA(vsd, intgroup=c("Tissue", "Fragment"), returnData=TRUE, ntop = 14464)

percentVar <- round(100 * attr(pcaData, "percentVar"))
PCA <- ggplot(pcaData, aes(PC1, PC2, color=Tissue, shape=Fragment)) +
  geom_point(size=2) +
  scale_color_manual(values = c("Aboral" = "mediumpurple1", "OralEpi" = "palegreen3"))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + theme_bw()

save_ggplot(PCA, "../output_RNA/differential_expression/PCA_allgenes")

PCA_small <- ggplot(pcaData, aes(PC1, PC2, color=Tissue)) +
  geom_point(size=2) +
  scale_color_manual(values = c("Aboral" = "mediumpurple1", "OralEpi" = "palegreen3"))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + theme_bw()

ggsave(filename = paste0("../output_RNA/differential_expression/PCA_allgenes_small", ".png"), plot = PCA_small, width = 4, height = 2.5, units = "in", dpi = 300)
```

```{r}
pcaData <- plotPCA(vsd, intgroup=c("Tissue", "Fragment"), returnData=TRUE)

percentVar <- round(100 * attr(pcaData, "percentVar"))
PCA <- ggplot(pcaData, aes(PC1, PC2, color=Tissue, shape=Fragment)) +
  geom_point(size=2) +
  scale_color_manual(values = c("Aboral" = "mediumpurple1", "OralEpi" = "palegreen3"))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + theme_bw()

save_ggplot(PCA, "../output_RNA/differential_expression/PCA")

PCA_small <- ggplot(pcaData, aes(PC1, PC2, color=Tissue)) +
  geom_point(size=2) +
  scale_color_manual(values = c("Aboral" = "mediumpurple1", "OralEpi" = "palegreen3"))+
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() + theme_bw()

ggsave(filename = paste0("../output_RNA/differential_expression/PCA_small", ".png"), plot = PCA_small, width = 4, height = 2.5, units = "in", dpi = 300)
```

Clearly, the majority of the variance in the data is explained by tissue type!

## Annotation data

Download annotation files from genome website
```{bash, eval=FALSE}

# wget files
wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.Conserved_Domain_Search_results.txt.gz

wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.EggNog_results.txt.gz

wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.KEGG_results.txt.gz

# move to references direcotry
mv *gz ../references

# unzip files
gunzip ../references/*gz
```

```{r}
EggNog <- read.delim("../references/Pocillopora_acuta_HIv2.genes.EggNog_results.txt") %>% dplyr::rename("query" = X.query)

CDSearch <- read.delim("../references/Pocillopora_acuta_HIv2.genes.Conserved_Domain_Search_results.txt", quote = "") %>% dplyr::rename("query" = X.Query)

KEGG <- read.delim("../references/Pocillopora_acuta_HIv2.genes.KEGG_results.txt", header = FALSE) %>% dplyr::rename("query" = V1, "KeggTerm" = V2)
```

```{r}
DE_05$query <- rownames(DE_05)
DE_05_annot <- DE_05 %>% left_join(CDSearch) %>% select(query,everything())
DE_05_eggnog <- DE_05 %>% left_join(EggNog) %>% select(query,everything())

write.csv(as.data.frame(DE_05_eggnog), 
          file="../output_RNA/differential_expression/DE_05_eggnog_annotation.csv")

annot_all <- as.data.frame(rownames(dds)) %>% dplyr::rename("query" = `rownames(dds)`) %>% left_join(CDSearch)

eggnog_all <- as.data.frame(rownames(dds)) %>% dplyr::rename("query" = `rownames(dds)`) %>% left_join(EggNog)
```

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])
gene_labels <- eggnog_all %>% select(query,PFAMs) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) %>% #replace NAs with "" for labelling purposes
  separate(PFAMs, into = c("PFAMs", "rest of name"), sep = ",(?=.*?,)", extra = "merge")
  
#view most significantly differentially expressed genes

select <- order(res$padj)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[select,"PFAMs"], fontsize_row = 5)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/top50_DE")

#view genes Higher in Aboral, Lower in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange,decreasing = TRUE)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_Aboral <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[select,"PFAMs"], fontsize_row = 5)
up_Aboral
save_ggplot(up_Aboral, "../output_RNA/differential_expression/up_Aboral")

#view genes Lower in Aboral, Higher in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_OralEpi <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[select,"PFAMs"], fontsize_row = 5)
up_OralEpi
save_ggplot(up_OralEpi, "../output_RNA/differential_expression/up_OralEpi")
```

## Genes of Interest

```{r}
MarkerGenes <- read.csv("../references/Pacuta_MarkerGenes_Levy2021.csv") %>% dplyr::rename("query" = 1, "List" = 2, "definition" = 3) %>% filter(List !="Toolkit")

MarkerGenes$def_short <- ifelse(nchar(MarkerGenes$definition) > 20, 
                            paste0(substr(MarkerGenes$definition, 1, 17), "..."), 
                            MarkerGenes$definition)
```

```{r}
Biomin <- read.csv("../references/Pacuta_Biomin.csv") %>% dplyr::rename("query" = Pocillopora_acuta_best_hit) %>% select(-c(accessionnumber.geneID, Ref))

Biomin <- Biomin %>%
  group_by(query,List) %>%
  summarize(definition = paste(unique(definition), collapse = ","))

Biomin$def_short <- ifelse(nchar(Biomin$definition) > 40, 
                            paste0(substr(Biomin$definition, 1, 37), "..."), 
                            Biomin$definition)
```



```{r}
DE_05$query <- rownames(DE_05)
resOrdered$query <- rownames(resOrdered)


DE_05_biomin <- DE_05 %>% left_join(Biomin) %>% select(query,everything()) %>% drop_na()
DESeq_biomin <- as.data.frame(resOrdered) %>% left_join(Biomin) %>% select(query,everything()) %>% drop_na()
DE_05_marker <- DE_05 %>% left_join(MarkerGenes) %>% select(query,everything())  %>% drop_na()

write.csv(as.data.frame(DE_05_biomin), 
          file="../output_RNA/differential_expression/DE_05_biomin_annotation.csv")

write.csv(as.data.frame(DE_05_marker), 
          file="../output_RNA/differential_expression/DE_05_markergene_annotation.csv")

biomin_all_counts <- as.data.frame(counts(dds)) %>% mutate(query = rownames(dds)) %>% select(query,everything()) %>% left_join(Biomin) 
biomin_all_res <- as.data.frame(resLFC) %>% mutate(query = rownames(resLFC)) %>% select(query,everything()) %>% left_join(Biomin) 

markers_all_counts <- as.data.frame(counts(dds)) %>% mutate(query = rownames(dds)) %>% select(query,everything()) %>% left_join(MarkerGenes) 
markers_all_res <- as.data.frame(resLFC) %>% mutate(query = rownames(resLFC)) %>% select(query,everything()) %>% left_join(MarkerGenes) 
```

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])

#view biomin genes that are differntially expressed

z_scores <- t(scale(t(assay(vsd)[DE_05_biomin$query, ]), center = TRUE, scale = TRUE))
DE_biomin <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = DE_05_biomin$def_short, fontsize_row = 5)
DE_biomin
save_ggplot(DE_biomin, "../output_RNA/differential_expression/DE_biomin")

#view marker genes that are differntially expressed

z_scores <- t(scale(t(assay(vsd)[DE_05_marker$query, ]), center = TRUE, scale = TRUE))
DE_marker <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,cutree_rows = 5,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = DE_05_marker$List, fontsize_row = 4)
DE_marker
save_ggplot(DE_marker, "../output_RNA/differential_expression/DE_marker")

DE_05_marker_grouped <- DE_05_marker %>% arrange(List) %>% mutate(List = as.factor(List))

z_scores <- t(scale(t(assay(vsd)[DE_05_marker_grouped$quer, ]), center = TRUE, scale = TRUE))
DE_05_marker_grouped_plot <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = DE_05_marker_grouped$List, fontsize_row = 5)

DE_05_marker_grouped_plot
save_ggplot(DE_05_marker_grouped_plot, "../output_RNA/differential_expression/DE_05_marker_grouped")
```

### Visualizing Differential Expression

```{r}
Biomin_volcano <- EnhancedVolcano(biomin_all_res, 
    lab = biomin_all_res$def_short,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.01,
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    pointSize = 1,
    labSize = 2,boxedLabels = TRUE,max.overlaps = 40)

save_ggplot(Biomin_volcano, "../output_RNA/differential_expression/Biomin_volcano")

Marker_volcano <- EnhancedVolcano(markers_all_res, 
    lab = markers_all_res$List,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.01,
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    pointSize = 1,
    labSize = 2,boxedLabels = TRUE,max.overlaps = 60)

save_ggplot(Marker_volcano, "../output_RNA/differential_expression/Marker_volcano")

Marker_volcano_names <- EnhancedVolcano(markers_all_res, 
    lab = markers_all_res$def_short,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.01,
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    pointSize = 1,
    labSize = 2,boxedLabels = TRUE,max.overlaps = 60)

save_ggplot(Marker_volcano_names, "../output_RNA/differential_expression/Marker_volcano_names")


EnhancedVolcano(resLFC, 
    lab = rownames(resLFC),
    x = 'log2FoldChange',
    y = 'pvalue')

volcano_plain <- EnhancedVolcano(resLFC, 
    lab = NA,
    x = 'log2FoldChange',
    y = 'padj',
    pCutoff = 0.05,
    title="",
    subtitle="",
    drawConnectors = TRUE,
    widthConnectors = 0.75,
    pointSize = 1,
    labSize = 2,boxedLabels = TRUE,max.overlaps = 60)

save_ggplot(volcano_plain, "../output_RNA/differential_expression/volcano_plain",width = 4, height = 6, units = "in", dpi = 300)
save_ggplot(volcano_plain, "../output_RNA/differential_expression/volcano_plain")
```

## Trying to annotate un-annotated DE genes:

```{bash, eval = FALSE}
# wget protein sequence reference file
wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.pep.faa.gz

# move to references direcotry
mv *gz ../references

# unzip files
gunzip ../references/*gz

#get the names of all the DEGs from the first column of the DEG csv file
tail -n +2 ../output_RNA/differential_expression/DEG_05.csv | cut -d',' -f1 | tr -d '"' > ../output_RNA/differential_expression/DEG_05_names.csv

#grep this file against the protein fasta file, first with wc -l to make sure the number of lines is correct (should be your number of DEGs)
grep -f ../output_RNA/differential_expression/DEG_05_names.csv ../references/Pocillopora_acuta_HIv2.genes.pep.faa | wc -l

#grep each header with the protein sequence after ("-A 1") and save to a new file
grep  -A 1 -f ../output_RNA/differential_expression/DEG_05_names.csv ../references/Pocillopora_acuta_HIv2.genes.pep.faa > ../output_RNA/differential_expression/DEG_05_seqs.txt
```

### nr database BLAST of DE_05 genes only

On andromeda:

Blastp-ing only the DE genes against the entire nr database (will take a while)

```{bash eval=FALSE}
cd ../scripts
nano DEG_05_blast.sh
```

```{bash eval=FALSE}
#!/bin/bash
#SBATCH --job-name="DE_blast"
#SBATCH -t 240:00:00
#SBATCH --export=NONE
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails
#SBATCH --mail-user=zdellaert@uri.edu #your email to send notifications
#SBATCH --mem=500GB
#SBATCH --error=../scripts/outs_errs/"%x_error.%j" #write out slurm error reports
#SBATCH --output=../scripts/outs_errs/"%x_output.%j" #write out any program outpus
#SBATCH --account=putnamlab
#SBATCH --nodes=2 --ntasks-per-node=24

module load BLAST+/2.15.0-gompi-2023a

cd ../output_RNA/differential_expression #set working directory
mkdir blast
cd blast

#nr database location andromeda: /data/shared/ncbi-db/.ncbirc 
# points to current location: cat /data/shared/ncbi-db/.ncbirc
# [BLAST]
# BLASTDB=/data/shared/ncbi-db/2024-11-10

blastp -query ../DEG_05_seqs.txt -db nr -out DEG_05_blast_results.txt -outfmt 0 -evalue 1E-05 \
-num_threads 48 \
-max_target_seqs 10

blastp -query ../DEG_05_seqs.txt -db nr -out DEG_05_blast_results_tab.txt -outfmt 6 -evalue 1E-05 \
-num_threads 48 \
-max_target_seqs 1 \
-max_hsps 1

echo "Blast complete!" $(date)
```

```{bash, eval=FALSE}
sbatch DEG_05_blast.sh
```

```{bash, eval=FALSE}
cd ../output_RNA/differential_expression/blast

wc -l DEG_05_blast_results_tab.txt #3537 of the 3606 genes were annotated

#get just the NCBI database accession numbers for the blast results
cut -f2 DEG_05_blast_results_tab.txt > DEG_05_blast_accessions.txt

#remove any duplicates
sort -u  DEG_05_blast_accessions.txt > unique_DEG_05_blast_accessions.txt

wc -l unique_DEG_05_blast_accessions.txt #3404 of the 3537 annotations were unique

while read acc; do
  curl -s "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=$acc&rettype=gp&retmode=text" \
  | grep "DEFINITION" | sed 's/DEFINITION  //g' | awk -v id="$acc" '{print id "\t" $0}'
done < unique_DEG_05_blast_accessions.txt > DEG_05_blast_names.txt

wc -l DEG_05_blast_names.txt #3396 ; unsure why 8 are missing.

join -1 2 -2 1 -t $'\t' <(sort -k2 DEG_05_blast_results_tab.txt) <(sort DEG_05_blast_names.txt) > annotated_DEG_05_blast_results_tab.txt
```

### SwissProt of ALL genes

On unity:

swissprot based on https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/20-Apul-gene-annotation.Rmd and
https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/20-Pmea-gene-annotation.Rmd

Steven's notebook post here: https://sr320.github.io/tumbling-oysters/posts/sr320-27-go/

```{bash, eval=FALSE}
mkdir ../references/blast_dbs
```

```{bash, eval=FALSE}
cd ../references/blast_dbs
curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
mv uniprot_sprot.fasta.gz uniprot_sprot_r2024_10_02.fasta.gz
gunzip -k uniprot_sprot_r2024_10_02.fasta.gz
rm uniprot_sprot_r2024_10_02.fasta.gz

head uniprot_sprot_r2024_10_02.fasta
echo "Number of Sequences"
grep -c ">" uniprot_sprot_r2024_10_02.fasta
# 572214 sequences
```

```{bash, eval=FALSE}
module load blast-plus/2.14.1

makeblastdb \
-in ../references/blast_dbs/uniprot_sprot_r2024_10_02.fasta \
-dbtype prot \
-out ../references/blast_dbs/uniprot_sprot_r2024_10_02
```

```{bash, eval=FALSE}
cd ../scripts
nano blastp_SwissProt.sh
```

```{bash eval=FALSE}
#!/bin/bash
#SBATCH -t 18:00:00
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=48
#SBATCH --mem=500GB
#SBATCH --export=NONE
#SBATCH --error=../scripts/outs_errs/"%x_error.%j" #write out slurm error reports
#SBATCH --output=../scripts/outs_errs/"%x_output.%j" #write out any program outpus
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails
#SBATCH -D /project/pi_hputnam_uri_edu/zdellaert/LaserCoral #set working directory

module load blast-plus/2.14.1

cd references/
mkdir annotation

fasta="Pocillopora_acuta_HIv2.genes.pep.faa"

blastp \
-query $fasta \
-db blast_dbs/uniprot_sprot_r2024_10_02 \
-out annotation/blastp_SwissProt_out.tab \
-evalue 1E-05 \
-num_threads 48 \
-max_target_seqs 1 \
-max_hsps 1 \
-outfmt 6

echo "Blast complete!" $(date)
```

```{bash, eval=FALSE}
cd references/annotation/

tr '|' '\t' <  blastp_SwissProt_out.tab >  blastp_SwissProt_out_sep.tab
```

```{bash, eval=FALSE}
cd ../references/annotation/

curl -H "Accept: text/plain; format=tsv" "https://rest.uniprot.org/uniprotkb/stream?fields=accession%2Creviewed%2Cid%2Cprotein_name%2Cgene_names%2Corganism_name%2Clength%2Cgo_p%2Cgo%2Cgo_id%2Cgo_c%2Cgo_f&format=tsv&query=%28reviewed%3Atrue%29" -o SwissProt-Annot-GO_111524.tsv

wc -l SwissProt-Annot-GO_111524.tsv
#572215
```

All code below based on https://github.com/urol-e5/deep-dive/blob/main/D-Apul/code/20-Apul-gene-annotation.Rmd and
https://github.com/urol-e5/deep-dive/blob/main/F-Pmea/code/20-Pmea-gene-annotation.Rmd

Steven's notebook post here: https://sr320.github.io/tumbling-oysters/posts/sr320-27-go/

```{r}
bltabl <- read.csv("../references/annotation/blastp_SwissProt_out_sep.tab", sep = '\t', header = FALSE)

spgo <- read.csv("../references/annotation/SwissProt-Annot-GO_111524.tsv", sep = '\t', header = TRUE)
```

```{r}
annot_tab <- left_join(bltabl, spgo, by = c("V3" = "Entry")) %>%
  select(
    query = V1,
    blast_hit = V3,
    evalue = V13,
    ProteinNames = Protein.names,
    BiologicalProcess = Gene.Ontology..biological.process.,
    GeneOntologyIDs = Gene.Ontology.IDs
  )

head(annot_tab)
```

```{r}
write.table(annot_tab, 
            file = "../references/annotation/protein-GO.tsv", 
            sep = "\t", 
            row.names = FALSE, 
            quote = FALSE)

DE_05_SwissProt <- DE_05 %>% left_join(annot_tab) %>% select(query,everything()) 

write.csv(as.data.frame(DE_05_SwissProt), 
          file="../output_RNA/differential_expression/DE_05_SwissProt_annotation.csv")
```

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])

DE_05_SwissProt$short_name <- ifelse(nchar(DE_05_SwissProt$ProteinNames) > 30, 
                            paste0(substr(DE_05_SwissProt$ProteinNames, 1, 27), "..."), 
                            DE_05_SwissProt$ProteinNames)

gene_labels <- DE_05_SwissProt %>% 
  select(query,short_name) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

#view most significantly differentially expressed genes

select <- order(res$padj)[1:50]
rownames(res)[select]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 6)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/top50_DE_SwissProt")

#view most significantly differentially expressed genes by LFC

select <- order(abs(res$log2FoldChange),decreasing = TRUE)[1:50]
rownames(res)[select]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 6)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/top50_LFC_DE_SwissProt")


#view genes Higher in Aboral, Lower in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange,decreasing = TRUE)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_Aboral <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 5)
up_Aboral
save_ggplot(up_Aboral, "../output_RNA/differential_expression/up_Aboral_SwissProt")

#view genes Lower in Aboral, Higher in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_OralEpi <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 5)
up_OralEpi
save_ggplot(up_OralEpi, "../output_RNA/differential_expression/up_OralEpi_SwissProt")
```

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])

DE_05_SwissProt$short_GO <- ifelse(nchar(DE_05_SwissProt$BiologicalProcess) > 30, 
                            paste0(substr(DE_05_SwissProt$BiologicalProcess, 1, 27), "..."), 
                            DE_05_SwissProt$BiologicalProcess)

gene_labels <- DE_05_SwissProt %>% select(query,short_GO) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

#view most significantly differentially expressed genes

select <- order(res$padj)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 5)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/top50_DE_Blast2GO")

#view genes Higher in Aboral, Lower in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange,decreasing = TRUE)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_Aboral <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 5)
up_Aboral
save_ggplot(up_Aboral, "../output_RNA/differential_expression/up_Aboral_Blast2GO")

#view genes Lower in Aboral, Higher in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_OralEpi <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row =gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 5)
up_OralEpi
save_ggplot(up_OralEpi, "../output_RNA/differential_expression/up_OralEpi_Blast2GO")
```



### Combined annotations manually

```{r}
Manual <- read.csv("../output_RNA/differential_expression/DE_05_Manual_annotation.csv") %>% dplyr::rename("query" = 2, "definition" = 3) %>% arrange(X) 
```

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])

gene_labels <- Manual %>% 
  select(query,Heatmap_Label) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

#view most significantly differentially expressed genes

select <- order(res$padj)[1:50]
rownames(res)[select]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 6)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/top50_DE_Manual")

#view most significantly differentially expressed genes by LFC

select <- order(abs(res$log2FoldChange),decreasing = TRUE)[1:50]
rownames(res)[select]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 6)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/top50_LFC_DE_Manual")


#view genes Higher in Aboral, Lower in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange,decreasing = TRUE)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_Aboral <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), labels_col = NA, annotation_colors = ann_colors,
         labels_row =gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 8.5)
up_Aboral
save_ggplot(up_Aboral, "../output_RNA/differential_expression/up_Aboral_Manual")

#view genes Lower in Aboral, Higher in OralEpi, ordered by log2FoldChange
select <- order(res$log2FoldChange)[1:50]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
up_OralEpi <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=FALSE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), labels_col = NA, annotation_colors = ann_colors,
         labels_row =gene_labels[match(rownames(res)[select],(gene_labels$query)),2], fontsize_row = 8.5)
up_OralEpi
save_ggplot(up_OralEpi, "../output_RNA/differential_expression/up_OralEpi_Manual")
```

## Expression of certain genes of interest: blast against P. acuta genome

### yin yang Transctiption factor

```{bash, eval=FALSE}
cd ../references

#download the genome protein fasta if you have not already
wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.pep.faa.gz

#unzip file
gunzip Pocillopora_acuta_HIv2.genes.pep.faa.gz
```

In unity

```{bash, eval=FALSE}
salloc -p cpu -c 8 --mem 32G

module load uri/main
module load BLAST+/2.15.0-gompi-2023a

#make blast_dbs directory if you haven't done so above
mkdir blast_dbs
cd blast_dbs

makeblastdb -in ../Pocillopora_acuta_HIv2.genes.pep.faa -out Pacuta_prot -dbtype prot

cd ../../output_RNA/differential_expression/

#make blast output directory if you haven't done so above
mkdir blast
cd blast

nano YinYang.txt
```

add accession numbers of interest:

- nematostella vectensis transcriptional repressor protein YY1 isoforms
  - XP_048585772.1
  - XP_048585773.1
  - XP_048585774.1

- human transcription factor YY2
  - NP_996806.2

- human transcription factor ZFP41 (REX-1)
  - NP_777560.2

```{bash, eval = FALSE}
XP_048585772.1
XP_048585773.1
XP_048585774.1
NP_996806.2
NP_777560.2
```

```{bash, eval = FALSE}
# Read the input file line by line and fetch FASTA sequences
while read -r accession; do
  if [[ -n "$accession" ]]; then
    echo "Fetching $accession..."
    curl -s "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=${accession}&rettype=fasta&retmode=text" >> "YinYang.fasta"
    echo >> "YinYang.fasta"  # Add a newline between sequences
    sleep 1  # Avoid hitting rate limits
  fi
done < "YinYang.txt"

# run blast with human-readable output
blastp -query YinYang.fasta -db ../../../references/blast_dbs/Pacuta_prot -out YinYang_blast_results.txt -outfmt 0

#looks like there are a lot of matches for each gene! I am going to do a tab search with a very low e-value cutoff:

# run blast with tabular output
blastp -query YinYang.fasta -db ../../../references/blast_dbs/Pacuta_prot -out YinYang_blast_results_tab.txt -outfmt 6 -evalue 1e-25
```

Great! Interestingly, the same gene, Pocillopora_acuta_HIv2___RNAseq.g25242.t1 is the top match for all 5 proteins I searched. Pocillopora_acuta_HIv2___TS.g24434.t1 is the second best match for all 5. That is interesting! Pocillopora_acuta_HIv2___RNAseq.g7583.t1 is also worth looking at. And Pocillopora_acuta_HIv2___TS.g21338.t1 matched all three YY1 isoforms.

```{r}
YinYangs <- c("Pocillopora_acuta_HIv2___RNAseq.g25242.t1",
              "Pocillopora_acuta_HIv2___TS.g24434.t1",
              "Pocillopora_acuta_HIv2___RNAseq.g7583.t1",
              "Pocillopora_acuta_HIv2___TS.g21338.t1")

for (i in 1:length(YinYangs)){
  plotCounts(dds, gene=YinYangs[i], intgroup=c("Tissue"),)
}

as.data.frame(resOrdered)[YinYangs,]
```

Okay! None of these genes are differentially expressed between the tissues. That is interesting and good to know. Pocillopora_acuta_HIv2___RNAseq.g25242.t1 has the highest basal expression of all the potential isoforms of this transcription factor.


### TRP Channels

```{r}
df <- as.data.frame(colData(dds)[,c("Tissue","Fragment")])

DE_05_SwissProt$short_name <- ifelse(nchar(DE_05_SwissProt$ProteinNames) > 80, 
                            paste0(substr(DE_05_SwissProt$ProteinNames, 1, 77), "..."), 
                            DE_05_SwissProt$ProteinNames)

gene_labels <- DE_05_SwissProt %>% 
  select(query,short_name) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

gene_labels <- Manual %>% 
  select(query,Heatmap_Label) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

TRP <- Manual %>% filter(grepl("transient", ProteinNames, ignore.case = TRUE)) 
select <- TRP$query

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select,(gene_labels$query)),2], fontsize_row = 12)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/TRP_DE_SwissProt", width = 6.43, height = 4.25, units = "in", dpi = 300)

annot_tab$short_name <- ifelse(nchar(annot_tab$ProteinNames) > 80, 
                            paste0(substr(annot_tab$ProteinNames, 1, 77), "..."), 
                            annot_tab$ProteinNames)

gene_labels <- annot_tab  %>% 
  select(query,short_name) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

TRP <- annot_tab %>% filter(grepl("transient", ProteinNames, ignore.case = TRUE))
select1 <- TRP$query
select <- match(select1,rownames(vsd))
select <- select[!is.na(select)]

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select1,gene_labels$query),2], fontsize_row = 6)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/TRP_SwissProt")
```

### biomin de heatmap nice
```{r}

gene_labels <- Manual %>% 
  select(query,Heatmap_Label) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes


DE_05_biomin_filtered <- DE_05_biomin %>% left_join(Manual,by="query" ) %>% filter(!(Heatmap_Label %in% c("Myosin-9", "Actin, cytoplasmic")))
  
select <- DE_05_biomin_filtered$query
z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
DE_biomin <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select,(gene_labels$query)),2], fontsize_row = 8, cutree_rows = 2)
DE_biomin
save_ggplot(DE_biomin, "../output_RNA/differential_expression/DE_biomin")

# Perform clustering
row_clusters <- hclust(dist(z_scores))
cluster_assignments <- cutree(row_clusters, k = 2) # Adjust k as needed

# Create a dataframe to manage clusters and reordering
clustered_data <- data.frame(
  query = select,
  Heatmap_Label = gene_labels[match(select, gene_labels$query), 2],
  cluster = cluster_assignments
)

# Reorder rows within each cluster alphabetically by their labels
clustered_data <- clustered_data %>%
  arrange(cluster, Heatmap_Label)

# Reorder the z_scores matrix and labels based on the new order
z_scores <- z_scores[match(clustered_data$query, rownames(z_scores)), ]
ordered_labels <- clustered_data$Heatmap_Label

# Generate heatmap with reordered rows and labels
DE_biomin <- pheatmap(
  z_scores,
  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
  cluster_rows = FALSE,  # Disable clustering since rows are pre-ordered
  cluster_cols = TRUE,
  show_rownames = TRUE,
  cutree_cols = 2,
  annotation_col = (df %>% select(Tissue)),
  annotation_colors = ann_colors,
  labels_row = ordered_labels,
  fontsize_row = 8
)
save_ggplot(DE_biomin, "../output_RNA/differential_expression/clusters_clean/DE_biomin")
```
###wnt and hox

```{r}
##
WNT <- Manual %>% filter(grepl("wnt", Heatmap_Label, ignore.case = TRUE)|grepl("frizzle", Heatmap_Label, ignore.case = TRUE)|grepl("homeobox", Heatmap_Label, ignore.case = TRUE)|grepl("hox", Heatmap_Label, ignore.case = TRUE))#|grepl("wnt", BiologicalProcess, ignore.case = TRUE))
select <- WNT$query

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select,(gene_labels$query)),2], fontsize_row = 8, cutree_rows = 2)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/WNT_Hox_DE_SwissProt")

# Perform clustering
# Perform clustering
row_clusters <- hclust(dist(z_scores))
cluster_assignments <- cutree(row_clusters, k = 2) # Adjust k for the number of clusters

# Create a dataframe for clustering and label management
clustered_data <- data.frame(
  query = select,
  Heatmap_Label = gene_labels[match(select, gene_labels$query), 2],
  cluster = cluster_assignments
)

# Reorder rows within each cluster alphabetically by their labels
clustered_data <- clustered_data %>%
  arrange(cluster, Heatmap_Label)

# Reorder the z_scores matrix and labels based on the new order
z_scores <- z_scores[match(clustered_data$query, rownames(z_scores)), ]
ordered_labels <- clustered_data$Heatmap_Label

# Ensure there is only 1 gap between clusters
#row_breaks <- which(cluster_assignments != cluster_assignments[1])  # Only where cluster switches

# Generate the heatmap with reordered rows and labels
top50_DE <- pheatmap(
  z_scores,
  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
  cluster_rows = FALSE,  # Disable clustering since rows are pre-ordered
  cluster_cols = TRUE,
  cutree_cols = 2,
  annotation_col = (df %>% select(Tissue)),
  annotation_colors = ann_colors,
  labels_row = ordered_labels,
  fontsize_row = 8,
  borders_color = "white",  # Optional border around clusters
  cutree_rows = 2,  # Keep row clustering if desired
  gaps_row = 24  # Add breaks only between clusters
)


# Save the heatmap
save_ggplot(top50_DE, "../output_RNA/differential_expression/clusters_clean/WNT_Hox_DE_SwissProt")

```

### bacteria genes

```{r}
mucin <- Manual %>% filter(grepl("mucin", ProteinNames, ignore.case = TRUE)|grepl("toll-", ProteinNames, ignore.case = TRUE)|grepl("ZP", ProteinNames, ignore.case = TRUE)|grepl("lectin", ProteinNames, ignore.case = TRUE)|grepl("nitric", Heatmap_Label, ignore.case = TRUE)) %>% filter(Heatmap_Label !="Cnidocyte marker protein (Collectin-11)")
select <- mucin$query

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select,(gene_labels$query)),2], fontsize_row = 8, cutree_rows = 2)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/bacteria_DE_SwissProt")


# Perform clustering
row_clusters <- hclust(dist(z_scores))
cluster_assignments <- cutree(row_clusters, k = 2) # Adjust k for the number of clusters

# Create a dataframe for clustering and label management
clustered_data <- data.frame(
  query = select,
  Heatmap_Label = gene_labels[match(select, gene_labels$query), 2],
  cluster = cluster_assignments
)

# Reorder rows within each cluster alphabetically by their labels
clustered_data <- clustered_data %>%
  arrange(cluster, Heatmap_Label)

# Reorder the z_scores matrix and labels based on the new order
z_scores <- z_scores[match(clustered_data$query, rownames(z_scores)), ]
ordered_labels <- clustered_data$Heatmap_Label

# Generate the heatmap with reordered rows and labels
mucins <- pheatmap(
  z_scores,
  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
  cluster_rows = FALSE,  # Disable clustering since rows are pre-ordered
  cluster_cols = TRUE,
  cutree_cols = 2,
  annotation_col = (df %>% select(Tissue)),
  annotation_colors = ann_colors,
  labels_row = ordered_labels,
  fontsize_row = 8,
  borders_color = "white",  # Optional border around clusters
  cutree_rows = 2,  # Keep row clustering if desired
  gaps_row = c(24,38)  # Add breaks only between clusters
)

# Save the heatmap
save_ggplot(mucins, "../output_RNA/differential_expression/clusters_clean/bacteria_DE_SwissProt")
```



### Symbiont related transporters

```{r}
transport <- Manual %>% filter(grepl("amino acid transport", BiologicalProcess, ignore.case = TRUE)|grepl("ammonium transmembrane transport", BiologicalProcess, ignore.case = TRUE))
select <- transport$query

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select,(gene_labels$query)),2], fontsize_row = 8, cutree_rows = 2)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/transport_DE_SwissProt")

# Perform clustering
row_clusters <- hclust(dist(z_scores))
cluster_assignments <- cutree(row_clusters, k = 2) # Adjust k for the number of clusters

# Create a dataframe for clustering and label management
clustered_data <- data.frame(
  query = select,
  Heatmap_Label = gene_labels[match(select, gene_labels$query), 2],
  cluster = cluster_assignments
)

# Reorder rows within each cluster alphabetically by their labels
clustered_data <- clustered_data %>%
  arrange(cluster, Heatmap_Label)

# Reorder the z_scores matrix and labels based on the new order
z_scores <- z_scores[match(clustered_data$query, rownames(z_scores)), ]
ordered_labels <- clustered_data$Heatmap_Label

# Generate the heatmap with reordered rows and labels
transporters_heat <- pheatmap(
  z_scores,
  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
  cluster_rows = FALSE,  # Disable clustering since rows are pre-ordered
  cluster_cols = TRUE,
  cutree_cols = 2,
  annotation_col = (df %>% select(Tissue)),
  annotation_colors = ann_colors,
  labels_row = ordered_labels,
  fontsize_row = 8,
  borders_color = "white",  # Optional border around clusters
  cutree_rows = 2,  # Keep row clustering if desired
  gaps_row = c(16)  # Add breaks only between clusters
)

# Save the heatmap
save_ggplot(transporters_heat, "../output_RNA/differential_expression/clusters_clean/transport_DE_SwissProt")


```

### temp and light sensing

```{r}
Manual <- read.csv("../output_RNA/differential_expression/DE_05_Manual_annotation.csv") %>% dplyr::rename("query" = 2, "definition" = 3) %>% arrange(X) 

gene_labels <- Manual %>% 
  select(query,Heatmap_Label) %>%
  mutate_all(~ ifelse(is.na(.), "", .)) #replace NAs with "" for labelling purposes

sensors <- Manual %>% filter(grepl("TRP", Heatmap_Label, ignore.case = TRUE)|grepl("cellular response to light stimulus", BiologicalProcess, ignore.case = TRUE)|grepl("detection of mechanical stimulus involved in sensory perception", BiologicalProcess, ignore.case = TRUE))


select <- sensors$query

z_scores <- t(scale(t(assay(vsd)[select, ]), center = TRUE, scale = TRUE))
top50_DE <- pheatmap(z_scores, color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200), cluster_rows=TRUE, show_rownames=TRUE,
         cluster_cols=TRUE, cutree_cols = 2,annotation_col=(df%>% select(Tissue)), annotation_colors = ann_colors,
         labels_row = gene_labels[match(select,(gene_labels$query)),2], fontsize_row = 8, cutree_rows = 3)
top50_DE
save_ggplot(top50_DE, "../output_RNA/differential_expression/sensors_DE_SwissProt")

# Perform clustering
row_clusters <- hclust(dist(z_scores))
cluster_assignments <- cutree(row_clusters, k = 3) # Adjust k for the number of clusters

# Create a dataframe for clustering and label management
clustered_data <- data.frame(
  query = select,
  Heatmap_Label = gene_labels[match(select, gene_labels$query), 2],
  cluster = cluster_assignments
)

# Reorder rows within each cluster alphabetically by their labels
clustered_data <- clustered_data %>%
  mutate(cluster = factor(cluster, levels = c(1, 3, 2))) %>%
  arrange(cluster, Heatmap_Label)

# Reorder the z_scores matrix and labels based on the new order
z_scores <- z_scores[match(clustered_data$query, rownames(z_scores)), ]
ordered_labels <- clustered_data$Heatmap_Label

# Generate the heatmap with reordered rows and labels
transporters_heat <- pheatmap(
  z_scores,
  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
  cluster_rows = FALSE,  # Disable clustering since rows are pre-ordered
  cluster_cols = TRUE,
  cutree_cols = 2,
  annotation_col = (df %>% select(Tissue)),
  annotation_colors = ann_colors,
  labels_row = ordered_labels,
  fontsize_row = 8,
  borders_color = "white",  # Optional border around clusters
  cutree_rows = 2,  # Keep row clustering if desired
  gaps_row = c(10,18)  # Add breaks only between clusters
)

# Save the heatmap
save_ggplot(transporters_heat, "../output_RNA/differential_expression/clusters_clean/sensors_DE_SwissProt")

```

### random
```{r}
# Create annotation data frame
df <- as.data.frame(colData(dds)[, c("Tissue", "Fragment")])

# Function to generate short names for proteins
generate_short_name <- function(data) {
  data %>%
    mutate(short_name = ifelse(nchar(ProteinNames) > 60, 
                               paste0(substr(ProteinNames, 1, 57), "..."), 
                               ProteinNames))
}

# Function to create gene labels
create_gene_labels <- function(data) {
  data %>%
    select(query, short_name) %>%
    mutate_all(~ ifelse(is.na(.), "", .)) # Replace NAs with ""
}

# Function to generate z-scores for selected genes
calculate_z_scores <- function(data, selection, vsd_matrix) {
  selected_genes <- match(selection, rownames(vsd_matrix))
  selected_genes <- selected_genes[!is.na(selected_genes)] # Remove NAs
  t(scale(t(assay(vsd_matrix)[selected_genes, ]), center = TRUE, scale = TRUE))
}

# Function to create and save heatmap
create_heatmap <- function(z_scores, labels_row, annotation_col, annotation_colors, output_path) {
  heatmap <- pheatmap(z_scores, 
                      color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
                      cluster_rows = TRUE, 
                      show_rownames = TRUE,
                      cluster_cols = TRUE, 
                      cutree_cols = 2,
                      annotation_col = annotation_col,
                      annotation_colors = annotation_colors,
                      labels_row = labels_row, 
                      fontsize_row = 6)
  save_ggplot(heatmap, output_path)
}

# Process DE_05_SwissProt
DE_05_SwissProt <- generate_short_name(DE_05_SwissProt)
gene_labels <- create_gene_labels(DE_05_SwissProt)
TRP <- DE_05_SwissProt %>% filter(grepl("transient", ProteinNames, ignore.case = TRUE))
select <- TRP$query
z_scores <- calculate_z_scores(DE_05_SwissProt, select, vsd)
create_heatmap(z_scores, 
               labels_row = gene_labels[match(select, gene_labels$query), 2], 
               annotation_col = df %>% select(Tissue), 
               annotation_colors = ann_colors, 
               output_path = "../output_RNA/differential_expression/TRP_DE_SwissProt")

# Process annot_tab
annot_tab <- generate_short_name(annot_tab)
gene_labels <- create_gene_labels(annot_tab)
TRP <- annot_tab %>% filter(grepl("transient", ProteinNames, ignore.case = TRUE))
TRP <- left_join(TRP, as.data.frame(resOrdered)) %>% filter(!is.na(log2FoldChange))
select1 <- TRP$query
z_scores <- calculate_z_scores(annot_tab, select1, vsd)
create_heatmap(z_scores, 
               labels_row = TRP$short_name, 
               annotation_col = df %>% select(Tissue), 
               annotation_colors = ann_colors, 
               output_path = "../output_RNA/differential_expression/TRP_SwissProt")


significant <- ifelse(row.names(z_scores) %in% DE_05$query, "Significant", "Not Significant")

# Add this as a new annotation
row_annotation <- data.frame(Significance = significant)
rownames(row_annotation) <- rownames(z_scores)
row_annotation_colors <- list(Significance = c("Significant" = "red", "Not Significant" = "grey"))

heatmap <- pheatmap(z_scores,  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(200),
                      cluster_rows = TRUE, 
                      show_rownames = TRUE,
                      cluster_cols = TRUE, 
                      cutree_cols = 2,
                    cutree_rows = 5,
                      annotation_col = df %>% select(Tissue),
                      annotation_colors = ann_colors,
                    annotation_row = row_annotation,
          annotation_row_colors = row_annotation_colors,
                      labels_row = TRP$short_name, 
                      fontsize_row = 12)
  save_ggplot(heatmap, "../output_RNA/differential_expression/TRP_SwissProt")
  
# Process DE_05_SwissProt
DE_05_SwissProt <- generate_short_name(DE_05_SwissProt)
gene_labels <- create_gene_labels(DE_05_SwissProt)
GFP <- DE_05_SwissProt %>% filter(grepl("biolum", BiologicalProcess, ignore.case = TRUE))
select <- GFP$query
z_scores <- calculate_z_scores(DE_05_SwissProt, select, vsd)
create_heatmap(z_scores, 
               labels_row = gene_labels[match(select, gene_labels$query), 2], 
               annotation_col = df %>% select(Tissue), 
               annotation_colors = ann_colors, 
               output_path = "../output_RNA/differential_expression/GFP_DE_SwissProt")

# Process annot_tab
annot_tab <- generate_short_name(annot_tab)
gene_labels <- create_gene_labels(annot_tab)
GFP <- annot_tab %>% filter(grepl("biolum", BiologicalProcess, ignore.case = TRUE))
select1 <- GFP$query
z_scores <- calculate_z_scores(annot_tab, select1, vsd)
create_heatmap(z_scores, 
               labels_row = gene_labels[match(select1, gene_labels$query), 2], 
               annotation_col = df %>% select(Tissue), 
               annotation_colors = ann_colors, 
               output_path = "../output_RNA/differential_expression/GFP_SwissProt")

```


### Aboral Oral Larvae Paper - Pdam genome

#### Join our data to Pdam annotations

```{r load_annotation}
Annot_Pdam <- read.csv("../references/annotation/blastp_Pdam_out.tab", sep = '\t', header = FALSE) %>% select(c(1,2)) %>% dplyr::rename("protein_id" = "V2", "query" = "V1")
```

```{r}
Manual_Pdam <- left_join(Manual, Annot_Pdam) 
```

```{r}
library(readxl)
larval_aboral_enriched <- read_excel("~/Downloads/media-2.xlsx", sheet = "pocillopora_aboral_enriched_05_")

Manual_Pdam_upaboral <- Manual_Pdam %>% filter(log2FoldChange > 0)
inner_join(larval_aboral_enriched, Manual_Pdam_upaboral, by=c("gene ID"="protein_id")) %>% dim()

#18 genes overlap from the 126 in the paper

larval_oral_enriched <- read_excel("~/Downloads/media-2.xlsx", sheet = "pocillopora_oral_enriched_05_lf")

Manual_Pdam_uporal <- Manual_Pdam %>% filter(log2FoldChange < 0)
inner_join(larval_oral_enriched, Manual_Pdam_uporal, by=c("gene ID"="protein_id")) %>% dim()

inner_join(larval_oral_enriched, Manual_Pdam_uporal, by=c("gene ID"="protein_id")) %>% View()

#only 6/83 genes overlap


larval_aboral_clusters <- read_excel("~/Downloads/media-4.xlsx") %>% filter(!is.na(pocillopora))

larval_aboral_clusters_Pacuta <- inner_join(larval_aboral_clusters, Manual_Pdam, by=c("pocillopora"="protein_id"))

#my genes upregulated in oral tissue are high in mucous cells, cool
```




## Updating Renv environment:

After youve confirmed your code works as expected, use renv::snapshot() to record the packages and their sources in the lockfile.

```{r, eval=FALSE}
renv::snapshot()
```
