---
title: "KEGG-Pathways"
author: "Zoe Dellaert"
date: "2024-12-09"
output: html_document
---

## KEGG Pathway analysis of LCM RNA Data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Managing Packages Using Renv

To run this code in my project using the renv environment, run the following lines of code

```{r, eval=FALSE}
install.packages("renv") #install the package on the new computer (may not be necessary if renv bootstraps itself as expected)
renv::restore() #reinstall all the package versions in the renv lockfile
```

## Load packages

```{r packages}
require("pathview")
require("KEGGREST")
require("rtracklayer")
require("tidyverse")

sessionInfo() #provides list of loaded packages and version of R.
```

## Load in differential expression data

```{r load_data}
DESeq_allgenes <- read.csv("../output_RNA/differential_expression/DESeq_results.csv", header = TRUE) %>% dplyr::rename("query" ="X")
```

## BLAST Pacuta genes to Pdam genome (in bash)

```{bash eval=FALSE}
cd ../scripts
nano Pdam_blast.sh
```

```{bash eval=FALSE}
#!/bin/bash
#SBATCH --job-name="Pdam_blast"
#SBATCH -t 240:00:00
#SBATCH --export=NONE
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails
#SBATCH --mail-user=zdellaert@uri.edu #your email to send notifications
#SBATCH --mem=100GB
#SBATCH --error=../scripts/outs_errs/"%x_error.%j" #write out slurm error reports
#SBATCH --output=../scripts/outs_errs/"%x_output.%j" #write out any program outpus
#SBATCH --nodes=2 --ntasks-per-node=24

module load BLAST+/2.15.0-gompi-2023a

cd ../references

makeblastdb -in Pdam_RefSeq_protein.faa -out blast_dbs/Pdam_prot -dbtype prot

blastp -query Pocillopora_acuta_HIv2.genes.pep.faa -db blast_dbs/Pdam_prot -out annotation/blastp_Pdam_out.tab -outfmt 6 -evalue 1E-05 \
-num_threads 48 \
-max_target_seqs 1 \
-max_hsps 1

echo "Blast complete!" $(date)
```

```{bash, eval=FALSE}
sbatch Pdam_blast.sh
```

## Join our data to Pdam annotations

```{r load_annotation}
Pdam_gtf <- import("../references/Pdam_RefSeq.gtf")
Pdam_gtf_proteins <- as.data.frame(Pdam_gtf) %>%
                        filter(type == "CDS") %>% 
                        select(protein_id, gene_id) %>% 
                        distinct()

Annot_Pdam <- read.csv("../references/annotation/blastp_Pdam_out.tab", sep = '\t', header = FALSE) %>% 
                        select(c(1,2)) %>%
                        dplyr::rename("protein_id" = "V2", "query" = "V1")

Annot_Pdam <- left_join(Annot_Pdam, Pdam_gtf_proteins)
```

Join to all genes data:
```{r join_all}
allgenes_Pdam <- left_join(DESeq_allgenes, Annot_Pdam)

# remove rows for duplicate values in the column gene_id, keep one with largest abs(log2foldchange)

allgenes_Pdam <- allgenes_Pdam %>%
  group_by(gene_id) %>%
  slice_max(order_by = abs(log2FoldChange), n = 1) %>%
  ungroup() %>% drop_na() %>% 
  mutate(numeric_gene = gsub("LOC", "", gene_id) %>% as.numeric())

LFC_allgenes <- setNames(allgenes_Pdam$log2FoldChange, allgenes_Pdam$numeric_gene)
```

Join to DE data:
```{r join_de}
DE_05_Pdam <- allgenes_Pdam %>% filter(padj <0.05 & abs(log2FoldChange) > 1)

DE_05_LFC <- setNames(DE_05_Pdam$log2FoldChange, DE_05_Pdam$numeric_gene)
```

What pathways are common in our DE genes?
```{r de_pathways}
pdam_pathways <- keggLink("pathway", "pdam")
pdam_pathways <- data.frame(pdam_pathways, names(pdam_pathways)) 
pdam_pathways <- pdam_pathways %>% dplyr::rename("numeric_gene" = "names.pdam_pathways.")
pdam_pathways$numeric_gene <- gsub("pdam:", "", pdam_pathways$numeric_gene) %>% as.numeric()

DE_05_Pdam_pathways <- left_join(DE_05_Pdam, pdam_pathways, relationship = "many-to-many") %>% drop_na()

DE_05_Pdam_pathways_count <- DE_05_Pdam_pathways %>% dplyr::count(pdam_pathways) %>% arrange(-n)

head(DE_05_Pdam_pathways_count)

DE_05_Pdam_pathways_count <- left_join(allgenes_Pdam, pdam_pathways, relationship = "many-to-many") %>% drop_na() %>% dplyr::count(pdam_pathways) %>% arrange(-n)

head(DE_05_Pdam_pathways_count)

UpOral_Pdam_pathways_count <- DE_05_Pdam_pathways %>% filter(log2FoldChange <0) %>% dplyr::count(pdam_pathways) %>% arrange(-n)
UpAboral_Pdam_pathways_count <- DE_05_Pdam_pathways %>% filter(log2FoldChange >0) %>% dplyr::count(pdam_pathways) %>% arrange(-n)

head(UpOral_Pdam_pathways_count)
head(UpAboral_Pdam_pathways_count)
```

## Visualize KEGG Pathways of interest

### Pathways of interest:

1. Wnt signaling pathway - https://www.genome.jp/entry/pathway+pdam04310
2. mTOR signaling pathway - https://www.genome.jp/entry/pathway+pdam04150
3. Neuroactive ligand-receptor interaction -https://www.genome.jp/entry/pathway+pdam04080
4. Motor proteins - https://www.genome.jp/entry/pathway+pdam04814
5. Lysosome - https://www.genome.jp/entry/pathway+pdam04142
6. Cytoskeleton in muscle cells - https://www.genome.jp/entry/pathway+pdam04820

### Code

```{r pathview}
setwd("../output_RNA/differential_expression/KEGG")

pathview_plot <- function(pathway_id, suffix, LFC_vector) {
  pathview(
    gene.data = LFC_vector,
    pathway.id = pathway_id,
    species = "pdam",
    gene.idtype = "KEGG",
    kegg.native = TRUE,
    out.suffix = suffix,
    low = "green",
    high = "mediumpurple1"
  )
}

# Specify pathways of interest
pathways <- c("pdam04310", "pdam04150", "pdam04080", "pdam04814", "pdam04142", "pdam04820","pdam04082","pdam04512","pdam04145")

# Visualize for DE_05 data and all genes
for (pathway in pathways) {
  pathview_plot(pathway, "DE_05", DE_05_LFC)
  pathview_plot(pathway, "allgenes", LFC_allgenes)
}
```

```{r}
library(clusterProfiler)
library(enrichplot)

kk_aboral <- enrichKEGG(gene         = names(DE_05_LFC)[DE_05_LFC > 1],
                 organism     = 'pdam',
                 pAdjustMethod = "none",
                 pvalueCutoff = 0.01)

kk_oral <- enrichKEGG(gene         = names(DE_05_LFC)[DE_05_LFC < -1],
                 organism     = 'pdam',
                 pAdjustMethod = "none",
                 pvalueCutoff = 0.01)

kegg_gene_list <- sort(LFC_allgenes, decreasing = TRUE)

kk_gse <- gseKEGG(geneList     = kegg_gene_list,
               organism     = "pdam",
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               maxGSSize =150)
```

```{r}
dotplot(kk_aboral, showCategory = 10, title = "Enriched Pathways")
dotplot(kk_oral, showCategory = 10, title = "Enriched Pathways")

dotplot(kk_gse, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)
ridgeplot(kk_gse) + labs(x = "enrichment distribution")
```


## Updating Renv environment:

After youâ€™ve confirmed your code works as expected, use renv::snapshot() to record the packages and their sources in the lockfile.

```{r, eval=FALSE}
renv::snapshot()
```