---
title: "Transcription-Factors"
author: "Zoe Dellaert"
date: "2024-12-09"
output:
  github_document:
    toc: true
    toc_depth: 3
    number_sections: true
    html_preview: true
always_allow_html: true
---


## Transcription Factor analysis of LCM RNA Data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Managing Packages Using Renv

To run this code in my project using the renv environment, run the following lines of code

```{r, eval=FALSE}
install.packages("renv") #install the package on the new computer (may not be necessary if renv bootstraps itself as expected)
renv::restore() #reinstall all the package versions in the renv lockfile
```

## Load packages

```{r packages}
require("rtracklayer")
require("Biostrings")
require("GenomicRanges")
require("tidyverse")

sessionInfo() #provides list of loaded packages and version of R.
```

## Load in reference files and differential expression data

In the next chunk I am loadinf in my DESeq data. These results are ordered by adjusted p-value. As a reminder, negative LFC = higher in Aboral tissue, and positive LFC = higher in Oral tissue.

```{r load_data}
DESeq <- read.csv("../output_RNA/differential_expression/DESeq_results.csv", header = TRUE) %>% dplyr::rename("query" ="X")
DE_05 <- DESeq %>% filter(padj < 0.05)
```

I want to identify promoter sequences of the differentially expressed genes to identify transcription factor binding sites. Below I load in the *Pocillopora acuta* genome and gff annotation files, which are in the references folder of this repo and were downloaded as described here: https://github.com/zdellaert/LaserCoral/blob/main/code/RNA-seq-bioinf.md.

```{r references}
# Import the P. acuta genome
genome_fasta <- "../references/Pocillopora_acuta_HIv2.assembly.fasta"
genome <- readDNAStringSet(genome_fasta, format = "fasta")

# Import the P actua genome gff3 file
gff_file <- "../references/Pocillopora_acuta_HIv2.genes.gff3"
gff <- import(gff_file, format = "gff")

#What types of entries are in the gff
unique(gff$type)

#keep only transcript entries, to have one entry per gene
gff <- gff[gff$type=="transcript"]

#the gff is missing chromosome lengths, so I am adding them here:
chromosome_lengths <- seqlengths(genome)
head(chromosome_lengths)

head(seqlengths(gff))

seqlengths(gff) <- chromosome_lengths[names(chromosome_lengths) %in% seqlevels(gff)]

head(seqlengths(gff))
```

Filter the gff for all expressed genes in the dataset (filtered, this is not all genes in the genome)

```{r}
gff_expressed <- gff[gff$ID %in% DESeq$query] 
```

## Define promoter regions

I am basing this idea off of the following paper:

"extract the putative promoter sequence, defined as the 500 bp immediately upstream of the predicted transcription start site in the gene model."

Cleves PA, Krediet CJ, Lehnert EM, Onishi M, Pringle JR. Insights into coral bleaching under heat stress from analysis of gene expression in a sea anemone model system. Proceedings of the National Academy of Sciences. 2020 Nov 17;117(46):28906–17. (https://www.pnas.org/doi/10.1073/pnas.2015737117)

I am going to try using the promoters function of the GenomicRanges Package (manual here: https://bioconductor.org/packages/devel/bioc/manuals/GenomicRanges/man/GenomicRanges.pdf):

The function is described as such:

promoters: assumes that the ranges in x represent transcript regions and returns the ranges of the corresponding promoter regions. The result is another GenomicRanges derivative parallel to the input, that is, of the same length as x and with the i-th element in the output corresponding to the i-th element in the input.

The promoter regions extend around the transcription start sites (TSS) which are located at start(x) for ranges on the + or * strand, and at end(x) for ranges on the - strand. The upstream and downstream arguments define the number of nucleotides in the 5’ and 3’ direction, respectively. More precisely, the output range is defined as

> (start(x) - upstream) to (start(x) + downstream - 1)

for ranges on the + or * strand, and as

> (end(x) - downstream + 1) to (end(x) + upstream)

for ranges on the - strand. Be aware that the returned object might contain **out-of-bound ranges** i.e. ranges that start before the first nucleotide position and/or end after the last nucleotide position of the underlying sequence.

```{r}
promoters_500_UP <- promoters(gff_expressed, upstream = 500, downstream = 0)

# are any of these regions out of bounds?

summary(start(promoters_500_UP))

#yes! I have some regions starting at -499, which is not good.

#trim any promoters that exit the boundaries of the chromosomes, none of mine did but just in case
promoters_500_UP <- GenomicRanges::trim(promoters_500_UP)

summary(start(promoters_500_UP))

# also remove any rows where the promoter sequence had a width of 0
data.frame(promoters_500_UP[promoters_500_UP@ranges@width == 0])$ID

promoters_500_UP <- promoters_500_UP[promoters_500_UP@ranges@width != 0]
```

This removed these four genes:
"Pocillopora_acuta_HIv2___RNAseq.g8520.t1", "Pocillopora_acuta_HIv2___RNAseq.g23688.t1",  "Pocillopora_acuta_HIv2___RNAseq.g3131.t1" , and "Pocillopora_acuta_HIv2___RNAseq.g28711.t1" 

Then, extract these sequences, by chromosome, using the function "DNAStringSet":

```{r}
#create a DNAStringSet object to house all the sequences
promoters_all <- DNAStringSet()

for (chromosome in unique(seqnames(promoters_500_UP))) {
  
  # filter the genome, one chromosome at a time
  genome_filtered <- genome[genome@ranges@NAMES %in% chromosome]
  
  # one chromosome at a time, filter the gff dataframe
  promoters_500_UP_filtered <- data.frame(promoters_500_UP) %>% filter(seqnames == chromosome)
  
  # set up a list for this chromosome
  promoters_chromosome <- DNAStringSetList()
  
  # then iterate over every row (transcript) for that chromosome
  for (i in 1:nrow(promoters_500_UP_filtered)) {
    
    #use the DNAStringSet function to subset the chromosome into the promoter regions of interest
    promoter <- DNAStringSet(genome_filtered,
                             start= promoters_500_UP_filtered$start[i],
                             end= promoters_500_UP_filtered$end[i])
    
    #carry over the transcript name (what we really care about!)
    names(promoter) <- promoters_500_UP_filtered$ID[i]
    
    #append this extracted sequence to the promoter list for this chromosome
    promoters_chromosome <-  c(promoters_chromosome, promoter)
  }
  
  #append this chromosome list to the DNAStringSetList for all chromosomes
  promoters_all <- c(promoters_all, unlist(promoters_chromosome))
}

writeXStringSet(promoters_all, filepath = "../output_RNA/differential_expression/TFs/promoters_500_upstream.fasta")

promoters_upAboral <- promoters_all[promoters_all@ranges@NAMES %in% DE_05[which(DE_05$log2FoldChange > 0),]$query]

promoters_upOralEpi <- promoters_all[promoters_all@ranges@NAMES %in% DE_05[which(DE_05$log2FoldChange < 0),]$query]

writeXStringSet(promoters_upAboral, filepath = "../output_RNA/differential_expression/TFs/promoters_500_upstream_upAboral.fasta")
writeXStringSet(promoters_upOralEpi, filepath = "../output_RNA/differential_expression/TFs/promoters_500_upstream_upOralEpi.fasta")

# I could define differentially expressed as having a log fold change cutoff as well
```


## Run MEME in Linux environment using the [Docker image of MEMESuite](https://hub.docker.com/r/memesuite/memesuite) via Singularity

alternatively submit jobs via their webserver: https://meme-suite.org/meme/

### In unity:

```{bash, eval=FALSE}
cd ../scripts
nano MEME.sh
```


```{bash, eval=FALSE}
#!/usr/bin/env bash
#SBATCH --export=NONE
#SBATCH --nodes=1 --ntasks-per-node=24
#SBATCH --signal=2
#SBATCH --no-requeue
#SBATCH --mem=200GB
#SBATCH --error=../scripts/outs_errs/"%x_error.%j"
#SBATCH --output=../scripts/outs_errs/"%x_output.%j"
#SBATCH -t 24:00:00
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails

SINGULARITY_IMAGE="docker://memesuite/memesuite:latest"

#Download Motif databases
cd ../references
wget https://meme-suite.org/meme/meme-software/Databases/motifs/motif_databases.12.25.tgz
tar -xvzf motif_databases.12.25.tgz
mv motif_databases motif_dbs

cd ../output_RNA/differential_expression/TFs

module load apptainer/latest

# run MEME on the differentially expressed gene fasta files, only report motifs with e-value < 0.05

singularity exec --cleanenv $SINGULARITY_IMAGE meme promoters_500_upstream_upAboral.fasta -dna -maxw 25 -mod anr -evt 0.05 -p 6 -oc meme_output_upAboral -revcomp

singularity exec --cleanenv $SINGULARITY_IMAGE meme promoters_500_upstream_upOralEpi.fasta -dna -maxw 25 -mod anr -evt 0.05 -p 6 -oc meme_output_upOralEpi -revcomp

# run TOMTOM on the MEME-identified motifs from above

singularity exec --cleanenv $SINGULARITY_IMAGE tomtom -no-ssc -oc tomtom_output_upAboral -min-overlap 5 -dist pearson -thresh 0.05 meme_output_upAboral/meme.txt ../../../references/motif_dbs/JASPAR/JASPAR2022_CORE_non-redundant_v2.meme

singularity exec --cleanenv $SINGULARITY_IMAGE tomtom -no-ssc -oc tomtom_output_upOralEpi -min-overlap 5 -dist pearson -thresh 0.05 meme_output_upOralEpi/meme.txt ../../../references/motif_dbs/JASPAR/JASPAR2022_CORE_non-redundant_v2.meme
```

**up_Aboral**: Based on the TOMTOM results, only the motif had any significant (q < 0.05) matches to the JASPAR database. This poly-T motif is not interesting to me.

<img height="100" alt="motif1" src="https://github.com/zdellaert/LaserCoral/blob/main/output_RNA/differential_expression/TFs/meme_output_upAboral/logo1.png?raw=true">

**up_OralEpi**: Based on the TOMTOM results, only three motifs had significant (q < 0.05) matches to the JASPAR database:

1. Poly-C, not interesting: <img height="100" alt="motif1" src="https://github.com/zdellaert/LaserCoral/blob/main/output_RNA/differential_expression/TFs/meme_output_upOralEpi/logo2.png?raw=true">
2. Poly-T, not interesting: <img height="100" alt="motif1" src="https://github.com/zdellaert/LaserCoral/blob/main/output_RNA/differential_expression/TFs/meme_output_upOralEpi/logo3.png?raw=true">
3. CCGCCATBTTK (MEME-6): <img height="100" alt="motif1" src="https://github.com/zdellaert/LaserCoral/blob/main/output_RNA/differential_expression/TFs/meme_output_upOralEpi/logo6.png?raw=true">
  - Both of the matches matched to the **Reverse complement** of this motif.
  - Match 1: MA1651.1 (ZFP42)
    - <img height="300" alt="motif1" src="https://github.com/zdellaert/LaserCoral/blob/main/output_RNA/differential_expression/TFs/tomtom_output_upOralEpi/MEME_6_Match1.png?raw=true">
    - Database 	JASPAR2022_CORE_non-redundant_v2
    - p-value 	5.17e-06
    - E-value 	1.01e-02
    - q-value 	1.93e-02
    - Overlap 	11
    - Offset 	6
    - Orientation 	Reverse Complement
  - Match 2:
    - <img height="300" alt="motif1" src="https://github.com/zdellaert/LaserCoral/blob/main/output_RNA/differential_expression/TFs/tomtom_output_upOralEpi/MEME_6_Match2.png?raw=true">
    - Name 	MA0095.3 (Yy1)
    - Database 	JASPAR2022_CORE_non-redundant_v2
    - p-value 	9.87e-06
    - E-value 	1.93e-02
    - q-value 	1.93e-02
    - Overlap 	10
    - Offset 	-1
    - Orientation 	Reverse Complement

I am going to run FIMO to quantify this motif against all the genes.

```{bash, eval=FALSE}
cd ../scripts
nano MEME_FIMO.sh
```

```{bash, eval=FALSE}
#!/usr/bin/env bash
#SBATCH --export=NONE
#SBATCH --nodes=1 --ntasks-per-node=24
#SBATCH --signal=2
#SBATCH --no-requeue
#SBATCH --mem=200GB
#SBATCH --error=../scripts/outs_errs/"%x_error.%j"
#SBATCH --output=../scripts/outs_errs/"%x_output.%j"
#SBATCH -t 24:00:00
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails

SINGULARITY_IMAGE="docker://memesuite/memesuite:latest"

cd ../output_RNA/differential_expression/TFs

module load apptainer/latest

# run FIMO 

singularity exec --cleanenv $SINGULARITY_IMAGE fimo --oc fimo_output_upOralEpi --motif "CCGCCATBTTK" meme_output_upOralEpi/meme.txt promoters_500_upstream.fasta
```

Visualize log-fold change of genes that have this motif in their promoter sequences:

```{r}
fimo <- read.table("../output_RNA/differential_expression/TFs/fimo_output_upOralEpi/fimo.tsv", header = TRUE)

fimo_filtered <- fimo %>% filter(q.value < 0.05)  

# Merge FIMO results with DESeq data 
fimo_merged <- fimo_filtered %>% full_join(DESeq, by = c("sequence_name" = "query"))

#summarize

fimo_counts <- fimo_merged %>% group_by(sequence_name, log2FoldChange, padj,  motif_id) %>% count() %>%
  mutate(n = if_else(is.na(motif_id), 0, n)) 

fimo_counts_DE <- fimo_counts %>% filter(padj < 0.05)

ggplot(fimo_counts, aes(x = factor(n), y = log2FoldChange, fill = factor(n))) +
  geom_jitter(width = 0.2, alpha = 0.3, aes(color = factor(n))) +
  geom_violin(alpha = 0.7) + 
  theme_minimal() +
  labs(
    title = "Log2 Fold Change by Motif Presence/Absence",
    x = "Motif Copies ",
    y = "Log2 Fold Change"
  ) 

ggplot(fimo_counts_DE, aes(x = factor(n), y = log2FoldChange, fill = factor(n))) +
  geom_jitter(width = 0.2, alpha = 0.3, aes(color = factor(n))) +
  geom_violin(alpha = 0.7) + 
  theme_minimal() +
  labs(
    title = "Log2 Fold Change by Motif Presence/Absence, significantly DE genes",
    x = "Motif Copies ",
    y = "Log2 Fold Change"
  ) 

fimo_presence_absence <- fimo_counts %>% mutate(n = if_else(n>0, 1, n))

fimo_presence_absence_DE <- fimo_presence_absence %>% filter(padj < 0.05)

ggplot(fimo_presence_absence, aes(x = factor(n), y = log2FoldChange, fill = factor(n))) +
  geom_jitter(width = 0.2, alpha = 0.3, aes(color = factor(n))) +
  geom_violin(alpha = 0.7) + 
  theme_minimal() +
  labs(
    title = "Log2 Fold Change by Motif Presence/Absence",
    x = "Motif Copies ",
    y = "Log2 Fold Change"
  ) 

ggplot(fimo_presence_absence_DE, aes(x = factor(n), y = log2FoldChange, fill = factor(n))) +
  geom_jitter(width = 0.2, alpha = 0.3, aes(color = factor(n))) +
  geom_violin(alpha = 0.7) + 
  theme_minimal() +
  labs(
    title = "Log2 Fold Change by Motif Presence/Absence, significantly DE genes",
    x = "Motif Copies ",
    y = "Log2 Fold Change"
  ) 
```

Wow! It looks like the presence of this motif may be reflected in log2FoldChange! The genes with 1 or 2 sites have predominately negative fold change.

```{r}
# Fisher's exact test for presence/absence
fisher_table <- table(fimo_presence_absence_DE$n, fimo_presence_absence_DE$log2FoldChange > 0)
fisher_table
fisher.test(fisher_table)
```

```{r}
logistic_model <- glm((log2FoldChange > 0) ~ n, data = fimo_presence_absence_DE, family = binomial)
summary(logistic_model)
```


```{r}
plot_data <- fimo_counts %>%
  mutate(motif_count = case_when(
    n == 0 ~ "0 Motifs",
    n == 1 ~ "1 Motif",
    n > 1 ~ ">1 Motifs"
  )) %>%
  mutate(motif_count = factor(motif_count, levels = c("0 Motifs", "1 Motif", ">1 Motifs")))

plot <- ggplot(plot_data, aes(x = motif_count, y = log2FoldChange, fill = motif_count)) +
  geom_jitter(aes(color = motif_count), width = 0.2, size = 1.5, alpha = 0.6) +  
  geom_violin(alpha = 0.7, width = 0.5) +  
  stat_summary(fun = mean, geom = "point", shape = 21,  size = 2, fill = "white", color = "black") + 
  scale_fill_brewer(palette = "Set2") +  
  scale_color_brewer(palette = "Set2") + 
  labs(
    title = "Log2 Fold Change by Motif Presence",
    subtitle = "All Genes",
    x = "Number of Motifs in Promoter",
    y = "Log2 Fold Change",
    fill = "Motif Count",
    color = "Motif Count"
  ) +
  theme_minimal(base_size = 14) + 
  theme(
    legend.position = "none",  
    plot.title = element_text(face = "bold", size = 16),
    plot.subtitle = element_text(size = 14),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14),
    panel.grid.major = element_line(size = 0.5, linetype = "dashed", color = "gray80")
  )

# Save the plot as a high-quality PNG
ggsave("../output_RNA/differential_expression/TFs/fimo_motif6.png", plot, width = 8, height = 6, dpi = 300)

# Display the plot
print(plot)

plot_data <- fimo_counts_DE %>%
  mutate(motif_count = case_when(
    n == 0 ~ "0 Motifs",
    n == 1 ~ "1 Motif",
    n > 1 ~ ">1 Motifs"
  )) %>%
  mutate(motif_count = factor(motif_count, levels = c("0 Motifs", "1 Motif", ">1 Motifs")))

plot <- ggplot(plot_data, aes(x = motif_count, y = log2FoldChange, fill = motif_count)) +
  geom_jitter(aes(color = motif_count), width = 0.2, size = 1.5, alpha = 0.6) +  
  geom_violin(alpha = 0.7, width = 0.5) + 
  stat_summary(fun = mean, geom = "point", shape = 21, size = 2, fill = "white", color = "black") +  
  scale_fill_brewer(palette = "Set2") + 
  scale_color_brewer(palette = "Set2") +  
  labs(
    title = "Log2 Fold Change by Motif Presence",
    subtitle = "Significantly DE Genes",
    x = "Number of Motifs in Promoter",
    y = "Log2 Fold Change",
    fill = "Motif Count",
    color = "Motif Count"
  ) +
  theme_minimal(base_size = 14) +  
  theme(
    legend.position = "none", 
    plot.title = element_text(face = "bold", size = 16),
    plot.subtitle = element_text(size = 14),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14),
    panel.grid.major = element_line(size = 0.5, linetype = "dashed", color = "gray80")
  )

ggsave("../output_RNA/differential_expression/TFs/fimo_motif6_DE.png", plot, width = 8, height = 6, dpi = 300)

# Display the plot
print(plot)
```



## Updating Renv environment:

After you’ve confirmed your code works as expected, use renv::snapshot() to record the packages and their sources in the lockfile.

```{r, eval=FALSE}
renv::snapshot()
```