---
title: "Semantic Enrichment"
author: "Zoe Dellaert"
date: "2025-01-11"
output:
  html_document: default
  github_document:
    toc: true
    toc_depth: 3
    number_sections: true
    html_preview: true
always_allow_html: true
---

## Gene Ontology Analysis analysis of LCM RNA Data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Managing Packages Using Renv

To run this code in my project using the renv environment, run the following lines of code

```{r, eval=FALSE}
install.packages("renv") #install the package on the new computer (may not be necessary if renv bootstraps itself as expected)
renv::restore() #reinstall all the package versions in the renv lockfile
```

## Load packages

```{r packages}
library("ViSEAGO")
require("topGO")
require("tidyverse")
#BiocManager::install("GO.db")

sessionInfo() #provides list of loaded packages and version of R.
```

## Description of pipeline

I am going to perform functional enrichment of GO terms using [ViSEAGO](https://biodatamining.biomedcentral.com/articles/10.1186/s13040-019-0204-1). 

I am following this vignette: [http://bioconductor.unipi.it/packages/devel/bioc/vignettes/ViSEAGO/inst/doc/ViSEAGO.html](http://bioconductor.unipi.it/packages/devel/bioc/vignettes/ViSEAGO/inst/doc/ViSEAGO.html).

## Load in reference files and differential expression data

In the next chunk I am loading in my DESeq data. These results are ordered by adjusted p-value. As a reminder, negative LFC = higher in Oral tissue, and positive LFC = higher in Aboral tissue.

```{r load_data}
#load in DESeq results
DESeq <- read.csv("../output_RNA/differential_expression/DESeq_results.csv", header = TRUE) %>% dplyr::rename("query" ="X")

#make dataframes of just differentially expressed genes for each LFC direction - filtering a little more stringent, abs(LFC) >2
DE_05_Aboral <- DESeq %>% filter(padj < 0.05 & log2FoldChange > 1)
DE_05_OralEpi <- DESeq %>% filter(padj < 0.05& log2FoldChange < -1)

#load in annotation data 
annot_tab <- read.delim("../references/annotation/protein-GO.tsv") %>% dplyr::rename(GOs = GeneOntologyIDs)

#filter annotation data for just expressed genes with GO annotations
annot_tab <- annot_tab %>% filter(query %in% DESeq$query) 

annot_tab$GOs <- gsub("; ", ";", annot_tab$GOs)
annot_tab$GOs[annot_tab$GOs==""] <- NA
annot_tab <- annot_tab %>% filter(!is.na(GOs))

nrow(annot_tab)
nrow(annot_tab)/nrow(DESeq)
```

10638/14464 genes in our dataset have GO information in this file. That is 74%.

```{r}
sum(annot_tab$query %in% DE_05_Aboral$query)
sum(annot_tab$query %in% DE_05_Aboral$query)/nrow(DE_05_Aboral)

sum(annot_tab$query %in% DE_05_OralEpi$query)
sum(annot_tab$query %in% DE_05_OralEpi$query)/nrow(DE_05_OralEpi)
```

379/558 genes that are significantly upregulated in the Aboral tissue have annotation information. That is 68% of the genes.

796/1216 genes that are significantly upregulated in the Oral Epidermis tissue have annotation information. That is 71% of the genes.

## Create custom GO annotation file for ViSEAGO

```{r}
##Get a list of GO Terms for all genes
annots <- annot_tab %>% dplyr::select(query,GOs) %>% dplyr::rename("GO.terms" = GOs)

# format into the format required by ViSEAGO for custom mappings
Custom_GOs <- annots %>%
  # Separate GO terms into individual rows
  separate_rows(GO.terms, sep = ";") %>%
  # Add necessary columns
  mutate(
    taxid = "pacuta",
    gene_symbol = query,
    evidence = "SwissProt"
  ) %>%
  # Rename columns
  dplyr::rename(
    gene_id = query,
    GOID = GO.terms
  ) %>%
  dplyr::select(taxid, gene_id, gene_symbol, GOID, evidence)

Custom_GOs_valid <- Custom_GOs %>% filter(GOID %in% keys(GO.db))

write.table(Custom_GOs_valid, "../output_RNA/differential_expression/semantic-enrichment/custom_GOs.txt",row.names = FALSE, sep = "\t", quote = FALSE,col.names=TRUE)

length(unique(Custom_GOs$gene_id))
length(unique(Custom_GOs_valid$gene_id))
```

We seem to have lost one gene when filtering for valid GO terms, so I need to account for that below.

### load the file into ViSEAGO

```{r}
Custom_Pacuta <- ViSEAGO::Custom2GO("../output_RNA/differential_expression/semantic-enrichment/custom_GOs.txt")

myGENE2GO_Pacuta <- ViSEAGO::annotate(
    id="pacuta",
    Custom_Pacuta
)
```
    
## Create gene lists for enrichment

```{r}
selection <- DESeq %>% filter(query %in% Custom_GOs_valid$gene_id) %>% 
                       mutate(DE_05_Aboral = ifelse(query %in% DE_05_Aboral$query, 1,0)) %>%
                       mutate(DE_05_Oral = ifelse(query %in% DE_05_OralEpi$query, 1,0)) %>%
                       mutate(expressed = 1)

selection_Aboral <- selection %>% pull(DE_05_Aboral) %>% as.factor()
names(selection_Aboral) <- selection %>% pull(query)

selection_Oral <- selection %>% pull(DE_05_Oral) %>% as.factor()
names(selection_Oral) <- selection %>% pull(query)

expressed <- selection %>% pull(expressed) %>% as.factor()
names(expressed) <- selection %>% pull(query)
```

## Oral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Oral)[selection_Oral==1]
background <- names(expressed)

BP_Oral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="BP",
    nodeSize=5
)

# perform TopGO test using classic algorithm
classic_Oral <- topGO::runTest(
    BP_Oral,
    algorithm ="classic",
    statistic = "fisher"
)

# perform TopGO test using elim algorithm
elim_Oral <- topGO::runTest(
    BP_Oral,
    algorithm ="elim",
    statistic = "fisher"
)

BP_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(#Oral_classic = c("BP_Oral", "classic_Oral"),
                 Oral_elim = c("BP_Oral", "elim_Oral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
BP_Results
ViSEAGO::show_table(BP_Results)

# print the merged table in a file
ViSEAGO::show_table(
    BP_Results,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Oral.csv"
)
```

 - enrichment pvalue cutoff:
        Oral_classic : 0.01
        Oral_elim : 0.01
- enrich GOs (in at least one list): 171 GO terms of 2 conditions.
        Oral_classic : 171 terms
        Oral_elim : 58 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=BP_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2 <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Oral_cluster_heatmap_Wang_wardD2.csv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOclusters")
```

## Aboral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Aboral)[selection_Aboral==1]
background <- names(expressed)

BP_Aboral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="BP",
    nodeSize=5
)

# perform TopGO test using classic algorithm
classic_Aboral <- topGO::runTest(
    BP_Aboral,
    algorithm ="classic",
    statistic = "fisher"
)

# perform TopGO test using elim algorithm
elim_Aboral <- topGO::runTest(
    BP_Aboral,
    algorithm ="elim",
    statistic = "fisher"
)

BP_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(#Aboral_classic = c("BP_Aboral", "classic_Aboral"),
                 Aboral_elim = c("BP_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
BP_Results
ViSEAGO::show_table(BP_Results)

# print the merged table in a file
ViSEAGO::show_table(
    BP_Results,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Aboral.csv"
)
```

 - enrichment pvalue cutoff:
        Aboral_classic : 0.01
        Aboral_elim : 0.01
- enrich GOs (in at least one list): 347 GO terms of 2 conditions.
        Aboral_classic : 347 terms
        Aboral_elim : 118 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=BP_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2 <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Aboral_cluster_heatmap_Wang_wardD2.csv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOclusters")
```

## Both tissues together:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
BP_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Oral_elim = c("BP_Oral", "elim_Oral"),
                 Aboral_elim = c("BP_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
BP_Results
ViSEAGO::show_table(BP_Results)

# print the merged table in a file
ViSEAGO::show_table(
    BP_Results,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05.csv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
        Aboral_elim : 0.01
- enrich GOs (in at least one list): 176 GO terms of 2 conditions.
        Oral_elim : 58 terms
        Aboral_elim : 118 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=BP_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```
#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2 <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 4
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_cluster_heatmap_Wang_wardD2.csv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOclusters")
```


```{r}
# GOclusters heatmap
Wang_clusters_wardD2<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOclusters")
```

## Custom plotting of results

```{r}
clustered_allDEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) %>%
  select(cluster, Cluster.name)

cluster_map
```

### Manually fix cluster names based on GO terms inside each cluster

```{r}
cluster_map_fixed <- cluster_map %>%
  mutate(Cluster.name = str_replace(Cluster.name, "Cluster 2: developmental process", "Cluster 2: cell fate determination"),
         Cluster.name = str_replace(Cluster.name, "Cluster 4: anatomical structure development", "Cluster 4: regeneration & tissue morphogenesis"),
         Cluster.name = str_replace(Cluster.name, "Cluster 5: multicellular organism development", "Cluster 5: multicellular organism pattern formation"),
         Cluster.name = str_replace(Cluster.name, "Cluster 9: transport", "Cluster 9: metal ion transmembrane transport"),
         Cluster.name = str_replace(Cluster.name, "Cluster 10: transport", "Cluster 10: amino acid & small molecule transmembrane transport"),
        Cluster.name = str_replace(Cluster.name, "Cluster 12: nervous system process", "Cluster 12: sensory perception"),
        Cluster.name = str_replace(Cluster.name, "Cluster 16: biological_process", "Cluster 16: regeneration & wound healing"),
        Cluster.name = str_replace(Cluster.name, "Cluster 18: secretion by cell", "Cluster 18: neurotransmitter and hormone secretion"),
        Cluster.name = str_replace(Cluster.name, "Cluster 24: external encapsulating structure organization", "Cluster 24: extracellular matrix organization"),
        Cluster.name = str_replace(Cluster.name, "Cluster 25: biological_process", "Cluster 25: cell organization and homeostasis")
         )
```


```{r}
clustered_allDEGs_enrichedGO <- clustered_allDEGs_enrichedGO %>%
  left_join(cluster_map_fixed, by = c("GO.cluster" = "cluster"))

write.csv(clustered_allDEGs_enrichedGO,"../output_RNA/differential_expression/semantic-enrichment/DE_05_clusters_named.csv")

library(DT)

# Interactive table with scrolling
datatable(clustered_allDEGs_enrichedGO,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))
```

### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  select(Cluster.name, Oral_elim.Significant_genes, Aboral_elim.Significant_genes) %>%
  mutate(
    # Clean cluster names
    Cluster.name = gsub("^Cluster \\d+:\\s*", "", Cluster.name),
    Oral_terms = ifelse(!is.na(Oral_elim.Significant_genes), 1, 0),
    Aboral_terms = ifelse(!is.na(Aboral_elim.Significant_genes), 1, 0)
  ) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Oral_terms),
    Aboral_terms = sum(Aboral_terms),
    .groups = "drop"
  ) %>%
  filter(Oral_terms + Aboral_terms > 0) %>%
  mutate(
    deviation = Oral_terms - Aboral_terms,
    Aboral_terms = -Aboral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

aboral_to_equal <- sum(transition_data$deviation < 0) + 0.5
equal_to_oral <- sum(transition_data$deviation <= 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(aboral_to_equal, equal_to_oral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#2E8B57", "Aboral" = "#9370DB")) +
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 11)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/GO_enrichment_bidirectional.png", p, width = 10, height = 8, dpi = 300)
print(p)
```


