---
title: "Semantic Enrichment"
author: "Zoe Dellaert"
date: "2025-01-11"
output:
  html_document: default
  github_document:
    toc: true
    toc_depth: 3
    number_sections: true
    html_preview: true
always_allow_html: true
---

## Gene Ontology Analysis analysis of LCM RNA Data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Managing Packages Using Renv

To run this code in my project using the renv environment, run the following lines of code

```{r, eval=FALSE}
install.packages("renv") #install the package on the new computer (may not be necessary if renv bootstraps itself as expected)
renv::restore() #reinstall all the package versions in the renv lockfile
```

## Load packages

```{r packages}
library("ViSEAGO")
require("topGO")
require("tidyverse")
#BiocManager::install("GO.db")

sessionInfo() #provides list of loaded packages and version of R.
```

## Description of pipeline

I am going to perform functional enrichment of GO terms using [ViSEAGO](https://biodatamining.biomedcentral.com/articles/10.1186/s13040-019-0204-1). 

I am following this vignette: [http://bioconductor.unipi.it/packages/devel/bioc/vignettes/ViSEAGO/inst/doc/ViSEAGO.html](http://bioconductor.unipi.it/packages/devel/bioc/vignettes/ViSEAGO/inst/doc/ViSEAGO.html).

## Load in reference files and differential expression data

In the next chunk I am loading in my DESeq data. These results are ordered by adjusted p-value. As a reminder, negative LFC = higher in Oral tissue, and positive LFC = higher in Aboral tissue.

```{r load_data}
#load in DESeq results
DESeq <- read.csv("../output_RNA/differential_expression/DESeq_results.csv", header = TRUE) %>% dplyr::rename("query" ="X")

#make dataframes of just differentially expressed genes for each LFC direction - filtering a little more stringent, abs(LFC) >2
DE_05_Aboral <- DESeq %>% filter(padj < 0.05 & log2FoldChange < -1)
DE_05_OralEpi <- DESeq %>% filter(padj < 0.05& log2FoldChange > 1)

#load in annotation data 
annot_tab <- read.delim("../references/annotation/protein-GO.tsv") %>% dplyr::rename(GOs = GeneOntologyIDs)

#filter annotation data for just expressed genes with GO annotations
annot_tab <- annot_tab %>% filter(query %in% DESeq$query) 

annot_tab$GOs <- gsub("; ", ";", annot_tab$GOs)
annot_tab$GOs[annot_tab$GOs==""] <- NA
annot_tab <- annot_tab %>% filter(!is.na(GOs))

nrow(annot_tab)
nrow(annot_tab)/nrow(DESeq)
```

10638/14464 genes in our dataset have GO information in this file. That is 74%.

```{r}
sum(annot_tab$query %in% DE_05_Aboral$query)
sum(annot_tab$query %in% DE_05_Aboral$query)/nrow(DE_05_Aboral)

sum(annot_tab$query %in% DE_05_OralEpi$query)
sum(annot_tab$query %in% DE_05_OralEpi$query)/nrow(DE_05_OralEpi)
```

379/558 genes that are significantly upregulated in the Aboral tissue have annotation information. That is 68% of the genes.

796/1216 genes that are significantly upregulated in the Oral Epidermis tissue have annotation information. That is 71% of the genes.

## Create custom GO annotation file for ViSEAGO

```{r}
##Get a list of GO Terms for all genes
annots <- annot_tab %>% dplyr::select(query,GOs,ProteinNames) %>% dplyr::rename("GO.terms" = GOs)

# format into the format required by ViSEAGO for custom mappings
Custom_GOs <- annots %>%
  # Separate GO terms into individual rows
  separate_rows(GO.terms, sep = ";") %>%
  # Add necessary columns
  mutate(
    taxid = "pacuta",
    gene_symbol = ProteinNames,
    evidence = "SwissProt"
  ) %>%
  # Rename columns
  dplyr::rename(
    gene_id = query,
    GOID = GO.terms
  ) %>%
  dplyr::select(taxid, gene_id, gene_symbol, GOID, evidence)

Custom_GOs_valid <- Custom_GOs %>% filter(GOID %in% keys(GO.db))

write.table(Custom_GOs_valid, "../output_RNA/differential_expression/semantic-enrichment/custom_GOs.txt",row.names = FALSE, sep = "\t", quote = FALSE,col.names=TRUE)

length(unique(Custom_GOs$gene_id))
length(unique(Custom_GOs_valid$gene_id))
```

We seem to have lost one gene when filtering for valid GO terms, so I need to account for that below.

### load the file into ViSEAGO

```{r}
Custom_Pacuta <- ViSEAGO::Custom2GO("../output_RNA/differential_expression/semantic-enrichment/custom_GOs.txt")

myGENE2GO_Pacuta <- ViSEAGO::annotate(
    id="pacuta",
    Custom_Pacuta
)
```
    
## Create gene lists for enrichment

```{r}
selection <- DESeq %>% filter(query %in% Custom_GOs_valid$gene_id) %>% 
                       mutate(DE_05_Aboral = ifelse(query %in% DE_05_Aboral$query, 1,0)) %>%
                       mutate(DE_05_Oral = ifelse(query %in% DE_05_OralEpi$query, 1,0)) %>%
                       mutate(expressed = 1)

selection_Aboral <- selection %>% pull(DE_05_Aboral) %>% as.factor()
names(selection_Aboral) <- selection %>% pull(query)

selection_Oral <- selection %>% pull(DE_05_Oral) %>% as.factor()
names(selection_Oral) <- selection %>% pull(query)

expressed <- selection %>% pull(expressed) %>% as.factor()
names(expressed) <- selection %>% pull(query)
```

# BP
## Oral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Oral)[selection_Oral==1]
background <- names(expressed)

BP_Oral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="BP",
    nodeSize=5
)

# perform TopGO test using elim algorithm
elim_Oral <- topGO::runTest(
    BP_Oral,
    algorithm ="elim",
    statistic = "fisher"
)

BP_Results <- ViSEAGO::merge_enrich_terms(
    Input = list( Oral_elim = c("BP_Oral", "elim_Oral")))
```



### Visualize and save initial results

```{r}
# display the merged table
BP_Results
ViSEAGO::show_table(BP_Results)

# print the merged table in a file
ViSEAGO::show_table(
    BP_Results,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Oral.tsv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
- enrich GOs (in at least one list):
        Oral_elim : 67 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=BP_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2_Oral <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    "GOterms")

ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    height=1000,
    width=700,
    "GOterms", file="../output_RNA/differential_expression/semantic-enrichment/DE_05_Oral_cluster_heatmap_Wang_wardD2.png")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2_Oral)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2_Oral,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Oral_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Oral,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2_Oral<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2_Oral,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Oral,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2_Oral <-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2_Oral,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    "GOclusters")
```

```{r}
DE_05_SwissProt <- read.csv("../output_RNA/differential_expression/DE_05_SwissProt_annotation.csv")
DESeq_SwissProt <- read.csv("../output_RNA/differential_expression/DESeq_SwissProt_annotation.csv")

Oral_enriched_clustered <- Wang_clusters_wardD2_Oral@enrich_GOs@data

top_cluster <- Oral_enriched_clustered %>% arrange(Oral_elim.pvalue) %>% head(1) %>% pull(GO.cluster)

Oral_enriched_clustered %>% dplyr::filter(GO.cluster == top_cluster)
```

```{r}
BP_go_terms <- BP_Oral@graph@nodes
genes_by_GO <- genesInTerm(BP_Oral, BP_go_terms)


cluster_8 <- Oral_enriched_clustered %>% dplyr::filter(GO.cluster == 8) %>% pull(GO.ID)
cluster_8_9 <- c(cluster_8, "GO:0050982")
  
cluster_8_9_genes <- unique(unlist(genes_by_GO[cluster_8_9]))

write.csv(cluster_8_9_genes,"../output_RNA/differential_expression/semantic-enrichment/cluster_gene_lists/Oral_enrich_BP_Sensory.csv")

DE_05_SwissProt %>% filter(query %in%unique(unlist(genes_by_GO[Oral_enriched_clustered %>% pull(GO.ID)]))) %>% View()
```


## Aboral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Aboral)[selection_Aboral==1]
background <- names(expressed)

BP_Aboral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="BP",
    nodeSize=5
)

# perform TopGO test using elim algorithm
elim_Aboral <- topGO::runTest(
    BP_Aboral,
    algorithm ="elim",
    statistic = "fisher"
)

BP_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Aboral_elim = c("BP_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
BP_Results
ViSEAGO::show_table(BP_Results)

# print the merged table in a file
ViSEAGO::show_table(
    BP_Results,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Aboral.tsv"
)
```

 - enrichment pvalue cutoff:
        Aboral_elim : 0.01
- enrich GOs (in at least one list):
        Aboral_elim : 118 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=BP_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2_Aboral <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 1,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2_Aboral)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2_Aboral,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_Aboral_cluster_heatmap_Wang_wardD2.tsv"
)

ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    height=1500,
    width=700,
    "GOterms", file="../output_RNA/differential_expression/semantic-enrichment/DE_05_Aboral_cluster_heatmap_Wang_wardD2.png")

```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Aboral,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2_Aboral<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2_Aboral,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Aboral,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2_Aboral<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2_Aboral,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOclusters")
```

```{r}
Aboral_enriched_clustered <- Wang_clusters_wardD2_Aboral@enrich_GOs@data

top_cluster <- Aboral_enriched_clustered %>% arrange(Aboral_elim.pvalue) %>% head(1) %>% pull(GO.cluster)

Aboral_enriched_clustered %>% dplyr::filter(GO.cluster == top_cluster)

ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOterms")
```

## Plotting of both tissues, pre-clustered by tissue

### Custom plotting of results

```{r}
clustered_Oral_DEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2_Oral@enrich_GOs@data) 

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2_Oral,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_",""),
    Cluster.term = str_replace(cluster_full,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) %>%
  dplyr::select(cluster, Cluster.name)

clustered_Oral_DEGs_enrichedGO <- clustered_Oral_DEGs_enrichedGO %>%
  left_join(cluster_map, by = c("GO.cluster" = "cluster")) %>%
  mutate(Tissue="Oral")

clustered_Aboral_DEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2_Aboral@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2_Aboral,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_",""),
    Cluster.term = str_replace(cluster_full,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) %>%
  dplyr::select(cluster, Cluster.name)
clustered_Aboral_DEGs_enrichedGO <- clustered_Aboral_DEGs_enrichedGO %>%
  left_join(cluster_map, by = c("GO.cluster" = "cluster")) %>%
  mutate(Tissue="Aboral")


clustered_allDEGs_enrichedGO <- rbind(clustered_Oral_DEGs_enrichedGO %>%
                                        dplyr::select(-contains("Oral")),
                                      clustered_Aboral_DEGs_enrichedGO %>%
                                        dplyr::select(-contains("Aboral")))

```

### NEW: Plotting of each tissue cluster separately

#### extract cluster trees
```{r}
library(ggdendro)
library(patchwork)
dist_mat <- slot(Wang_clusters_wardD2_Oral, "clusters_dist")[["BMA"]]
hc <- hclust(dist_mat, method = "ward.D2")
dend <- as.dendrogram(hc)

dend_data <- ggdendro::dendro_data(dend, type = "rectangle")

dend_labels <- dend_data$labels %>%
    mutate(
    cluster = str_extract(label, "\\d+"),
    Cluster.name = str_replace(label,".*?_.*?_",""),
    Cluster.term = str_replace(label,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) 

bar_plot <- clustered_allDEGs_enrichedGO %>%  
filter(Tissue == "Oral") %>%
  arrange(as.numeric(GO.cluster)) %>%
  mutate(Cluster_label = factor(Cluster.name, levels = rev(dend_labels$Cluster.name))) %>%
  ggplot(aes(x = Cluster_label)) +
  stat_count(width = .8, fill="#FD8D3C", alpha = 0.8) +
  coord_flip() +
  theme_classic(base_size = 9) +
  scale_y_continuous(
    breaks = scales::breaks_width(2),
    expand = c(0, 0),
    limits = c(0, NA)
  ) + 
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Oral epidermis", y = "Terms in Cluster")

dend_plot <- ggplot() +
  geom_segment(
    data = dend_data$segments,
    aes(x = y, y = x, xend = yend, yend = xend),
    linewidth = 0.6,
    color = "#2C3E50"
  ) +
  scale_y_reverse(
    breaks = seq_along(dend_labels$Cluster.name),
    expand = c(0.02, 0.02)
  ) +
  scale_x_reverse(expand = c(0.02, 0)) +
  theme_void() +
  theme(
    plot.margin = margin(t = 30, r = 0, b = 22, l = 5)
  )

# Combine plots with better proportions
final <- dend_plot + bar_plot +
  plot_layout(widths = c(0.8, 3.5))

Wang_clusters_wardD2_Oral@heatmap$GOclusters_static

ggsave("../output_RNA/differential_expression/semantic-enrichment/BP_Dend_Clusters_Oral.png", final, width = 5, height = 6.5, dpi = 300)
```

```{r}
library(ggdendro)
dist_mat <- slot(Wang_clusters_wardD2_Aboral, "clusters_dist")[["BMA"]]
hc <- hclust(dist_mat, method = "ward.D2")
dend <- as.dendrogram(hc)

dend_data <- ggdendro::dendro_data(dend, type = "rectangle")

dend_labels <- dend_data$labels %>%
    mutate(
    cluster = str_extract(label, "\\d+"),
    Cluster.name = str_replace(label,".*?_.*?_",""),
    Cluster.term = str_replace(label,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) 

bar_plot <- clustered_allDEGs_enrichedGO %>%  
filter(Tissue == "Aboral") %>%
  arrange(as.numeric(GO.cluster)) %>%
  mutate(Cluster_label = factor(Cluster.name, levels = rev(dend_labels$Cluster.name))) %>%
  ggplot(aes(x = Cluster_label)) +
  stat_count(width = .8, fill="#756BB1", alpha = 0.8) +
  coord_flip() +
  theme_classic(base_size = 9) +
  scale_y_continuous(
    breaks = scales::breaks_width(2),
    expand = c(0, 0),
    limits = c(0, NA)
  ) + 
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Aboral", y = "Terms in Cluster")

dend_plot <- ggplot() +
  geom_segment(
    data = dend_data$segments,
    aes(x = y, y = x, xend = yend, yend = xend),
    linewidth = 0.6,
    color = "#2C3E50"
  ) +
  scale_y_reverse(
    breaks = seq_along(dend_labels$Cluster.name),
    expand = c(0.02, 0.02)
  ) +
  scale_x_reverse(expand = c(0.02, 0)) +
  theme_void() +
  theme(
    plot.margin = margin(t = 30, r = 0, b = 22, l = 5)
  )

# Combine plots with better proportions
final <- dend_plot + bar_plot +
  plot_layout(widths = c(0.8, 3.5))

Wang_clusters_wardD2_Oral@heatmap$GOclusters_static

ggsave("../output_RNA/differential_expression/semantic-enrichment/BP_Dend_Clusters_Aboral.png", final, width = 5, height = 6.5, dpi = 300)
```

#### bar plots
```{r}
p1 <- clustered_allDEGs_enrichedGO %>%  
filter(Tissue == "Aboral") %>%
  arrange(as.numeric(GO.cluster)) %>%
  mutate(Cluster_label = factor(Cluster.name, levels = rev(unique(Cluster.name)))) %>%
  ggplot(aes(x = Cluster_label)) +
  stat_count(width = .8, fill="#756BB1", alpha = 0.8) +
  coord_flip() +
  scale_y_continuous(breaks = scales::breaks_width(2)) +
  theme_bw(base_size = 9) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Aboral", y = "Terms in Cluster")

p2 <- clustered_allDEGs_enrichedGO %>%  
filter(Tissue == "Oral") %>%
  arrange(as.numeric(GO.cluster)) %>%
  mutate(Cluster_label = factor(Cluster.name, levels = rev(unique(Cluster.name)))) %>%
  ggplot(aes(x = Cluster_label)) +
  stat_count(width = .8, fill="#FD8D3C", alpha = 0.8) +
  coord_flip() +
  theme_bw(base_size = 9) +
  scale_y_continuous(breaks = scales::breaks_width(2)) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Oral epidermis", y = "Terms in Cluster")

plot <- p1 + p2 +
  plot_annotation(
    title = "Semantic Similarity Clusters of All Enriched (p < 0.01) BP GO Terms",
    theme = theme(plot.title = element_text(face = "bold", size = 9, hjust = 0.5))
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/BP_Clusters.png", plot, width = 7, height = 6.5, dpi = 300)
```

### NEW: Plotting of both tissues, pre-clustered by tissue

```{r}
top_Oral_per_Cluster <- clustered_Oral_DEGs_enrichedGO %>% group_by(GO.cluster) %>% arrange(Oral_elim.pvalue) %>% slice(1) %>% ungroup() %>% dplyr::select(GO.cluster,GO.ID,term) %>% rename("top_sig_GO" = `GO.ID`,"top_sig_term" = `term`)

plotting_oral <- left_join(clustered_Oral_DEGs_enrichedGO,top_Oral_per_Cluster) %>%
  arrange(as.numeric(GO.cluster)) %>%
                        mutate(is_top = GO.ID == top_sig_GO,
                               pvalue = Oral_elim.pvalue,
                                term_wrapped = str_wrap(`term`, width = 25),
                                 cluster_label = factor(str_wrap(Cluster.name,
                                        width = 20), levels = unique(str_wrap(Cluster.name,
                                        width = 20)))) %>%
  add_count(GO.cluster, name = "n_terms")

top_Aboral_per_Cluster <- clustered_Aboral_DEGs_enrichedGO %>% group_by(GO.cluster) %>% arrange(Aboral_elim.pvalue) %>% slice(1) %>% ungroup() %>% dplyr::select(GO.cluster,GO.ID,term) %>% rename("top_sig_GO" = `GO.ID`,"top_sig_term" = `term`)

plotting_aboral <- left_join(clustered_Aboral_DEGs_enrichedGO,top_Aboral_per_Cluster)%>%
                      arrange(as.numeric(GO.cluster)) %>%
                        mutate(is_top = GO.ID == top_sig_GO,
                               pvalue = Aboral_elim.pvalue,
                                term_wrapped = str_wrap(`term`, width = 25),
                                 cluster_label = factor(str_wrap(Cluster.name,
                                        width = 20), levels = unique(str_wrap(Cluster.name,
                                        width = 20)))) %>%
  add_count(GO.cluster, name = "n_terms")
  

combined_plotting <- rbind(plotting_oral %>%
                                        dplyr::select(-contains("Oral")),
                                      plotting_aboral %>%
                                        dplyr::select(-contains("Aboral")))
```

```{r}
library(ggrepel)
p_oral <- ggplot(plotting_oral, 
         aes(x = `Oral_elim.-log10_pvalue`, 
             y = reorder(GO.ID, `Oral_elim.-log10_pvalue`),
             color = factor(GO.cluster),
             size = `Oral_elim.-log10_pvalue`)) +
    geom_segment(aes(xend = 0, yend = reorder(GO.ID, `Oral_elim.-log10_pvalue`)),
                 alpha = 0.5, linewidth = 0.8) +
    geom_point(alpha = 0.8) +
    facet_grid(cluster_label ~ ., 
               scales = "free_y", space = "free_y") +
    geom_text_repel(data = plotting_oral %>% filter(is_top),
                  aes(label = term_wrapped),
                  size = 2.5,
                  hjust = 0,
                  nudge_x = 0.3,
                  direction = "y",
                  color = "black",
                  fontface = "bold",
                  segment.color = "gray40",
                  segment.size = 0.3,
                  box.padding = 0.3,
                  max.overlaps = 30) +
    labs(title = "Oral Epidermis Tissue",
         x = "-log10(p-value)",
         y = NULL) +
    theme_minimal(base_size = 10) +
    theme(plot.title = element_text(face = "bold", size = 12),
          axis.text.y = element_text(size = 7.5),
          strip.text.y = element_text(size = 8.5, angle = 0, hjust = 0, face = "bold"),
          legend.position = "none",
          panel.spacing = unit(0.5, "lines"),
          panel.grid.major.y = element_blank(),
          panel.grid.minor = element_blank()) +
    scale_size_continuous(range = c(2, 5))

p_aboral <- ggplot(plotting_aboral, 
         aes(x = `Aboral_elim.-log10_pvalue`, 
             y = reorder(GO.ID, `Aboral_elim.-log10_pvalue`),
             color = factor(GO.cluster),
             size = `Aboral_elim.-log10_pvalue`)) +
    geom_segment(aes(xend = 0, yend = reorder(GO.ID, `Aboral_elim.-log10_pvalue`)),
                 alpha = 0.5, linewidth = 0.8) +
    geom_point(alpha = 0.8) +
    facet_grid(cluster_label ~ ., 
               scales = "free_y", space = "free_y") +
  geom_text_repel(data = plotting_aboral %>% filter(is_top),
                  aes(label = term_wrapped),
                  size = 2.5,
                  hjust = 0,
                  nudge_x = 0.3,
                  direction = "y",
                  color = "black",
                  fontface = "bold",
                  segment.color = "gray40",
                  segment.size = 0.3,
                  box.padding = 0.3,
                  max.overlaps = 30) + 
  labs(title = "Aboral Tissue",
         x = "-log10(p-value)",
         y = NULL) +
    theme_minimal(base_size = 10) +
    theme(plot.title = element_text(face = "bold", size = 12),
          axis.text.y = element_text(size = 7.5),
          strip.text.y = element_text(size = 8.5, angle = 0, hjust = 0, face = "bold"),
          legend.position = "none",
          panel.spacing = unit(0.5, "lines"),
          panel.grid.major.y = element_blank(),
          panel.grid.minor = element_blank()) +
    scale_size_continuous(range = c(2, 5))
library(patchwork)
p2 <- p_oral + p_aboral +
  plot_annotation(
    title = "Enriched GO Terms by Semantic Cluster",
    subtitle = "Diamond = most significant term per cluster | Size = significance",
    theme = theme(plot.title = element_text(face = "bold", size = 14))
  )


ggsave("../output_RNA/differential_expression/semantic-enrichment/All_sig_BP_terms_by_cluster.pdf", p2, width = 14, height = 16)
```

```{r}
p_oral <-ggplot(plotting_oral, 
       aes(x = `Oral_elim.-log10_pvalue`, 
           y = reorder(GO.ID, `Oral_elim.-log10_pvalue`),
           size = `Oral_elim.-log10_pvalue`)) +
  geom_segment(aes(xend = 0, yend = reorder(GO.ID, `Oral_elim.-log10_pvalue`)),
               alpha = 0.4, linewidth = 0.6, color = "#FD8D3C") +
  geom_point(alpha = 0.7, color = "#FD8D3C") +
    facet_grid(cluster_label ~ ., 
               scales = "free_y", space = "free_y") +
    geom_text_repel(data = plotting_oral %>% filter(is_top),
                  aes(label = str_wrap(term, width = 20)),
                  size = 2,
                  hjust = 0,
                  nudge_x = 0.5,
                  nudge_y = -0.5,
                  direction = "y",
                  color = "black",
                  segment.color = "gray50",
                  segment.size = 0.5,
                  max.overlaps = 30) +
    labs(title = "Oral Epidermis Tissue",
         x = "-log10(p-value)",
         y = NULL) +
    theme_minimal(base_size = 11) +
    theme(plot.title = element_text(face = "bold", size = 12),
          axis.text.y = element_text(size = 7.5),
          strip.text.y = element_text(size = 7.5, angle = 0, hjust = 0, face = "bold"),
          legend.position = "none",
          panel.spacing = unit(.15, "lines"),
          panel.grid.major.y = element_blank(),
          panel.grid.minor = element_blank()) +
    scale_size_continuous(range = c(2, 5))

p_aboral <- ggplot(plotting_aboral, 
         aes(x = `Aboral_elim.-log10_pvalue`, 
             y = reorder(GO.ID, `Aboral_elim.-log10_pvalue`),
             color = factor(GO.cluster),
             size = `Aboral_elim.-log10_pvalue`)) +
    geom_segment(aes(xend = 0, yend = reorder(GO.ID, `Aboral_elim.-log10_pvalue`)),
               alpha = 0.4, linewidth = 0.6, color = "#756BB1") +
  geom_point(alpha = 0.7, color = "#756BB1") +
      facet_grid(cluster_label ~ ., 
               scales = "free_y", space = "free_y") +
  geom_text_repel(data = plotting_aboral %>% filter(is_top),
                  aes(label = str_wrap(term, width = 20)),
                  size = 2,
                  hjust = 0,
                  nudge_x = 0.5,
                  nudge_y = -0.5,
                  direction = "y",
                  color = "black",
                  segment.color = "gray50",
                  segment.size = 0.5,
                  max.overlaps = 30) +
  labs(title = "Aboral Tissue",
         x = "-log10(p-value)",
         y = NULL) +
   theme_minimal(base_size = 11) +
    theme(plot.title = element_text(face = "bold", size = 12),
          axis.text.y = element_text(size = 7.5),
          strip.text.y = element_text(size = 7.5, angle = 0, hjust = 0, face = "bold"),
          legend.position = "none",
          panel.spacing = unit(.15, "lines"),
          panel.grid.major.y = element_blank(),
          panel.grid.minor = element_blank()) +
    scale_size_continuous(range = c(2, 5))

p2 <- p_oral + p_aboral +
  plot_annotation(
    title = "Tissue-specific GO term enrichment patterns",
    theme = theme(plot.title = element_text(face = "bold", size = 14))
  )


ggsave("../output_RNA/differential_expression/semantic-enrichment/All_sig_BP_terms_by_cluster_colored.pdf", p2, width = 14, height = 16)
ggsave("../output_RNA/differential_expression/semantic-enrichment/All_sig_BP_terms_by_cluster_colored2.png", p2, width = 8, height = 12, dpi = 300)
```



### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  dplyr::select(Cluster.name, Tissue) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Tissue=="Oral"),
    Aboral_terms = sum(Tissue=="Aboral"),
    .groups = "drop"
  ) %>%
  mutate(
    deviation = Aboral_terms - Oral_terms,
    Oral_terms = -Oral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

oral_to_equal <- sum(transition_data$deviation >= 0) + 0.5
equal_to_aboral <- sum(transition_data$deviation > 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/GO_enrichment_bidirectional_clustered_separate.png", p, width = 10, height = 8, dpi = 300)
```

## Both tissues together:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
BP_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Oral_elim = c("BP_Oral", "elim_Oral"),
                 Aboral_elim = c("BP_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
BP_Results
ViSEAGO::show_table(BP_Results)

# print the merged table in a file
ViSEAGO::show_table(
    BP_Results,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05.tsv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
        Aboral_elim : 0.01
- enrich GOs (in at least one list): 185 GO terms of 2 conditions.
        Oral_elim : 67 terms
        Aboral_elim : 118 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=BP_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2 <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 4
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2,
    "../output_RNA/differential_expression/semantic-enrichment/DE_05_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOclusters")
```


```{r}
# GOclusters heatmap
Wang_clusters_wardD2<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOclusters")
```

## Custom plotting of results

```{r}
clustered_allDEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) %>%
  dplyr::select(cluster, Cluster.name)

cluster_map
```

### Manually fix cluster names based on GO terms inside each cluster

```{r}
cluster_map_fixed <- cluster_map #%>%
  # mutate(Cluster.name = str_replace(Cluster.name, "Cluster 2: developmental process", "Cluster 2: cell fate determination"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 4: anatomical structure development", "Cluster 4: regeneration & tissue morphogenesis"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 5: multicellular organism development", "Cluster 5: multicellular organism pattern formation"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 9: transport", "Cluster 9: metal ion transmembrane transport"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 10: transport", "Cluster 10: amino acid & small molecule transmembrane transport"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 12: nervous system process", "Cluster 12: sensory perception"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 16: biological_process", "Cluster 16: regeneration & wound healing"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 18: secretion by cell", "Cluster 18: neurotransmitter and hormone secretion"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 24: external encapsulating structure organization", "Cluster 24: extracellular matrix organization"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 25: biological_process", "Cluster 25: cell organization and homeostasis")
  #        )
```


```{r}
clustered_allDEGs_enrichedGO <- clustered_allDEGs_enrichedGO %>%
  left_join(cluster_map_fixed, by = c("GO.cluster" = "cluster"))

write.csv(clustered_allDEGs_enrichedGO,"../output_RNA/differential_expression/semantic-enrichment/DE_05_clusters_named.csv")

library(DT)

# Interactive table with scrolling
datatable(clustered_allDEGs_enrichedGO,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))
```

### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  dplyr::select(Cluster.name, Oral_elim.Significant_genes, Aboral_elim.Significant_genes) %>%
  mutate(
    # Clean cluster names
    #Cluster.name = gsub("^Cluster \\d+:\\s*", "", Cluster.name),
    Oral_terms = ifelse(!is.na(Oral_elim.Significant_genes), 1, 0),
    Aboral_terms = ifelse(!is.na(Aboral_elim.Significant_genes), 1, 0)
  ) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Oral_terms),
    Aboral_terms = sum(Aboral_terms),
    .groups = "drop"
  ) %>%
  filter(Oral_terms + Aboral_terms > 0) %>%
  mutate(
    deviation = Aboral_terms - Oral_terms,
    Oral_terms = -Oral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

datatable(cluster_data,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

oral_to_equal <- sum(transition_data$deviation >= 0) + 0.5
equal_to_aboral <- sum(transition_data$deviation > 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/GO_enrichment_bidirectional.png", p, width = 10, height = 8, dpi = 300)

print(p)
# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) +  #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12, hjust=0)
  ) +
  scale_y_continuous(labels = abs) +
  scale_x_discrete(position = "top") +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/GO_enrichment_bidirectional_rightaxis.png", p, width = 10, height = 8, dpi = 300)

print(p)

```

```{r}
# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  #coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "right",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12,angle = 65, hjust = 1),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

p

ggsave("../output_RNA/differential_expression/semantic-enrichment/GO_enrichment_bidirectional_horiz.png", p, width = 12, height = 8, dpi = 300)
```


# CC
## Oral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Oral)[selection_Oral==1]
background <- names(expressed)

CC_Oral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="CC",
    nodeSize=5
)

# perform TopGO test using elim algorithm
elim_Oral <- topGO::runTest(
    CC_Oral,
    algorithm ="elim",
    statistic = "fisher"
)

CC_Results <- ViSEAGO::merge_enrich_terms(
    Input = list( Oral_elim = c("CC_Oral", "elim_Oral")))
```


### Visualize and save initial results

```{r}
# display the merged table
CC_Results
ViSEAGO::show_table(CC_Results)

# print the merged table in a file
ViSEAGO::show_table(
    CC_Results,
    "../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_Oral.tsv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
- enrich GOs (in at least one list):
        Oral_elim : 19 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=CC_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2_Oral <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2_Oral)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2_Oral,
    "../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_Oral_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Oral,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2_Oral<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2_Oral,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Oral,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2_Oral <-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2_Oral,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    "GOclusters")
```

## Aboral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Aboral)[selection_Aboral==1]
background <- names(expressed)

CC_Aboral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="CC",
    nodeSize=5
)

# perform TopGO test using elim algorithm
elim_Aboral <- topGO::runTest(
    CC_Aboral,
    algorithm ="elim",
    statistic = "fisher"
)

CC_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Aboral_elim = c("CC_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
CC_Results
ViSEAGO::show_table(CC_Results)

# print the merged table in a file
ViSEAGO::show_table(
    CC_Results,
    "../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_Aboral.tsv"
)
```

 - enrichment pvalue cutoff:
        Aboral_elim : 0.01
- enrich GOs (in at least one list):
        Aboral_elim : 23 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=CC_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2_Aboral <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2_Aboral)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2_Aboral,
    "../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_Aboral_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Aboral,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2_Aboral<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2_Aboral,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Aboral,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2_Aboral<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2_Aboral,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOclusters")
```

## Plotting of both tissues, pre-clustered by tissue

## Custom plotting of results

```{r}
clustered_Oral_DEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2_Oral@enrich_GOs@data) 

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2_Oral,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_",""),
    Cluster.term = str_replace(cluster_full,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) #%>%
  #mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) #%>%
  #dplyr::select(cluster, Cluster.name)

clustered_Oral_DEGs_enrichedGO <- clustered_Oral_DEGs_enrichedGO %>%
  left_join(cluster_map, by = c("GO.cluster" = "cluster")) %>%
  mutate(Tissue="Oral")

clustered_Aboral_DEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2_Aboral@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2_Aboral,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_",""),
    Cluster.term = str_replace(cluster_full,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) #%>%
  #mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) #%>%
  #dplyr::select(cluster, Cluster.name)
clustered_Aboral_DEGs_enrichedGO <- clustered_Aboral_DEGs_enrichedGO %>%
  left_join(cluster_map, by = c("GO.cluster" = "cluster")) %>%
  mutate(Tissue="Aboral")


clustered_allDEGs_enrichedGO <- rbind(clustered_Oral_DEGs_enrichedGO %>%
                                        dplyr::select(-contains("Oral")),
                                      clustered_Aboral_DEGs_enrichedGO %>%
                                        dplyr::select(-contains("Aboral")))

```



### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  dplyr::select(Cluster.name, Tissue) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Tissue=="Oral"),
    Aboral_terms = sum(Tissue=="Aboral"),
    .groups = "drop"
  ) %>%
  mutate(
    deviation = Aboral_terms - Oral_terms,
    Oral_terms = -Oral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

oral_to_equal <- sum(transition_data$deviation >= 0) + 0.5
equal_to_aboral <- sum(transition_data$deviation > 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/CC_GO_enrichment_bidirectional_clustered_separate.png", p, width = 10, height = 8, dpi = 300)
```

## Both tissues together:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
CC_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Oral_elim = c("CC_Oral", "elim_Oral"),
                 Aboral_elim = c("CC_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
CC_Results
ViSEAGO::show_table(CC_Results)

# print the merged table in a file
ViSEAGO::show_table(
    CC_Results,
    "../output_RNA/differential_expression/semantic-enrichment/CC_DE_05.tsv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
        Aboral_elim : 0.01
- enrich GOs (in at least one list): 40 GO terms of 2 conditions.
        Oral_elim : 19 terms
        Aboral_elim : 23 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=CC_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2 <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 4
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2,
    "../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOclusters")
```


```{r}
# GOclusters heatmap
Wang_clusters_wardD2<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOclusters")
```

## Custom plotting of results

```{r}
clustered_allDEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) %>%
  select(cluster, Cluster.name)

cluster_map
```

### Manually fix cluster names based on GO terms inside each cluster

```{r}
cluster_map_fixed <- cluster_map #%>%
  # mutate(Cluster.name = str_replace(Cluster.name, "Cluster 2: developmental process", "Cluster 2: cell fate determination"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 4: anatomical structure development", "Cluster 4: regeneration & tissue morphogenesis"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 5: multicellular organism development", "Cluster 5: multicellular organism pattern formation"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 9: transport", "Cluster 9: metal ion transmembrane transport"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 10: transport", "Cluster 10: amino acid & small molecule transmembrane transport"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 12: nervous system process", "Cluster 12: sensory perception"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 16: biological_process", "Cluster 16: regeneration & wound healing"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 18: secretion by cell", "Cluster 18: neurotransmitter and hormone secretion"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 24: external encapsulating structure organization", "Cluster 24: extracellular matrix organization"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 25: biological_process", "Cluster 25: cell organization and homeostasis")
  #        )
```


```{r}
clustered_allDEGs_enrichedGO <- clustered_allDEGs_enrichedGO %>%
  left_join(cluster_map_fixed, by = c("GO.cluster" = "cluster"))

write.csv(clustered_allDEGs_enrichedGO,"../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_clusters_named.csv")

library(DT)

# Interactive table with scrolling
datatable(clustered_allDEGs_enrichedGO,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))
```

### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  select(Cluster.name, Oral_elim.Significant_genes, Aboral_elim.Significant_genes) %>%
  mutate(
    # Clean cluster names
    #Cluster.name = gsub("^Cluster \\d+:\\s*", "", Cluster.name),
    Oral_terms = ifelse(!is.na(Oral_elim.Significant_genes), 1, 0),
    Aboral_terms = ifelse(!is.na(Aboral_elim.Significant_genes), 1, 0)
  ) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Oral_terms),
    Aboral_terms = sum(Aboral_terms),
    .groups = "drop"
  ) %>%
  filter(Oral_terms + Aboral_terms > 0) %>%
  mutate(
    deviation = Aboral_terms - Oral_terms,
    Oral_terms = -Oral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

datatable(cluster_data,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

oral_to_equal <- sum(transition_data$deviation >= 0) + 0.5
equal_to_aboral <- sum(transition_data$deviation > 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/CC_GO_enrichment_bidirectional.png", p, width = 10, height = 8, dpi = 300)

print(p)
# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) +  #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12, hjust=0)
  ) +
  scale_y_continuous(labels = abs) +
  scale_x_discrete(position = "top") +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/CC_GO_enrichment_bidirectional_rightaxis.png", p, width = 10, height = 8, dpi = 300)

print(p)

```

```{r}
# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  #coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "right",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12,angle = 65, hjust = 1),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

p

ggsave("../output_RNA/differential_expression/semantic-enrichment/CC_GO_enrichment_bidirectional_horiz.png", p, width = 12, height = 8, dpi = 300)
```


# MF
## Oral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Oral)[selection_Oral==1]
background <- names(expressed)

MF_Oral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="MF",
    nodeSize=5
)

# perform TopGO test using elim algorithm
elim_Oral <- topGO::runTest(
    MF_Oral,
    algorithm ="elim",
    statistic = "fisher"
)

MF_Results <- ViSEAGO::merge_enrich_terms(
    Input = list( Oral_elim = c("MF_Oral", "elim_Oral")))
```


### Visualize and save initial results

```{r}
# display the merged table
MF_Results
ViSEAGO::show_table(MF_Results)

# print the merged table in a file
ViSEAGO::show_table(
    MF_Results,
    "../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_Oral.tsv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
- enrich GOs (in at least one list):
        Oral_elim : 32 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=MF_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2_Oral <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2_Oral)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2_Oral,
    "../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_Oral_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Oral,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2_Oral<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2_Oral,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Oral,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2_Oral <-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2_Oral,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Oral,
    "GOclusters")
```

## Aboral Epidermis:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
# create viseago object
selection <- names(selection_Aboral)[selection_Aboral==1]
background <- names(expressed)

MF_Aboral <- ViSEAGO::create_topGOdata(
    geneSel=selection,
    allGenes=background,
    gene2GO=myGENE2GO_Pacuta, 
    ont="MF",
    nodeSize=5
)

# perform TopGO test using elim algorithm
elim_Aboral <- topGO::runTest(
    MF_Aboral,
    algorithm ="elim",
    statistic = "fisher"
)

MF_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Aboral_elim = c("MF_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
MF_Results
ViSEAGO::show_table(MF_Results)

# print the merged table in a file
ViSEAGO::show_table(
    MF_Results,
    "../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_Aboral.tsv"
)
```

 - enrichment pvalue cutoff:
        Aboral_elim : 0.01
- enrich GOs (in at least one list):
        Aboral_elim : 28 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=MF_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2_Aboral <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 2
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2_Aboral)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2_Aboral,
    "../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_Aboral_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Aboral,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2_Aboral<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2_Aboral,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2_Aboral,
    "GOclusters")
```

```{r}
# GOclusters heatmap
Wang_clusters_wardD2_Aboral<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2_Aboral,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2_Aboral,
    "GOclusters")
```

## Plotting of both tissues, pre-clustered by tissue

## Custom plotting of results

```{r}
clustered_Oral_DEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2_Oral@enrich_GOs@data) 

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2_Oral,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_",""),
    Cluster.term = str_replace(cluster_full,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) #%>%
  #mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) #%>%
  #dplyr::select(cluster, Cluster.name)

clustered_Oral_DEGs_enrichedGO <- clustered_Oral_DEGs_enrichedGO %>%
  left_join(cluster_map, by = c("GO.cluster" = "cluster")) %>%
  mutate(Tissue="Oral")

clustered_Aboral_DEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2_Aboral@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2_Aboral,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_",""),
    Cluster.term = str_replace(cluster_full,"^\\d+_",""),
    Cluster.term = str_replace(Cluster.term,"_.*$","")
  ) #%>%
  #mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) #%>%
  #dplyr::select(cluster, Cluster.name)
clustered_Aboral_DEGs_enrichedGO <- clustered_Aboral_DEGs_enrichedGO %>%
  left_join(cluster_map, by = c("GO.cluster" = "cluster")) %>%
  mutate(Tissue="Aboral")


clustered_allDEGs_enrichedGO <- rbind(clustered_Oral_DEGs_enrichedGO %>%
                                        dplyr::select(-contains("Oral")),
                                      clustered_Aboral_DEGs_enrichedGO %>%
                                        dplyr::select(-contains("Aboral")))

```



### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  dplyr::select(Cluster.name, Tissue) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Tissue=="Oral"),
    Aboral_terms = sum(Tissue=="Aboral"),
    .groups = "drop"
  ) %>%
  mutate(
    deviation = Aboral_terms - Oral_terms,
    Oral_terms = -Oral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

oral_to_equal <- sum(transition_data$deviation >= 0) + 0.5
equal_to_aboral <- sum(transition_data$deviation > 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/MF_GO_enrichment_bidirectional_clustered_separate.png", p, width = 10, height = 8, dpi = 300)
```

## Both tissues together:

### create topGO objects and perform enrichment using topGO wrapped by ViSEAGO

```{r}
MF_Results <- ViSEAGO::merge_enrich_terms(
    Input = list(Oral_elim = c("MF_Oral", "elim_Oral"),
                 Aboral_elim = c("MF_Aboral", "elim_Aboral"))
)
```


### Visualize and save initial results

```{r}
# display the merged table
MF_Results
ViSEAGO::show_table(MF_Results)

# print the merged table in a file
ViSEAGO::show_table(
    MF_Results,
    "../output_RNA/differential_expression/semantic-enrichment/MF_DE_05.tsv"
)
```

 - enrichment pvalue cutoff:
        Oral_elim : 0.01
        Aboral_elim : 0.01
- enrich GOs (in at least one list): 59 GO terms of 2 conditions.
        Oral_elim : 32 terms
        Aboral_elim : 28 terms

### Semantic similarity --> Cluster: Use Wang graph-based method to identify similarity between GO terms

```{r}
# initialize 
myGOs<-ViSEAGO::build_GO_SS(
    gene2GO=myGENE2GO_Pacuta,
    enrich_GO_terms=MF_Results
)

myGOs <- ViSEAGO::compute_SS_distances(
    myGOs,
   distance=c("Wang")
)

myGOs
```

#### Multi Dimensional Scaling Plot of the GO-terms based on semantic similarity

```{r, eval=FALSE}
# display MDSplot
ViSEAGO::MDSplot(myGOs,
                 "GOterms")
```

#### Heatmap + table of GO terms, with clusters aggregated by hclust with ward.D2 method

```{r}
# GOterms heatmap with the default parameters
Wang_clusters_wardD2 <- ViSEAGO::GOterms_heatmap(
  myGOs,
  showIC = TRUE,
  showGOlabels = TRUE,
  GO.tree = list(
    tree = list(distance = "Wang", aggreg.method = "ward.D2"),
    cut = list(
      dynamic = list(
        pamStage = TRUE,
        pamRespectsDendro = TRUE,
        deepSplit = 2,
        minClusterSize = 4
      )
    )
  ),
  samples.tree = NULL
)

# Display the clusters-heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOterms")

# Display the clusters-heatmap table
ViSEAGO::show_table(Wang_clusters_wardD2)

# Print the clusters-heatmap table
ViSEAGO::show_table(
    Wang_clusters_wardD2,
    "../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_cluster_heatmap_Wang_wardD2.tsv"
)
```

#### Multi Dimensional Scaling Plot of the enriched GO-terms based on semantic similarity, colored by cluster

```{r, eval=FALSE}
# display colored MDSplot
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOterms")
```

### Further summarizing the Semantic Similarity-based clusters

```{r}
# calculate semantic similarities between clusters of GO terms
Wang_clusters_wardD2<-ViSEAGO::compute_SS_distances(
    Wang_clusters_wardD2,
    distance=c("max", "avg","rcmax", "BMA")
)
```

```{r, eval=FALSE}
# MDSplot - one point per cluster
ViSEAGO::MDSplot(
    Wang_clusters_wardD2,
    "GOclusters")
```


```{r}
# GOclusters heatmap
Wang_clusters_wardD2<-ViSEAGO::GOclusters_heatmap(
    Wang_clusters_wardD2,
    tree=list(
        distance="BMA",
        aggreg.method="ward.D2"
    )
)

# display the GOClusters heatmap
ViSEAGO::show_heatmap(
    Wang_clusters_wardD2,
    "GOclusters")
```

## Custom plotting of results

```{r}
clustered_allDEGs_enrichedGO <- as.data.frame(Wang_clusters_wardD2@enrich_GOs@data)

cluster_map <- data.frame(cluster_full = rownames(as.matrix(slot(Wang_clusters_wardD2,"clusters_dist")[["BMA"]])))
cluster_map <- cluster_map %>% 
  mutate(
    cluster = str_extract(cluster_full, "\\d+"),
    Cluster.name = str_replace(cluster_full,".*?_.*?_","")
  ) %>%
  mutate(Cluster.name = paste0("Cluster ", cluster, ": ", Cluster.name)) %>%
  select(cluster, Cluster.name)

cluster_map
```

### Manually fix cluster names based on GO terms inside each cluster

```{r}
cluster_map_fixed <- cluster_map #%>%
  # mutate(Cluster.name = str_replace(Cluster.name, "Cluster 2: developmental process", "Cluster 2: cell fate determination"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 4: anatomical structure development", "Cluster 4: regeneration & tissue morphogenesis"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 5: multicellular organism development", "Cluster 5: multicellular organism pattern formation"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 9: transport", "Cluster 9: metal ion transmembrane transport"),
  #        Cluster.name = str_replace(Cluster.name, "Cluster 10: transport", "Cluster 10: amino acid & small molecule transmembrane transport"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 12: nervous system process", "Cluster 12: sensory perception"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 16: biological_process", "Cluster 16: regeneration & wound healing"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 18: secretion by cell", "Cluster 18: neurotransmitter and hormone secretion"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 24: external encapsulating structure organization", "Cluster 24: extracellular matrix organization"),
  #       Cluster.name = str_replace(Cluster.name, "Cluster 25: biological_process", "Cluster 25: cell organization and homeostasis")
  #        )
```


```{r}
clustered_allDEGs_enrichedGO <- clustered_allDEGs_enrichedGO %>%
  left_join(cluster_map_fixed, by = c("GO.cluster" = "cluster"))

write.csv(clustered_allDEGs_enrichedGO,"../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_clusters_named.csv")

library(DT)

# Interactive table with scrolling
datatable(clustered_allDEGs_enrichedGO,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))
```

### Plot

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)

# Create bidirectional data
cluster_data <- clustered_allDEGs_enrichedGO %>%
  select(Cluster.name, Oral_elim.Significant_genes, Aboral_elim.Significant_genes) %>%
  mutate(
    # Clean cluster names
    #Cluster.name = gsub("^Cluster \\d+:\\s*", "", Cluster.name),
    Oral_terms = ifelse(!is.na(Oral_elim.Significant_genes), 1, 0),
    Aboral_terms = ifelse(!is.na(Aboral_elim.Significant_genes), 1, 0)
  ) %>%
  group_by(Cluster.name) %>%
  summarise(
    Oral_terms = sum(Oral_terms),
    Aboral_terms = sum(Aboral_terms),
    .groups = "drop"
  ) %>%
  filter(Oral_terms + Aboral_terms > 0) %>%
  mutate(
    deviation = Aboral_terms - Oral_terms,
    Oral_terms = -Oral_terms  # Make negative for bidirectional plot
  ) %>%
  pivot_longer(cols = c("Oral_terms", "Aboral_terms"), 
               names_to = "Tissue", values_to = "n_terms") %>%
  mutate(Tissue = ifelse(Tissue == "Oral_terms", "Oral epidermis", "Aboral"))

datatable(cluster_data,
          options = list(
            pageLength = 10,   # rows per page
            scrollX = TRUE,    # horizontal scroll if wide
            scrollY = "400px"  # vertical scroll
          ))

# Calculate transition points for dashed lines
transition_data <- cluster_data %>%
  distinct(Cluster.name, deviation) %>%
  arrange(deviation)

oral_to_equal <- sum(transition_data$deviation >= 0) + 0.5
equal_to_aboral <- sum(transition_data$deviation > 0) + 0.5

# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/MF_GO_enrichment_bidirectional.png", p, width = 10, height = 8, dpi = 300)

print(p)
# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) +  #green: "#2E8B57"
  coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "bottom",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12),
    axis.text.y = element_text(size = 12, hjust=0)
  ) +
  scale_y_continuous(labels = abs) +
  scale_x_discrete(position = "top") +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/MF_GO_enrichment_bidirectional_rightaxis.png", p, width = 10, height = 8, dpi = 300)

print(p)

```

```{r}
# Plot
p <- ggplot(cluster_data, aes(x = reorder(Cluster.name, -deviation), y = n_terms, fill = Tissue)) +
  geom_col(width = 0.7, alpha = 0.8) +
  geom_hline(yintercept = 0, color = "black", linewidth = 0.8) +
  geom_vline(xintercept = c(oral_to_equal, equal_to_aboral), 
             color = "grey60", linetype = "dashed", linewidth = 0.6) +
  scale_fill_manual(values = c("Oral epidermis" = "#FD8D3C", "Aboral" = "#756BB1")) + #green: "#2E8B57"
  #coord_flip() +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor = element_blank(),
    legend.position = "right",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    axis.text.x = element_text(size = 12,angle = 65, hjust = 1),
    axis.text.y = element_text(size = 12)
  ) +
  scale_y_continuous(labels = abs) +
  labs(
    x = "Gene clusters",
    y = "Number of enriched GO terms",
    title = "Tissue-specific GO term enrichment patterns",
    subtitle = "Clusters ordered by net enrichment (Oral vs. Aboral)"
  )

p

ggsave("../output_RNA/differential_expression/semantic-enrichment/MF_GO_enrichment_bidirectional_horiz.png", p, width = 12, height = 8, dpi = 300)
```



## Combine everything into one big plot:

```{r}
BP_GOs_Aboral <- read.delim(file="../output_RNA/differential_expression/semantic-enrichment/DE_05_Aboral.tsv") %>% mutate(ontology="BP")
MF_GOs_Aboral <- read.delim(file="../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_Aboral.tsv")   %>% mutate(ontology="MF")
CC_GOs_Aboral <- read.delim(file="../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_Aboral.tsv") %>% mutate(ontology="CC")

BP_GOs_Oral <- read.delim(file="../output_RNA/differential_expression/semantic-enrichment/DE_05_Oral.tsv") %>% mutate(ontology="BP")
MF_GOs_Oral <- read.delim(file="../output_RNA/differential_expression/semantic-enrichment/MF_DE_05_Oral.tsv") %>% mutate(ontology="MF")
CC_GOs_Oral <- read.delim(file="../output_RNA/differential_expression/semantic-enrichment/CC_DE_05_Oral.tsv") %>% mutate(ontology="CC")

oral_combined <- rbind(BP_GOs_Oral,MF_GOs_Oral,CC_GOs_Oral) %>%
                        rename_with(~ str_replace_all(.x, "Oral_elim|\\.", "")) %>% mutate(Tissue="Oral")
aboral_combined <- rbind(BP_GOs_Aboral,MF_GOs_Aboral,CC_GOs_Aboral) %>%
                        rename_with(~ str_replace_all(.x, "Aboral_elim|\\.", "")) %>% mutate(Tissue="Aboral")

tissues_combined <- rbind(oral_combined,aboral_combined) %>%
                          mutate(freq = as.numeric(str_remove(genes_frequency,pattern="\\%.*")))
```

```{r}
aboral_cols <- c(
  BP = "#756BB1",
  CC = "#9E9AC8",
  MF = "#CBC9E2"
)

oral_cols <- c(
  BP = "#FD8D3C",
  CC = "#FDAE6B",
  MF = "#FDD0A2"
)

top10_pval <- tissues_combined %>%
  filter(ontology=="BP") %>%
  group_by(Tissue,ontology) %>%
  slice_max(order_by = desc(pvalue), n = 30, with_ties = FALSE) %>% 
  arrange(Tissue, ontology, log10_pvalue) %>%
  ungroup() %>%
  mutate(term=str_wrap(`term`, width = 40))

p1 <- top10_pval %>% filter(Tissue == "Aboral") %>%
  mutate(term = fct_reorder(term, log10_pvalue)) %>% 
  ggplot(aes(x = term, y = log10_pvalue, fill = ontology)) +
  geom_col(width = .8, fill="#756BB1", alpha = 0.8) +
  facet_grid(ontology ~ ., scales = "free_y", space = "free_y") +
  coord_flip() +
 # scale_fill_manual(values = aboral_cols) +
  theme_bw(base_size = 9) +
  #scale_y_continuous(limits = c(0, 35), expand = expansion(mult = c(0, 0.05)))+
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Aboral", y = "-log10(p-value)")

p2 <- top10_pval %>% filter(Tissue == "Oral") %>%
  mutate(term = fct_reorder(term, log10_pvalue)) %>% 
  ggplot(aes(x = term, y = log10_pvalue, fill = ontology)) +
  geom_col(width = .8, fill="#FD8D3C", alpha = 0.8) +
  facet_grid(ontology ~ ., scales = "free_y", space = "free_y") +
  coord_flip() +
 # scale_fill_manual(values = oral_cols) +
 # scale_y_continuous(limits = c(0, 35), expand = expansion(mult = c(0, 0.05)))+
  theme_bw(base_size = 9) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
     strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Oral epidermis", y = "-log10(p-value)")

plot <- p1 + p2 +
  plot_annotation(
    title = "Top 30 enriched GO terms by p-value",
    theme = theme(plot.title = element_text(face = "bold", size = 9, hjust = 0.5))
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/Top_BP_30.png", plot, width = 7, height = 6.5, dpi = 300)
```

```{r}
top10_freq <- tissues_combined %>% group_by(Tissue,ontology) %>%
  slice_max(order_by = freq, n = 10, with_ties = FALSE) %>% 
  arrange(Tissue, ontology, log10_pvalue) %>%
  ungroup() %>%
  mutate(term=str_wrap(`term`, width = 40))


p1 <- top10_freq %>% filter(Tissue == "Aboral") %>%
  mutate(term = fct_reorder(term, freq)) %>% 
  ggplot(aes(x = term, y = freq, fill = ontology)) +
  geom_col(width = 0.7, alpha = 0.85) +
  facet_grid(ontology ~ ., scales = "free_y", space = "free_y") +
  coord_flip() +
  scale_fill_manual(values = aboral_cols) +
  theme_bw(base_size = 9) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
     strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Aboral", y = "% DEGs with GO term of\nall genes with GO term")

p2 <- top10_freq %>% filter(Tissue == "Oral") %>%
  mutate(term = fct_reorder(term, freq)) %>% 
  ggplot(aes(x = term, y = freq, fill = ontology)) +
  geom_col(width = 0.7, alpha = 0.85) +
  facet_grid(ontology ~ ., scales = "free_y", space = "free_y") +
  coord_flip() +
  scale_fill_manual(values = oral_cols) +
  theme_bw(base_size = 9) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7,color = "black"),
    strip.text = element_text(face = "bold", size = 9),
     strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "none",
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Oral epidermis", y = "% DEGs with GO term of\nall genes with GO term")


plot <- p1 + p2 +
  plot_annotation(
    title = "Top 10 enriched GO terms by frequency in tissue DEGs",
    theme = theme(plot.title = element_text(face = "bold", size = 9, hjust = 0.5))
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/All_Onts_frq.png", plot, width = 7, height = 6.5, dpi = 300)
```


```{r}
# Create gradient scales for each tissue
aboral_gradient <- colorRampPalette(c("#CBC9E2", "#756BB1"))(100)
oral_gradient <- colorRampPalette(c("#FDD0A2", "#FD8D3C"))(100)

p1 <- top10_freq %>% filter(Tissue == "Aboral") %>%
  mutate(term = fct_reorder(term, freq)) %>% 
  ggplot(aes(x = term, y = freq, fill = log10_pvalue)) +
  geom_col(width = 0.7, alpha = 0.85) +
  facet_grid(ontology ~ ., scales = "free_y", space = "free_y") +
  coord_flip() +
  scale_fill_gradientn(colors = aboral_gradient, name = "-log10(p-value)") +
  theme_bw(base_size = 9) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7, color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "right",
    legend.key.height = unit(0.8, "cm"),
    legend.key.width = unit(0.3, "cm"),
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Aboral", y = "% DEGs with GO term of\nall genes with GO term")

p2 <- top10_freq %>% filter(Tissue == "Oral") %>%
  mutate(term = fct_reorder(term, freq)) %>% 
  ggplot(aes(x = term, y = freq, fill = log10_pvalue)) +
  geom_col(width = 0.7, alpha = 0.85) +
  facet_grid(ontology ~ ., scales = "free_y", space = "free_y") +
  coord_flip() +
  scale_fill_gradientn(colors = oral_gradient, name = "-log10(p-value)") +
  theme_bw(base_size = 9) +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.major.x = element_line(color = "grey90", linewidth = 0.2),
    axis.title.y = element_blank(),
    axis.text.y = element_text(size = 7, lineheight = 0.7, color = "black"),
    strip.text = element_text(face = "bold", size = 9),
    strip.background = element_rect(fill = "white", color = "black"),
    legend.position = "right",
    legend.key.height = unit(0.8, "cm"),
    legend.key.width = unit(0.3, "cm"),
    plot.subtitle = element_text(face = "bold", size = 9, hjust = 0.5),
    panel.spacing.y = unit(0.2, "lines")
  ) + labs(subtitle = "Oral epidermis", y = "% DEGs with GO term of\nall genes with GO term")

plot <- p1 + p2 +
  plot_annotation(
    title = "Top 10 enriched GO terms by frequency in tissue DEGs",
    theme = theme(plot.title = element_text(face = "bold", size = 9, hjust = 0.5))
  )

ggsave("../output_RNA/differential_expression/semantic-enrichment/All_Onts_freq_pval.png", plot, width = 8, height = 6.5, dpi = 300)
```

