---
title: "07-Marker-Genes"
author: "Zoe Dellaert"
date: "2025-01-13"
output:
  github_document:
    toc: true
    toc_depth: 3
    number_sections: true
    html_preview: true
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Cell type Marker Gene Lists from other cnidarians

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Managing Packages Using Renv

To run this code in my project using the renv environment, run the following lines of code

```{r, eval=FALSE}
install.packages("renv") #install the package on the new computer (may not be necessary if renv bootstraps itself as expected)
renv::restore() #reinstall all the package versions in the renv lockfile
```

## Load packages

```{r}
require("tidyverse")
```


## Gathering marker genes of cell types from cnidarian single cell atlases

- [X] 1 Aiptasia (Cui et al., 2023; 10.1126/sciadv.adf7108)
- [X] 2 Hydra (Siebert et al., 2019; 10.1126/science.aav9314)
  - Note: their published marker genes in the supplement are based on their de novo transcriptome, but they also mapped to the genome and have a seurat object for that
- [X] 3 Hydractinia (Salamanca-Díaz et al., 2025; 10.1038/s41467-025-57168-z)
- [X] 4 Nematostella (Cole et al., 2024;10.1186/s12983-024-00529-z)
- [X] 5 Oculina (Valadez-Ingersoll et al., 2024; 10.1101/2024.06.20.599951)
- [X] 6 Orbicella (del Campo et al., 2024; 10.21203/rs.3.rs-5397639/v1)
- [X] 7 Pocillopora (Diao et al., 2025; 10.3389/fmars.2025.1551763)
- [X] 8 Stylophora (Levy et al., 2021; 10.1016/j.cell.2021.04.005)
- [X] 9 Xenia (Hu et al., 2020; 10.1038/s41586-020-2385-7)

### To do now:

- [ ] combine marker gene lists from all species
- [ ] ID all genes in the OG of each species of interest as belonging to that marker gene w/ species source

## Okay, lets do it?

1. Rerun orthofinder with the following genomes
   1. Aiptasia (Cui et al., 2023; 10.1126/sciadv.adf7108)
   2. Hydra (Siebert et al., 2019; 10.1126/science.aav9314)
   3. Hydractinia (Salamanca-Díaz et al., 2025; 10.1038/s41467-025-57168-z)
   4. Nematostella (Cole et al., 2024;10.1186/s12983-024-00529-z)
   5. Astrangia (used to map Oculina data by Valadez-Ingersoll et al., 2024; 10.1101/2024.06.20.599951)
   6. Orbicella (del Campo et al., 2024; 10.21203/rs.3.rs-5397639/v1)
   7. Pocillopora damicornis (Diao et al., 2025; 10.3389/fmars.2025.1551763)
   8. Stylophora (Levy et al., 2021; 10.1016/j.cell.2021.04.005)
   9. Xenia (Hu et al., 2020; 10.1038/s41586-020-2385-7)
   10. ***Pocillopora acuta***
   11. ***Porites compressa***
   12. ***Montipora capitata***
2.  Run broccoli on all those genomes
3.  


## Directory Structure: 

All of this will be self contained in a directory called "cnidarian_marker_genes" - can be the same between LaserCoral and MultiSp repos

```
cd reference_genes/cnidarian_marker_genes/
```

### Step 1: Aquire genome protein fasta sequences

```{bash}
mkdir cnidarian_protein_fastas
cd cnidarian_protein_fastas

# 1 - Aiptasia, assembled by Baumgarten et al., 2015 (10.1073/pnas.1513318112)
wget http://aiptasia.reefgenomics.org/download/aiptasia_genome.proteins.fa.gz

# 2 - Hydra, assembled by Steele et al 2015 (https://research.nhgri.nih.gov/hydra/)
wget https://research.nhgri.nih.gov/hydra/download/genemodels_proteins/hydra2.0_genemodels.aa.gz

# 3 - Hydractinia, assembled by Kon-Nanjo et al 2023 (10.1093/g3journal/jkad107)
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/029/227/915/GCF_029227915.1_HSymV2.1/GCF_029227915.1_HSymV2.1_protein.faa.gz

# 4 - Nematostella, assembled by Zimmermann et al., 2023 (10.1038/s41467-023-44080-7)
wget https://simrbase.stowers.org/files/pub/nematostella/Nvec/genomes/Nvec200/aligned/tcs_v2/20240221/NV2g.20240221.protein.fa

# 5 - Astrangia, assembled by Stankiewicz et al., 2025 (10.1093/g3journal/jkaf033)
wget -O apoculata_proteins.fasta.gz https://zenodo.org/records/14110456/files/apoculata_proteins.fasta.gz?download=1

# 6 - Orbicella, assembled by Young et al., 2024 (10.1186/s12864-024-10092-w)
wget -O Orbicella_faveolata_gen_17.proteins.fa https://zenodo.org/records/10151798/files/Orbicella_faveolata_gen_17.proteins.fa?download=1

# 7 - Pocillopora damicornis, assembled by Cunning et al., 2018 (10.1038/s41598-018-34459-8)
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/003/704/095/GCF_003704095.1_ASM370409v1/GCF_003704095.1_ASM370409v1_protein.faa.gz

# 8 - Stylophora, assembled by Voolstra et al., 2017 (10.1038/s41598-017-17484-x)
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/571/385/GCF_002571385.2_Stylophora_pistillata_v1.1/GCF_002571385.2_Stylophora_pistillata_v1.1_protein.faa.gz

# 9 - Xenia, assembled by Hu et al., 2020 (10.1038/s41586-020-2385-7)
wget https://cmo.carnegiescience.edu/endosymbiosis/genome/xenSp1.proteins.fa

# Pocillopora acuta, assembled by Stephens et al., 2022 (10.1093/gigascience/giac098)
wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.pep.faa.gz

# Porites compressa, assembled by Stephens et al., 2022 (10.1093/gigascience/giac098)
wget http://cyanophora.rutgers.edu/porites_compressa/Porites_compressa_HIv1.genes.pep.faa.gz

# Montipora capitata, assembled by Stephens et al., 2022 (10.1093/gigascience/giac098)
wget http://cyanophora.rutgers.edu/montipora/Montipora_capitata_HIv3.genes.pep.faa.gz

gunzip *

mv aiptasia_genome.proteins.fa Edia_proteins.faa
mv hydra2.0_genemodels.aa Hvul_proteins.faa
mv GCF_029227915.1_HSymV2.1_protein.faa Hsym_proteins.faa
mv NV2g.20240221.protein.fa Nvec_proteins.faa
mv apoculata_proteins.fasta Apoc_proteins.faa
mv Orbicella_faveolata_gen_17.proteins.fa Ofav_proteins.faa
mv GCF_003704095.1_ASM370409v1_protein.faa Pdam_proteins.faa
mv GCF_002571385.2_Stylophora_pistillata_v1.1_protein.faa Spis_proteins.faa
mv xenSp1.proteins.fa Xesp_proteins.faa
mv Pocillopora_acuta_HIv2.genes.pep.faa Pacu_proteins.faa
mv Porites_compressa_HIv1.genes.pep.faa Pcom_proteins.faa
mv Montipora_capitata_HIv3.genes.pep.faa Mcap_proteins.faa
```

for reference: *file_shortname*

```
Edia_proteins.faa
Hvul_proteins.faa
Hsym_proteins.faa
Nvec_proteins.faa
Apoc_proteins.faa
Ofav_proteins.faa
Pdam_proteins.faa
Spis_proteins.faa
Xesp_proteins.faa
Pacu_proteins.faa
Pcom_proteins.faa
Mcap_proteins.faa
```

## Run orthofinder on 12 cnidarian genomes

```
cd reference_genes/cnidarian_marker_genes/
```

```{bash, eval = FALSE}
mkdir scripts
mkdir scripts/outs_errs

nano scripts/orthofinder.sh
```

```{bash, eval = FALSE}
#!/usr/bin/env bash
#SBATCH --export=NONE
#SBATCH --ntasks=1 --cpus-per-task=20 #split one task over multiple CPU
#SBATCH --mem=500GB
#SBATCH -p cpu-preempt
#SBATCH -t 48:00:00
#SBATCH --mail-type=END,FAIL,TIME_LIMIT_80 #email you when job stops and/or fails or is nearing its time limit
#SBATCH --error=scripts/outs_errs/"%x_error.%j" #if your job fails, the error report will be put in this file
#SBATCH --output=scripts/outs_errs/"%x_output.%j" #once your job is completed, any final job report comments will be put in this file

mkdir -p output

# run orthofinder with 20 threads
/work/pi_hputnam_uri_edu/pgrams/OrthoFinder/orthofinder -f cnidarian_protein_fastas/ -t 20 -o output/orthofinder_results/
```

## Running broccoli on 12 cnidarian genomes

```
cd reference_genes/cnidarian_marker_genes/
nano scripts/broccoli.sh
```

```{bash, eval = FALSE}
#!/usr/bin/env bash
#SBATCH --export=NONE
#SBATCH --ntasks=1 --cpus-per-task=20 #split one task over multiple CPU
#SBATCH --mem=500GB
#SBATCH -p cpu-preempt
#SBATCH -t 48:00:00
#SBATCH --mail-type=END,FAIL,TIME_LIMIT_80 #email you when job stops and/or fails or is nearing its time limit
#SBATCH --error=scripts/outs_errs/"%x_error.%j" #if your job fails, the error report will be put in this file
#SBATCH --output=scripts/outs_errs/"%x_output.%j" #once your job is completed, any final job report comments will be put in this file

#load conda and activate conda environment
module load conda/latest
conda activate /work/pi_hputnam_uri_edu/conda/envs/env-broccoli

#load additional programs needed to run broccoli
module load uri/main
module load diamond/2.1.7
module load all/FastTree/2.1.11-GCCcore-12.3.0

mkdir -p output/broccoli

cd output/broccoli
python ~/broccoli.py -dir ../../cnidarian_protein_fastas/ -ext '.faa' -path_fasttree FastTree -threads 8
```
  
---

## Custom P. acuta BLASTP database:

```{bash, eval=FALSE}
cd ../references

#download the genome protein fasta if you have not already
wget http://cyanophora.rutgers.edu/Pocillopora_acuta/Pocillopora_acuta_HIv2.genes.pep.faa.gz

#unzip file
gunzip Pocillopora_acuta_HIv2.genes.pep.faa.gz

#request interactive computing node
salloc -p cpu -c 8 --mem 32G

#load BLAST module
module load uri/main
module load BLAST+/2.15.0-gompi-2023a

#make blast_dbs directory if you haven't done so above
mkdir blast_dbs
cd blast_dbs

makeblastdb -in ../Pocillopora_acuta_HIv2.genes.pep.faa -out Pacuta_prot -dbtype prot
```

## HOX genes 

```{bash eval=FALSE}
cd ../output_RNA/differential_expression/

#make blast output directory if you haven't done so above
mkdir blast
cd blast

nano Nematostella_Hox.txt
```

add accession numbers of interest (based on https://doi.org/10.1016/j.cub.2023.05.044):

Alx4	    Homeobox protein aristaless-like 4	             NV2g017442000.1
Anthox1	  Homeobox protein Anthox1	                       NV2g017961000.1
Anthox1a	Homeobox protein Anthox1a	                       NV2g012279000.1
Anthox6a	Homeobox protein Anthox6a	                       NV2g011327000.1
Anthox7	  Homeobox protein Anthox7	                       NV2g012004000.1
Anthox8a	Homeobox protein Anthox8a	                       NV2g012007000.1
Anthox8b	Homeobox protein Anthox8b	                       NV2g012008000.1
B-H1	    Homeobox protein B-H1                            NV2g025278000.1
Dlx	      Homeobox protein Dlx                             NV2g014640000.1
Dmbx1	    Diencephalon/mesencephalon homeobox protein 1	   NV2g003591000.1
Dmbx4   	Diencephalon/mesencephalon homeobox protein 1-B	 NV2g003592000.1
Dmbx5	    Diencephalon/mesencephalon homeobox protein 1-B	 NV2g003593000.1
Emx	      Homeobox protein EMX1	                           NV2g018403000.1
Gbx	      Gastrulation brain homeodomain                   NV2g011380000.1
Gsc	      Homeobox protein goosecoid	                     NV2g010922000.1
MoxB    	Homeobox protein Mox1	                           NV2g012940000.1
MoxC      Homeobox protein Mox1	                           NV2g012942000.1
MoxD    	Homeobox protein Mox1	                           NV2g012943000.1
Msx1	    Homeobox protein MSX-1                           NV2g012247000.1
Nkx2.2a1	Homeobox protein Nkx-2.2a                   	   NV2g011134000.1
Nkx2.2a2	Homeobox protein Nkx-2.2a	                       NV2g011262000.1
Nkx2.8    Homeobox protein Nkx-2.8                    	   NV2g011132000.1
Nkx3.2    Homeobox protein Nkx-3.2                         NV2g018723000.1
OtxA	    Homeobox protein OTX1	                           NV2g012793000.1
OtxB	    Homeobox protein OTX1 B                          NV2g012795000.1
OtxC	    Homeobox protein OTX1	                           NV2g012799000.1
Six4/5	  Homeobox protein SIX4/5	                         NV2g010728000.1
Tlx	      T-cell leukemia homeobox protein 2	             NV2g011935000.1
Uncx	    Homeobox protein unc-4 homolog                   NV2g006849000.1
Cdx	      HoxB3-like                                	     NV2g030001000.1
Lbx	      Transcription factor LBX2                   	   NV2g017667000.1

```{bash, eval = FALSE}
NV2g017442000.1
NV2g017961000.1
NV2g012279000.1
NV2g011327000.1
NV2g012004000.1
NV2g012007000.1
NV2g012008000.1
NV2g025278000.1
NV2g014640000.1
NV2g003591000.1
NV2g003592000.1
NV2g003593000.1
NV2g018403000.1
NV2g011380000.1
NV2g010922000.1
NV2g012940000.1
NV2g012942000.1
NV2g012943000.1
NV2g012247000.1
NV2g011134000.1
NV2g011262000.1
NV2g011132000.1
NV2g018723000.1
NV2g012793000.1
NV2g012795000.1
NV2g012799000.1
NV2g010728000.1
NV2g011935000.1
NV2g006849000.1
NV2g030001000.1
NV2g017667000.1
```

```{bash, eval = FALSE}
wget  https://genomes.stowers.org/files/pub/nematostella/Nvec/genomes/Nvec200/aligned/tcs_v2/20240221/NV2g.20240221.transcripts.fa

#grep each header with the protein sequence after (seqkit allows for flexibility wiht multi-line FASTAs) and save to a new file
seqkit grep --id-regexp "gene=([^ ]+)" -f Nematostella_Hox.txt NV2g.20240221.transcripts.fa -o Nematostella_Hox_seqs.txt
```


```{bash, eval = FALSE}
#request interactive computing node
salloc -p cpu -c 8 --mem 32G

#load BLAST module
module load uri/main
module load BLAST+/2.15.0-gompi-2023a

# These are nucleotide sequences, so will be blastx
# run blast with human-readable output
blastx -query Nematostella_Hox_seqs.txt -db ../../../references/blast_dbs/Pacuta_prot -out Nematostella_Hox_blast_results.txt -outfmt 0

#looks like there are a lot of matches for each gene! I am going to do a tab search with a very low e-value cutoff:

# run blast with tabular output
blastx -query Nematostella_Hox_seqs.txt -db ../../../references/blast_dbs/Pacuta_prot -out Nematostella_Hox_blast_results_tab_top3.txt -outfmt 6 -evalue 1e-20 -max_target_seqs 3

# reciprocal blast against Nematostella fasta to find best hit 

cut -f2 Nematostella_Hox_blast_results_tab_top3.txt | sort | uniq > Nematostella_Hox_blast_results.names
grep -A 1 -f Nematostella_Hox_blast_results.names ../../../references/Pocillopora_acuta_HIv2.genes.pep.faa > Nematostella_Hox_blast_results_Pacutaprot.txt
```

### Blast against Custom Nematostella BLASTP database:

```{bash, eval=FALSE}
#request interactive computing node
salloc -p cpu -c 8 --mem 32G

#load BLAST module
module load uri/main
module load BLAST+/2.15.0-gompi-2023a

cd ../references/blast_dbs
mkdir model_genomes
cd model_genomes

#download the genome protein fasta if you have not already
wget https://genomes.stowers.org/files/pub/nematostella/Nvec/genomes/Nvec200/aligned/tcs_v2/20240221/NV2g.20240221.protein.fa

cd ..

makeblastdb -in model_genomes/NV2g.20240221.protein.fa -out NVec_prot -dbtype prot

cd ../output_RNA/differential_expression/

# run blast with tabular output
blastp -query Nematostella_Hox_blast_results_Pacutaprot.txt -db ../../../references/blast_dbs/NVec_prot -out Nematostella_Hox_reciprocal_blast_results_tab_top1.txt -outfmt 6 -evalue 1e-20 -max_target_seqs 1
cut -f2 Nematostella_Hox_reciprocal_blast_results_tab_top1.txt | sort | uniq > Nematostella_Hox_reciprocal_blast_results_tab_top1_names.txt
grep -f  Nematostella_Hox_reciprocal_blast_results_tab_top1_names.txt NV2g.20240221.transcripts.fa > Nematostella_Hox_Pacuta_matches.txt
```

In R:

```{r}
Hox_genes <- read.csv("../output_RNA/differential_expression/blast/NV_Hox.csv", header = FALSE)
Hox_blast <- read.delim("../output_RNA/differential_expression/blast/Nematostella_Hox_reciprocal_blast_results_tab_top1.txt", header = FALSE)
Hox_transcripttogene <- read.delim("../output_RNA/differential_expression/blast/Nematostella_Hox_Pacuta_matches.txt", header = FALSE, sep=" ")

colnames(Hox_transcripttogene) <- c("transcript", "gene","CDs")
colnames(Hox_blast) <- c("Pocillopora_ID", "Nematostella_Transcript", "Perc_Identity", "Align_Length", "Mismatch", "Gap_Open",
                          "Q_start", "Q_end", "S_start", "S_end", "E_value", "Bit_score")
colnames(Hox_genes) <- c("Gene_Name", "Description", "NV2_ID", "UVienna_ID")

Hox_transcripttogene <- Hox_transcripttogene %>%
  mutate(transcript = gsub(">", "",transcript)) %>%
  mutate(gene = gsub("gene=", "",gene))

results = data.frame(Pacuta_gene = character(), 
                      Gene_Name = character(), 
                      Description = character(), 
                      NV2_ID = character(), 
                      UVienna_ID = character(), 
                      stringsAsFactors = FALSE)

for (i in 1:nrow(Hox_blast)){
  P_acuta = Hox_blast$Pocillopora_ID[i]
  Nt <- Hox_blast$Nematostella_Transcript[i]
  
  Ng <- Hox_transcripttogene$gene[Hox_transcripttogene$transcript == Nt]
  
  info <- Hox_genes[Hox_genes$NV2_ID==Ng,]
  info <- info %>% mutate(Pacuta_gene = P_acuta) %>% select(Pacuta_gene,everything())
  results <- rbind(results, info)
}

write.csv(results, file="../output_RNA/marker_genes/Hox_nematostella.csv")
```



## Code for accessing NCBI

```{bash, eval = FALSE}
# Read the input file line by line and fetch FASTA sequences
while read -r accession; do
  if [[ -n "$accession" ]]; then
    echo "Fetching $accession..."
    curl -s "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=${accession}&rettype=fasta&retmode=text" >> "YinYang.fasta"
    echo >> "YinYang.fasta"  # Add a newline between sequences
    sleep 1  # Avoid hitting rate limits
  fi
done < "YinYang.txt"
```


## Appendix: Program installations

### Install Orthofinder on unity

```{bash, eval = FALSE}
cd /work/pi_hputnam_uri_edu/pgrams

wget https://github.com/davidemms/OrthoFinder/releases/latest/download/OrthoFinder.tar.gz

tar xzvf OrthoFinder.tar.gz
cd OrthoFinder/

#print help 
./orthofinder -h
```

### Install Broccoli on unity

Set up conda environments and install program

```{bash, eval=FALSE}
# enter interactive session
salloc

# load conda
module load conda/latest

# create and activate environment
conda create --prefix /work/pi_hputnam_uri_edu/conda/envs/env-broccoli python=3.8 ete3
conda activate /work/pi_hputnam_uri_edu/conda/envs/env-broccoli

# clone broccoli repo into environment
cd /work/pi_hputnam_uri_edu/conda/envs/env-broccoli
git clone https://github.com/rderelle/Broccoli.git
cd Broccoli

# check that the install is sucessful by running the help command
python broccoli.py -help

# Make a symlink (shortcut) to home directory
ln -s /work/pi_hputnam_uri_edu/conda/envs/env-broccoli/Broccoli/broccoli.py ~/broccoli.py

# now it can run in any directory via
python ~/broccoli.py -help
```

