---
title: "Interproscan"
author: "Zoe Dellaert"
date: "2025-10-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Interproscan 

```{bash eval=FALSE}
cd /project/pi_hputnam_uri_edu/zdellaert/LaserCoral/scripts
nano DEG_05_interproscan.sh
```

```{bash eval=FALSE}
#!/usr/bin/env bash
#SBATCH --export=NONE
#SBATCH --nodes=1 --ntasks-per-node=24
#SBATCH --signal=2
#SBATCH --no-requeue
#SBATCH --mem=200GB
#SBATCH --error=../scripts/outs_errs/"%x_error.%j"
#SBATCH --output=../scripts/outs_errs/"%x_output.%j"
#SBATCH -t 2:00:00 --qos=short
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails

module load uri/main
module load InterProScan/5.73-104.0-foss-2024a

cd ../output_RNA/differential_expression #set working directory

mkdir -p interpro
cd interpro

interproscan.sh \
  -i ../DEG_05_seqs.txt \
  -f TSV,XML \
  -appl Pfam,SMART,CDD \
  -dp
  
interproscan.sh \
  -i ../expressed_genes_seqs.txt \
  -f TSV,XML \
  -appl Pfam,SMART,CDD \
  -dp
echo "Scan complete!" $(date)
```


```{r}
library(tidyverse)
library(ggplot2)
library(Polychrome)

#load in DESeq results
DESeq <- read.csv("../output_RNA/differential_expression/DESeq_results.csv", header = TRUE) %>% dplyr::rename("query" ="X")

#make dataframes of just differentially expressed genes for each LFC direction - filtering a little more stringent, abs(LFC) >2
DE_05 <- DESeq %>% filter(padj < 0.05 & abs(log2FoldChange) > 1)
DE_05_Aboral <- DESeq %>% filter(padj < 0.05 & log2FoldChange > 1)
DE_05_OralEpi <- DESeq %>% filter(padj < 0.05& log2FoldChange < -1)
```

```{r}
interpro <- read.delim("../output_RNA/differential_expression/interpro/expressed_genes_seqs.txt.tsv", header=FALSE) 
 
# https://interproscan-docs.readthedocs.io/en/v5/OutputFormats.html

colnames(interpro) <- c(
  "query",
  "seq_md5",
  "seq_length",
  "analysis",
  "hit_ID",
  "hit_name",
  "start", "stop",
  "e_value",
  "status",
  "date",
  "Interpro_accession",
  "Interpro_annotation",
  "GO_annotation",
  "pathway_annotation"
)

domains_plot <- interpro %>%
  filter(analysis == "Pfam") %>%
  select(query, hit_name, start, stop, seq_length) %>%
  distinct() %>%
  mutate(
    midpoint = (start + stop) / 2,
    width = stop - start
  ) %>% left_join(DESeq)

domains <- interpro %>% arrange(query,hit_ID) %>%
  filter(analysis == "Pfam") %>%
  select(query, hit_name) %>%
  distinct() 

head(domains)
```

### enrichment code based on levy et al 2025 -- testing for similarity of results

```{r}
#gsa_enrichment_load_pfam_list = function(pfam_architecture_file, do_unique = TRUE, do_remove_empty = TRUE, architecture_sep = " ") {
	
# load pfam species i
	pfm_i = read.table("~/Downloads/Pocdam_long.pep.pfamscan_archs.csv", col.names = c("gene","architecture"), sep = "\t")

pfm_i_l = lapply(pfm_i$architecture, function(v) unique(unlist(strsplit(v, split = " "))) )
names(pfm_i_l) = pfm_i$gene

pfm_i_l = pfm_i_l [ lengths(pfm_i_l) > 0 ]

```


```{r}
DE_05_OralEpi_top500 <- DE_05_OralEpi %>% arrange(log2FoldChange) %>% slice_head(n = 500)
annotation <- interpro %>%
  filter(analysis == "Pfam") %>%
  select(query, hit_name) %>%
  mutate(hit_name = str_replace_all(hit_name, " ", "_")) %>%
  group_by(query) %>%
  summarise(hit_names = list(unique(hit_name))) %>%
  deframe()  # converts to named list

genes_fg <- interpro %>% filter(query %in% DE_05_OralEpi_top500$query) %>% pull(query) %>% unique()
genes_bg <- names(annotation)

	# count distributions in background and foreground list
	dist_fg = as.data.frame(table(unlist(annotation [ genes_fg ])))
	dist_bg = as.data.frame(table(unlist(annotation)))
	
		# prepare table
		tab_in_all  = merge(dist_fg, dist_bg, by.x = "Var1", by.y = "Var1", all.x = TRUE, all.y = FALSE)
		tab_in_all  = tab_in_all[apply(tab_in_all!=0, 1, all),]
		colnames(tab_in_all) = c("annot","freq_in_fg","freq_in_bg")
		tab_in_all$total_annot_in_fg = length(genes_fg )
		tab_in_all$total_annot_in_bg = length(annotation)
		tab_in_all$fraction_in_fg = tab_in_all$freq_in_fg / tab_in_all$total_annot_in_fg
		tab_in_all$fraction_in_bg = tab_in_all$freq_in_bg / tab_in_all$total_annot_in_bg
		
		# hypergeometric test
		tab_in_all$pval = phyper(
			tab_in_all$freq_in_fg - 1,
			tab_in_all$freq_in_bg,
			tab_in_all$total_annot_in_bg - tab_in_all$freq_in_bg,
			tab_in_all$total_annot_in_fg,
			lower.tail = FALSE)
		
		# pvalue adjustment
		tab_in_all$pval_adj = p.adjust(tab_in_all$pval, method = "BH")
		
		# sort result by pvalue
		tab_in_all = tab_in_all [ order(tab_in_all$pval_adj) , ]
```


###  my enrichment code

```{r}
annotated_genes <- unique(domains$query)
length(annotated_genes)

annotated_genes_de <- domains %>% filter(query %in% DE_05$query) %>% pull(query) %>% unique()
length(annotated_genes_de)

domain_counts <- domains %>%
  mutate(is_DE = query %in% annotated_genes_de) %>%
  group_by(hit_name, is_DE) %>%
  summarize(n_genes = n_distinct(query), .groups = "drop") %>%
  pivot_wider(names_from = is_DE, values_from = n_genes, values_fill = 0) %>%
  rename(DE = `TRUE`, NonDE = `FALSE`) %>%
  mutate(Background = DE + NonDE) %>%
  filter(DE != 0)

# Hypergeometric enrichment
M <- length(annotated_genes)          # total annotated genes
N <- length(annotated_genes_de)           # number of DE genes

domain_enrichment <- domain_counts %>%
  rowwise() %>%
  mutate(
    k = DE,                     # number of DE genes with this domain
    K = Background,              # total number of genes with this domain
    p_value = phyper(q = DE - 1,
                     m = Background,
                     n = M - Background,
                     k = N,
                     lower.tail = FALSE)
  ) %>%
  ungroup() %>%
  mutate(padj = p.adjust(p_value, method = "BH")) %>%
  arrange(padj)

# Add enrichment ratio for interpretation
domain_enrichment <- domain_enrichment %>%
  mutate(
    DE_ratio = DE / N,
    Background_ratio = Background / M,
    enrichment_ratio = (DE / N) / (Background / M)
  )

# Filter significant or top N domains
top_domains <- domain_enrichment %>%
  filter(padj < 0.05, DE > 0) %>%
  arrange(padj) %>%
  slice_head(n = 20)

# Plot
ggplot(top_domains, aes(x = factor(hit_name, levels = rev(hit_name)), y = enrichment_ratio, color = -log10(padj),size = DE)) +
  geom_point() +
  coord_flip() +
  scale_fill_gradient(low = "skyblue", high = "darkblue") +
  labs(
    x = "Pfam Domain",
    y = "Enrichment Ratio (DE / Background)",
    fill = "-log10(padj)",
    title = "Pfam Domain Enrichment in Differentially Expressed Genes"
  ) +
  theme_minimal(base_size = 13) +
  theme(
    panel.grid.major.y = element_blank(),
    axis.text.y = element_text(size = 10)
  )
```

```{r}
annotated_genes <- unique(domains$query)
length(annotated_genes)

annotated_genes_de <- domains %>% filter(query %in% DE_05_Aboral$query) %>% pull(query) %>% unique()
length(annotated_genes_de)

domain_counts <- domains %>%
  mutate(is_DE = query %in% annotated_genes_de) %>%
  group_by(hit_name, is_DE) %>%
  summarize(n_genes = n_distinct(query), .groups = "drop") %>%
  pivot_wider(names_from = is_DE, values_from = n_genes, values_fill = 0) %>%
  rename(DE = `TRUE`, NonDE = `FALSE`) %>%
  mutate(Background = DE + NonDE) %>%
  filter(DE != 0)

# Hypergeometric enrichment
M <- length(annotated_genes)          # total annotated genes
N <- length(annotated_genes_de)           # number of DE genes

domain_enrichment <- domain_counts %>%
  rowwise() %>%
  mutate(
    k = DE,                     # number of DE genes with this domain
    K = Background,              # total number of genes with this domain
    p_value = phyper(q = DE - 1,
                     m = Background,
                     n = M - Background,
                     k = N,
                     lower.tail = FALSE)
  ) %>%
  ungroup() %>%
  mutate(padj = p.adjust(p_value, method = "BH")) %>%
  arrange(padj)

# Add enrichment ratio for interpretation
domain_enrichment <- domain_enrichment %>%
  mutate(
    DE_ratio = DE / N,
    Background_ratio = Background / M,
    enrichment_ratio = (DE / N) / (Background / M)
  )

# Filter significant or top N domains
top_domains <- domain_enrichment %>%
  filter(padj < 0.05, DE > 0) %>%
  arrange(padj) %>%
  slice_head(n = 20)

# Plot
ggplot(top_domains, aes(x = factor(hit_name, levels = rev(hit_name)), y = enrichment_ratio, color = -log10(padj),size = DE)) +
  geom_point() +
  coord_flip() +
  scale_fill_gradient(low = "skyblue", high = "darkblue") +
  labs(
    x = "Pfam Domain",
    y = "Enrichment Ratio (DE / Background)",
    fill = "-log10(padj)",
    title = "Pfam Domain Enrichment in Differentially Expressed Genes"
  ) +
  theme_minimal(base_size = 13) +
  theme(
    panel.grid.major.y = element_blank(),
    axis.text.y = element_text(size = 10)
  )
```

```{r}
annotated_genes <- unique(domains$query)
length(annotated_genes)

annotated_genes_de <- domains %>% filter(query %in% DE_05_OralEpi$query) %>% pull(query) %>% unique()
length(annotated_genes_de)

domain_counts <- domains %>%
  mutate(is_DE = query %in% annotated_genes_de) %>%
  group_by(hit_name, is_DE) %>%
  summarize(n_genes = n_distinct(query), .groups = "drop") %>%
  pivot_wider(names_from = is_DE, values_from = n_genes, values_fill = 0) %>%
  rename(DE = `TRUE`, NonDE = `FALSE`) %>%
  mutate(Background = DE + NonDE) %>%
  filter(DE != 0)

# Hypergeometric enrichment
M <- length(annotated_genes)          # total annotated genes
N <- length(annotated_genes_de)           # number of DE genes

domain_enrichment <- domain_counts %>%
  rowwise() %>%
  mutate(
    k = DE,                     # number of DE genes with this domain
    K = Background,              # total number of genes with this domain
    p_value = phyper(q = DE - 1,
                     m = Background,
                     n = M - Background,
                     k = N,
                     lower.tail = FALSE)
  ) %>%
  ungroup() %>%
  mutate(padj = p.adjust(p_value, method = "BH")) %>%
  arrange(padj)

# Add enrichment ratio for interpretation
domain_enrichment <- domain_enrichment %>%
  mutate(
    DE_ratio = DE / N,
    Background_ratio = Background / M,
    enrichment_ratio = (DE / N) / (Background / M)
  )

# Filter significant or top N domains
top_domains <- domain_enrichment %>%
  filter(padj < 0.05, DE > 0) %>%
  arrange(padj) %>%
  slice_head(n = 20)

# Plot
ggplot(top_domains, aes(x = factor(hit_name, levels = rev(hit_name)), y = enrichment_ratio, color = -log10(padj),size = DE)) +
  geom_point() +
  coord_flip() +
  scale_fill_gradient(low = "skyblue", high = "darkblue") +
  labs(
    x = "Pfam Domain",
    y = "Enrichment Ratio (DE / Background)",
    fill = "-log10(padj)",
    title = "Pfam Domain Enrichment in Differentially Expressed Genes"
  ) +
  theme_minimal(base_size = 13) +
  theme(
    panel.grid.major.y = element_blank(),
    axis.text.y = element_text(size = 10)
  )
```



```{r}
DE_05_OralEpi_top20 <- DE_05_OralEpi %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 10)
DE_05_Aboral_top20 <- DE_05_Aboral %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 10)
ZP_genes <- domains_plot %>% right_join(DESeq %>% select(query, log2FoldChange)) %>%
    filter(grepl("ZP-", hit_name))
  
gene_lists <- list(DE_05_OralEpi_top20,DE_05_Aboral_top20,ZP_genes)

for (list in gene_lists){
  
    plotting <- domains_plot %>%
      right_join(list %>% select(query, log2FoldChange)) %>%
      arrange(-abs(log2FoldChange)) %>%
      mutate(query = factor(query, levels = unique(query)))
    
    domain_colors <- createPalette(N = length(unique(plotting$hit_name[!is.na(plotting$hit_name)])),
                                   seedcolors = c("#4477AA", "#228833", "#EE6677", "#CCBB44", "#66CCEE", "#AA3377", "#BBBBBB"))
    
    domain_colors <- setNames(domain_colors, unique(plotting$hit_name[!is.na(plotting$hit_name)]))
    
    plot <- ggplot(plotting, aes(y = fct_rev(factor(query)))) +
      geom_segment(aes(x = 0, xend = seq_length, yend = fct_rev(factor(query))),
                   color = "gray90", linewidth = 1) +
      geom_rect(aes(
        xmin = start, xmax = stop,
        ymin = as.numeric(fct_rev(factor(query))) - 0.25,
        ymax = as.numeric(fct_rev(factor(query))) + 0.25,
        fill = hit_name
      ), color = "gray40", linewidth = 0.2) +
      scale_fill_manual(values = domain_colors) +
      scale_x_continuous(expand = expansion(mult = c(0.01, 0.05))) +
      labs(
        x = "Protein position (amino acids)",
        y = NULL,
        fill = "Pfam domain",
        title = "Domain architectures of top differentially expressed proteins in Oral Tissue (sorted by log2 fold change)"
      ) +
      theme_minimal(base_size = 7) +
      theme(
        panel.grid = element_blank(),
        axis.text.y = element_text(size = 5.5, face = "bold"),
        axis.text.x = element_text(size = 6),
        plot.title = element_text(size = 8, face = "bold", hjust = 0),
        legend.position = "bottom",
        legend.key.size = unit(0.3, "cm"),
        legend.text = element_text(size = 5)
      )
    
    print(plot)
}

plotting <- domains_plot %>%
  right_join(DESeq %>% select(query, log2FoldChange)) %>%
    filter(grepl("ZP-", hit_name)) %>% 
  arrange(-abs(log2FoldChange)) %>%
  mutate(query = factor(query, levels = unique(query)))

```



```{r}
DE_05_top20 <- DE_05_OralEpi %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 20) %>% 
                                  bind_rows(
                                              DE_05_Aboral %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 20))

# Define the Pfam IDs or names for ZP domains
zp_domains <- c("PF00100", "PF23344")

plotting <- domains_plot %>%
  right_join(DE_05_top20 %>% select(query, log2FoldChange)) %>%
  arrange(log2FoldChange) %>%
  mutate(query = factor(query, levels = unique(query))) %>%
  mutate(highlight = hit_ID %in% zp_domains)

ZP_top20_bothdir <- ggplot(plotting, aes(y = fct_rev(factor(query)))) +
  geom_segment(aes(x = 0, xend = seq_length, yend = fct_rev(factor(query))),
               color = "gray90", linewidth = 1) +
  geom_rect(aes(
    xmin = start, xmax = stop,
    ymin = as.numeric(fct_rev(factor(query))) - 0.25,
    ymax = as.numeric(fct_rev(factor(query))) + 0.25,
    fill = ifelse(highlight, hit_name, "Other"),
    alpha = highlight
  ),
  color = "gray40", linewidth = 0.2) +
  scale_alpha_manual(values = c(`TRUE` = 1, `FALSE` = 0.2), guide = "none") +
  scale_fill_manual(
    values = c(
      setNames(rep("#E64B35FF", length(unique(plotting$hit_name))), unique(plotting$hit_name)),
      "Other" = "gray90"
    ),
    guide = "none"
  ) +
  scale_x_continuous(expand = expansion(mult = c(0.01, 0.05))) +
  labs(
    x = "Protein position (amino acids)",
    y = NULL,
    title = "Highlighted Zona Pellucida domains in top DE Tissue proteins"
  ) +
  theme_minimal(base_size = 7) +
  theme(
    panel.grid = element_blank(),
    axis.text.y = element_text(size = 3.5, face = "bold"),
    axis.text.x = element_text(size = 6),
    plot.title = element_text(size = 8, face = "bold", hjust = 0),
    legend.position = "none"
  )

ZP_top20_bothdir
```

