---
title: "Interproscan"
author: "Zoe Dellaert"
date: "2025-10-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Interproscan 

```{bash eval=FALSE}
cd /project/pi_hputnam_uri_edu/zdellaert/LaserCoral/scripts
nano DEG_05_interproscan.sh
```

```{bash eval=FALSE}
#!/usr/bin/env bash
#SBATCH --export=NONE
#SBATCH --nodes=1 --ntasks-per-node=24
#SBATCH --signal=2
#SBATCH --no-requeue
#SBATCH --mem=200GB
#SBATCH --error=../scripts/outs_errs/"%x_error.%j"
#SBATCH --output=../scripts/outs_errs/"%x_output.%j"
#SBATCH -t 2:00:00 --qos=short
#SBATCH --mail-type=BEGIN,END,FAIL #email you when job starts, stops and/or fails

module load uri/main
module load InterProScan/5.73-104.0-foss-2024a

cd ../output_RNA/differential_expression #set working directory

mkdir -p interpro
cd interpro

interproscan.sh \
  -i ../DEG_05_seqs.txt \
  -f TSV,XML \
  -appl Pfam,SMART,CDD \
  -dp
  
interproscan.sh \
  -i ../expressed_genes_seqs.txt \
  -f TSV,XML \
  -appl Pfam,SMART,CDD \
  -dp
echo "Scan complete!" $(date)
```


```{r}
library(tidyverse)
library(ggplot2)
library(Polychrome)

#load in DESeq results
DESeq <- read.csv("../output_RNA/differential_expression/DESeq_results.csv", header = TRUE) %>% dplyr::rename("query" ="X")

#make dataframes of just differentially expressed genes for each LFC direction - filtering a little more stringent, abs(LFC) >2
DE_05 <- DESeq %>% filter(padj < 0.05 & abs(log2FoldChange) > 1)
DE_05_Aboral <- DESeq %>% filter(padj < 0.05 & log2FoldChange > 1)
DE_05_OralEpi <- DESeq %>% filter(padj < 0.05& log2FoldChange < -1)
```

```{r}
interpro <- read.delim("../output_RNA/differential_expression/interpro/expressed_genes_seqs.txt.tsv", header=FALSE) 
 
# https://interproscan-docs.readthedocs.io/en/v5/OutputFormats.html

colnames(interpro) <- c(
  "query",
  "seq_md5",
  "seq_length",
  "analysis",
  "hit_ID",
  "hit_name",
  "start", "stop",
  "e_value",
  "status",
  "date",
  "Interpro_accession",
  "Interpro_annotation",
  "GO_annotation",
  "pathway_annotation"
)

interpro <- interpro %>% filter(e_value < 0.01)

domains_plot <- interpro %>%
  filter(analysis == "Pfam") %>%
  select(query, hit_name, start, stop, seq_length) %>%
  distinct() %>%
  mutate(
    midpoint = (start + stop) / 2,
    width = stop - start
  ) %>% left_join(DESeq)

domains <- interpro %>% arrange(query,hit_ID) %>%
  filter(analysis == "Pfam") %>%
  select(query, hit_name) %>%
  distinct() 

head(domains)
```

```{r}
enrichmentplot <- function(domain_enrichment = domain_enrichment, n_top = 20, Tissue = "Oral Epidermis"){
  top_domains <- domain_enrichment %>%
  arrange(padj) %>%
  slice_head(n = n_top) %>%
  mutate(sig_level = case_when(
    padj < 0.001 ~ "padj < 0.001",
    padj < 0.01  ~ "padj < 0.01",
    padj < 0.05  ~ "padj < 0.05",
    TRUE         ~ "ns"
  ))
  
plot <- ggplot(top_domains, aes(x = reorder(hit_name, enrichment_ratio),y = enrichment_ratio)) +
  geom_segment(aes(xend = hit_name, y = 0, yend = enrichment_ratio),
               color = "grey80", linewidth = .8) +
  geom_point(aes(size = DE, color = sig_level)) +
  geom_text(aes(label = DE),
            hjust = 0.5, size = 3, color = "white",fontface = "bold") +
  coord_flip() +
  scale_color_manual(values = c("padj < 0.001" = "black","padj < 0.01" = "gray8","padj < 0.05" = "gray50", "ns" = "gray70"),
                     name = "Adjusted p-value") +
  scale_size_continuous(range = c(3, 8), name = "DE genes") +
  labs(
    x = NULL,
    y = "Enrichment Ratio (DE / Background)",
    title = paste("Pfam Domain Enrichment in DE Genes: ", Tissue)
  ) + theme_classic(base_size = 12)  + theme(plot.title = element_text(size = 13, face = "bold", hjust = 0.5),
                                           legend.margin = margin(l = 10),plot.margin = margin(10, 15, 10, 10))

print(plot)

plot2 <- ggplot(top_domains, aes(x = reorder(hit_name, DE_ratio),y = DE_ratio)) +
  geom_segment(aes(xend = hit_name, y = 0, yend = DE_ratio),
               color = "grey80", linewidth = .8) +
  geom_point(aes(size = DE, color = sig_level)) +
  geom_text(aes(label = DE),
            hjust = 0.5, size = 3, color = "white",fontface = "bold") +
  coord_flip() +
  scale_y_continuous(labels = scales::percent) + 
  scale_color_manual(values = c("padj < 0.001" = "black","padj < 0.01" = "gray8","padj < 0.05" = "gray50", "ns" = "gray70"),
                     name = "Adjusted p-value") +
  scale_size_continuous(range = c(3, 8), name = "DE genes") +
  labs(
    x = NULL,
    y = "% DE genes with Domain",
    title = paste("Pfam Domain Enrichment in DE Genes: ", Tissue)
  ) + theme_classic(base_size = 12)  + theme(plot.title = element_text(size = 13, face = "bold", hjust = 0.5),
                                           legend.margin = margin(l = 10),plot.margin = margin(10, 15, 10, 10))

print(plot2)
}
```

```{r}
annotated_genes <- unique(domains$query)
length(annotated_genes)

annotated_genes_de <- domains %>% filter(query %in% DE_05$query) %>% pull(query) %>% unique()
length(annotated_genes_de)

domain_counts <- domains %>%
  mutate(is_DE = query %in% annotated_genes_de) %>%
  group_by(hit_name, is_DE) %>%
  summarize(n_genes = n_distinct(query), .groups = "drop") %>%
  pivot_wider(names_from = is_DE, values_from = n_genes, values_fill = 0) %>%
  dplyr::rename(DE = `TRUE`, NonDE = `FALSE`) %>%
  mutate(Background = DE + NonDE) %>%
  filter(DE != 0)

# Hypergeometric enrichment
M <- length(annotated_genes)          # total annotated genes
N <- length(annotated_genes_de)           # number of DE genes

domain_enrichment_DE_05 <- domain_counts %>%
  rowwise() %>%
  mutate(p_value = phyper(q = DE - 1,
                     m = Background,
                     n = M - Background,
                     k = N,
                     lower.tail = FALSE)
  ) %>%
  ungroup() %>%
  mutate(padj = p.adjust(p_value, method = "BH")) %>%
  arrange(padj) %>%
  mutate(
    DE_ratio = DE / N,
    Background_ratio = Background / M,
    enrichment_ratio = (DE / N) / (Background / M)
  )

# Plot
enrichmentplot(domain_enrichment = domain_enrichment_DE_05, n_top = 20, Tissue = "All DE (both directions)")
```

```{r}
annotated_genes <- unique(domains$query)
length(annotated_genes)

annotated_genes_de <- domains %>% filter(query %in% DE_05_Aboral$query) %>% pull(query) %>% unique()
length(annotated_genes_de)

domain_counts <- domains %>%
  mutate(is_DE = query %in% annotated_genes_de) %>%
  group_by(hit_name, is_DE) %>%
  summarize(n_genes = n_distinct(query), .groups = "drop") %>%
  pivot_wider(names_from = is_DE, values_from = n_genes, values_fill = 0) %>%
  dplyr::rename(DE = `TRUE`, NonDE = `FALSE`) %>%
  mutate(Background = DE + NonDE) %>%
  filter(DE != 0)

# Hypergeometric enrichment
M <- length(annotated_genes)          # total annotated genes
N <- length(annotated_genes_de)           # number of DE genes

domain_enrichment_Aboral <- domain_counts %>%
  rowwise() %>%
  mutate(p_value = phyper(q = DE - 1,
                     m = Background,
                     n = M - Background,
                     k = N,
                     lower.tail = FALSE)
  ) %>%
  ungroup() %>%
  mutate(padj = p.adjust(p_value, method = "BH")) %>%
  arrange(padj) %>%
  mutate(
    DE_ratio = DE / N,
    Background_ratio = Background / M,
    enrichment_ratio = (DE / N) / (Background / M)
  )

# Plot
enrichmentplot(domain_enrichment = domain_enrichment_Aboral, n_top = 10, Tissue = "Aboral")
```

```{r}
annotated_genes <- unique(domains$query)
length(annotated_genes)

annotated_genes_de <- domains %>% filter(query %in% DE_05_OralEpi$query) %>% pull(query) %>% unique()
length(annotated_genes_de)

domain_counts <- domains %>%
  mutate(is_DE = query %in% annotated_genes_de) %>%
  group_by(hit_name, is_DE) %>%
  summarize(n_genes = n_distinct(query), .groups = "drop") %>%
  pivot_wider(names_from = is_DE, values_from = n_genes, values_fill = 0) %>%
  dplyr::rename(DE = `TRUE`, NonDE = `FALSE`) %>%
  mutate(Background = DE + NonDE) %>%
  filter(DE != 0)

# Hypergeometric enrichment
M <- length(annotated_genes)          # total annotated genes
N <- length(annotated_genes_de)           # number of DE genes

domain_enrichment_Oral <- domain_counts %>%
  rowwise() %>%
  mutate(p_value = phyper(q = DE - 1,
                     m = Background,
                     n = M - Background,
                     k = N,
                     lower.tail = FALSE)
  ) %>%
  ungroup() %>%
  mutate(padj = p.adjust(p_value, method = "BH")) %>%
  arrange(padj) %>%
  mutate(
    DE_ratio = DE / N,
    Background_ratio = Background / M,
    enrichment_ratio = (DE / N) / (Background / M)
  )

# Plot
enrichmentplot(domain_enrichment = domain_enrichment_Oral, n_top = 10, Tissue = "Oral Epidermis")
```

```{r}
library(scales)

enrichment_combined <- full_join(
  domain_enrichment_Oral,
  domain_enrichment_Aboral,
  by = c("hit_name", "Background"),
  suffix = c("_Oral", "_Aboral")) %>%
  mutate(
    DE_Oral = replace_na(DE_Oral, 0),
    DE_ratio_Oral = replace_na(DE_ratio_Oral, 0),
    enrichment_ratio_Oral = replace_na(enrichment_ratio_Oral, 0),
    padj_Oral = replace_na(padj_Oral, 1),
    DE_Aboral = replace_na(DE_Aboral, 0),
    DE_ratio_Aboral = replace_na(DE_ratio_Aboral, 0),
    enrichment_ratio_Aboral = replace_na(enrichment_ratio_Aboral, 0),
    padj_Aboral = replace_na(padj_Aboral, 1),
    nonDE_Oral = Background - DE_Oral - DE_Aboral,
    nonDE_Aboral = Background - DE_Oral - DE_Aboral)

write.csv(enrichment_combined, "../output_RNA/differential_expression/interpro/Pfam_enrichment_results.csv")

enrichment_combined$hit_name[
  enrichment_combined$hit_name == "ATP:guanido phosphotransferase, C-terminal catalytic domain"
] <- "ATP:guanido phosphotransferase,\nC-terminal catalytic domain"

enrichmentplot_bidirectional <- function(enrichment_combined, n_top = 20) {
  enrichment_combined_long <- enrichment_combined %>%
    mutate(
      sig_level_Oral = case_when(
        padj_Oral < 0.001 ~ "padj < 0.001",
        padj_Oral < 0.01  ~ "padj < 0.01",
        padj_Oral < 0.05  ~ "padj < 0.05",
        TRUE ~ "ns"
      ),
      sig_level_Aboral = case_when(
        padj_Aboral < 0.001 ~ "padj < 0.001",
        padj_Aboral < 0.01  ~ "padj < 0.01",
        padj_Aboral < 0.05  ~ "padj < 0.05",
        TRUE ~ "ns"
      )
    ) %>%
    select(
      hit_name,
      DE_Oral,
      DE_ratio_Oral,
      padj_Oral,
      sig_level_Oral,
      DE_Aboral,
      DE_ratio_Aboral,
      padj_Aboral,
      sig_level_Aboral
    ) %>%
    pivot_longer(
      cols = -hit_name,
      names_to = c(".value", "Tissue"),
      names_pattern = "(.*)_(Oral|Aboral)"
    )
  
  top_oral <- enrichment_combined_long %>%
    filter(Tissue == "Oral") %>%
    arrange(padj) %>%
    slice_head(n = n_top) %>%
    arrange(desc(DE)) %>%
    pull(hit_name)
  
  top_aboral <- enrichment_combined_long %>%
    filter(Tissue == "Aboral") %>%
    arrange(padj) %>%
    slice_head(n = n_top) %>%
    arrange(DE) %>%
    pull(hit_name)
  
  top_domains_both <- enrichment_combined_long %>%
    filter(hit_name %in% union(top_oral, top_aboral)) %>%
    mutate(
      DE_ratio = ifelse(Tissue == "Oral", -DE_ratio, DE_ratio),
      sig_tissue = paste(Tissue, sig_level),
      hit_name = factor(hit_name, levels = rev(c(
        top_oral, setdiff(top_aboral, top_oral)
      )))
    )
  
  max_val <- max(abs(top_domains_both$DE_ratio))
  
  # Plot both tissues with facetting
  p <-  ggplot(top_domains_both, aes(x = hit_name, y = DE_ratio)) +
    geom_hline(
      yintercept = 0,
      color = "black",
      linewidth = 0.6,
      linetype = "dashed") +
    geom_segment(
      aes(xend = hit_name, y = 0, yend = DE_ratio),
      color = "grey85",
      linewidth = 0.6) +
    geom_point(aes(size = DE, fill = sig_tissue),
               shape = 21,
               alpha = 0.9) +
    geom_text(
      aes(label = ifelse(DE == 0, "", DE)),
      size = 3,
      color = "black",
      fontface = "bold") +
    coord_flip() +
    scale_fill_manual(
      name = "Adjusted p-value × Tissue",
      values = c(
        "Oral padj < 0.001" = "#FD8D3C",
        "Oral padj < 0.01"  = "#FDAE6B",
        "Oral padj < 0.05"  = "#FEC69D",
        "Aboral padj < 0.001" = "mediumpurple1",
        "Aboral padj < 0.01"  = "#CBA7F8",
        "Aboral padj < 0.05"  = "#D4C0FF",
        "Oral ns" = "gray80",
        "Aboral ns" = "gray80")) +
    scale_y_continuous(
      labels = function(x)
        percent(abs(x), accuracy = 1),
      breaks = seq(-round(max_val, 1), round(max_val, 1), by = 0.025),
      minor_breaks = seq(-round(max_val, 1), round(max_val, 1), by = 0.01),
      limits = c(-max_val, max_val)) +
    scale_size_continuous(range = c(3, 12), name = "DE genes") +
    labs(
      x = NULL,
      y = "% Upregulated genes in tissue containing Domain",
      title = paste0(
        "Pfam Domain Enrichment in DE Genes: Oral vs Aboral (Top ",n_top," per tissue)")) +
    theme_classic(base_size = 12) +
    theme(plot.title = element_text(
        size = 13,
        face = "bold",
        hjust = 0.5),
      legend.position = "right",
      plot.margin = margin(10, 15, 10, 10),
      panel.grid.major.x = element_line(color = "grey80"))
  
  print(p)
}

enrichmentplot_bidirectional_v2 <- function(enrichment_combined, n_top = 20) {
  enrichment_combined_long <- enrichment_combined %>%
    mutate(
      sig_level_Oral = case_when(
        padj_Oral < 0.001 ~ "padj < 0.001",
        padj_Oral < 0.01  ~ "padj < 0.01",
        padj_Oral < 0.05  ~ "padj < 0.05",
        TRUE ~ "ns"),
      sig_level_Aboral = case_when(
        padj_Aboral < 0.001 ~ "padj < 0.001",
        padj_Aboral < 0.01  ~ "padj < 0.01",
        padj_Aboral < 0.05  ~ "padj < 0.05",
        TRUE ~ "ns")) %>%
    select(
      hit_name, nonDE_Oral,nonDE_Aboral,
      DE_Oral,
      enrichment_ratio_Oral,
      padj_Oral,
      sig_level_Oral,
      DE_Aboral,
      enrichment_ratio_Aboral,
      padj_Aboral,
      sig_level_Aboral) %>%
    pivot_longer(
      cols = -hit_name,
      names_to = c(".value", "Tissue"),
      names_pattern = "(.*)_(Oral|Aboral)")
  
  top_oral <- enrichment_combined_long %>%
    filter(Tissue == "Oral") %>%
    arrange(padj) %>%
    slice_head(n = n_top) %>%
    arrange(desc(enrichment_ratio)) %>%
    pull(hit_name)
  
  top_aboral <- enrichment_combined_long %>%
    filter(Tissue == "Aboral") %>%
    arrange(padj) %>%
    slice_head(n = n_top) %>%
    arrange(enrichment_ratio) %>%
    pull(hit_name)
  
  top_domains_both <- enrichment_combined_long %>%
    filter(hit_name %in% union(top_oral, top_aboral)) %>%
    mutate(enrichment_ratio = ifelse(Tissue == "Oral", -enrichment_ratio, enrichment_ratio),
      sig_tissue = paste(Tissue, sig_level),
      hit_name = factor(hit_name, levels = rev(c(top_oral, setdiff(top_aboral, top_oral)))))
  
  max_val <- ceiling(max(top_domains_both$enrichment_ratio[top_domains_both$enrichment_ratio>0])/5) * 5
  min_val <- ceiling(max(abs(top_domains_both$enrichment_ratio[top_domains_both$enrichment_ratio<0]))/5) * 5
  
  # Plot both tissues with facetting
  p <-  ggplot(top_domains_both, aes(x = hit_name, y = enrichment_ratio)) +
    geom_hline(
      yintercept = 0,
      color = "black",
      linewidth = 0.6,
      linetype = "dashed") +
    geom_segment(
      aes(xend = hit_name, y = 0, yend = enrichment_ratio),
      color = "grey85",
      linewidth = 0.6) +
    geom_point(aes(size = DE, fill = sig_tissue),shape = 21,alpha = 0.9) +
    geom_text(aes(label = ifelse(DE == 0, "", DE)),size = 3,color = "black",fontface = "bold") +
    geom_text(aes(y = -min_val-2, label = nonDE),size = 3,color = "black") +
  # Text annotation
  annotate("text", 
    x = length(unique(top_domains_both$hit_name)) + 0.75,y = -min_val - 2, 
    label = "# of non-\nDE genes",
    vjust = 0.5,  hjust = 0.5,  # Centered horizontally
    size = 2.25,fontface = "bold", color = "black",lineheight = 0.9) +
  coord_flip(clip = "off") + 
    scale_fill_manual(
      name = "Adjusted p-value × Tissue",
      values = c(
        "Oral padj < 0.001" = "#FD8D3C",
        "Oral padj < 0.01"  = "#FDAE6B",
        "Oral padj < 0.05"  = "#FEC69D",
        "Aboral padj < 0.001" = "mediumpurple1",
        "Aboral padj < 0.01"  = "#CBA7F8",
        "Aboral padj < 0.05"  = "#D4C0FF",
        "Oral ns" = "gray80",
        "Aboral ns" = "gray80" )) +
    scale_y_continuous( labels = function(x)
        abs(x), breaks = seq(-min_val, max_val, by = 5),
      limits = c(-min_val-2, max_val)) +
    scale_size_continuous(range = c(3, 10), name = "DE genes") +
    labs(
      x = NULL,
      y = "Enrichment Ratio (Domains DE / Domains in Background)",
      title = paste0(
        "Pfam Domain Enrichment in DE Genes: Oral vs Aboral (Top ",
        n_top,
        " per tissue)")) +
    theme_classic(base_size = 12) +
    theme(plot.title = element_text(size = 13,face = "bold",hjust = 0.5),legend.position = "right",
      plot.margin = margin(10, 10, 10, 10),
      panel.grid.major.x = element_line(color = "grey80"))
  
  print(p)
}

enrichmentplot_bidirectional(enrichment_combined, n_top = 10)

plot <- enrichmentplot_bidirectional_v2(enrichment_combined, n_top = 10)

print(plot)
ggsave(filename = "../output_RNA/differential_expression/Pfam_enriched_top10.png", plot = plot, width = 10, height = 7, units = "in", dpi = 300,bg=NULL)
ggsave(filename = "../output_RNA/differential_expression/pdf_figs/Pfam_enriched_top10.png.pdf", plot = plot, width = 10, height = 7, units = "in", dpi = 300,bg=NULL)
```


```{r}
DE_05_OralEpi_top20 <- DE_05_OralEpi %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 20)
DE_05_Aboral_top20 <- DE_05_Aboral %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 10)
ZP_genes <- domains_plot %>% right_join(DESeq %>% select(query, log2FoldChange)) %>%
    filter(grepl("ZP-", hit_name))
  
gene_lists <- list(DE_05_OralEpi_top20,DE_05_Aboral_top20,ZP_genes)

for (list in gene_lists){
  
    plotting <- domains_plot %>%
      right_join(list %>% select(query, log2FoldChange)) %>%
      arrange(-abs(log2FoldChange)) %>%
      mutate(query = factor(query, levels = unique(query)))
    
    domain_colors <- createPalette(N = length(unique(plotting$hit_name[!is.na(plotting$hit_name)])),
                                   seedcolors = c("#4477AA", "#228833", "#EE6677", "#CCBB44", "#66CCEE", "#AA3377", "#BBBBBB"))
    
    domain_colors <- setNames(domain_colors, unique(plotting$hit_name[!is.na(plotting$hit_name)]))
    
    plot <- ggplot(plotting, aes(y = fct_rev(factor(query)))) +
      geom_segment(aes(x = 0, xend = seq_length, yend = fct_rev(factor(query))),
                   color = "gray90", linewidth = 1) +
      geom_rect(aes(
        xmin = start, xmax = stop,
        ymin = as.numeric(fct_rev(factor(query))) - 0.25,
        ymax = as.numeric(fct_rev(factor(query))) + 0.25,
        fill = hit_name
      ), color = "gray40", linewidth = 0.2) +
      scale_fill_manual(values = domain_colors) +
      scale_x_continuous(expand = expansion(mult = c(0.01, 0.05))) +
      labs(
        x = "Protein position (amino acids)",
        y = NULL,
        fill = "Pfam domain",
        title = "Domain architectures of top differentially expressed proteins in Oral Tissue (sorted by log2 fold change)") +
      theme_minimal(base_size = 7) +
      theme(
        panel.grid = element_blank(),
        axis.text.y = element_text(size = 5.5, face = "bold"),
        axis.text.x = element_text(size = 6),
        plot.title = element_text(size = 8, face = "bold", hjust = 0),
        legend.position = "bottom",
        legend.key.size = unit(0.3, "cm"),
        legend.text = element_text(size = 5)
      )
    
    print(plot)
}

plotting <- domains_plot %>%
  right_join(DESeq %>% select(query, log2FoldChange)) %>%
    filter(grepl("ZP-", hit_name)) %>% 
  arrange(-abs(log2FoldChange)) %>%
  mutate(query = factor(query, levels = unique(query)))

```

```{r}
# Define the Pfam IDs or names for ZP domains
zp_genes <- domains %>% filter(hit_name == "ZP-N domain" | hit_name == "ZP-C domain")

annots <- read.delim("../references/annotation/protein-GO.tsv")

zp_genes <- domains %>% filter(hit_name == "ZP-N domain" | hit_name == "ZP-C domain") %>% left_join(annots) %>% left_join(DESeq)
```


```{r}
DE_05_top20 <- DE_05_OralEpi %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 20) %>% 
                                  bind_rows(
                                              DE_05_Aboral %>% arrange(-abs(log2FoldChange)) %>% slice_head(n = 20))

# Define the Pfam IDs or names for ZP domains
zp_domains <- c("ZP-N domain", "ZP-C domain")

plotting <- domains_plot %>%
  right_join(DE_05_top20 %>% select(query, log2FoldChange)) %>%
  arrange(log2FoldChange) %>%
  mutate(query = factor(query, levels = unique(query))) %>%
  mutate(highlight = hit_name %in% zp_domains)

ZP_top20_bothdir <- ggplot(plotting, aes(y = fct_rev(factor(query)))) +
  geom_segment(aes(x = 0, xend = seq_length, yend = fct_rev(factor(query))),
               color = "gray90", linewidth = 1) +
  geom_rect(aes(
    xmin = start, xmax = stop,
    ymin = as.numeric(fct_rev(factor(query))) - 0.25,
    ymax = as.numeric(fct_rev(factor(query))) + 0.25,
    fill = ifelse(highlight, hit_name, "Other"),
    alpha = highlight
  ),
  color = "gray40", linewidth = 0.2) +
  scale_alpha_manual(values = c(`TRUE` = 1, `FALSE` = 0.2), guide = "none") +
  scale_fill_manual(
    values = c(
      setNames(rep("#E64B35FF", length(unique(plotting$hit_name))), unique(plotting$hit_name)),
      "Other" = "gray90"
    ),
    guide = "none"
  ) +
  scale_x_continuous(expand = expansion(mult = c(0.01, 0.05))) +
  labs(
    x = "Protein position (amino acids)",
    y = NULL,
    title = "Highlighted Zona Pellucida domains in top DE Tissue proteins"
  ) +
  theme_minimal(base_size = 7) +
  theme(
    panel.grid = element_blank(),
    axis.text.y = element_text(size = 3.5, face = "bold"),
    axis.text.x = element_text(size = 6),
    plot.title = element_text(size = 8, face = "bold", hjust = 0),
    legend.position = "none"
  )

ZP_top20_bothdir
```